
---
title: 'Adipose Precursor: Analysis Part2 - pseudotime'
author: "Mike Morgan"
output: html_notebook
---

## Introduction
We have seen so far that the structure of these adipocyte precursors is complex.  There are not any clearly distinct sub-populations in the data that would indicate different cell types.  We expect these cells to be following some (or several) developmental trajectories, but we have little information on what markers should be found at either end of that developmental program.

I'll try to construct some meaningful pseudotime trajectories on these data.  First I'll use all 930 cells, and then split them into the GFP+ and GFP- groups to see if there are any qualitative and quantitative differences in these trajectories for these cells.

```{r, echo=FALSE, message=F, warning=FALSE}
suppressPackageStartupMessages(library(org.Mm.eg.db))
suppressPackageStartupMessages(library(reshape2))
suppressPackageStartupMessages(library(WGCNA))
suppressPackageStartupMessages(library(lsa))
suppressPackageStartupMessages(library(biomaRt))
suppressPackageStartupMessages(library(destiny))
suppressPackageStartupMessages(library(flashClust))
suppressPackageStartupMessages(library(ggrepel))
suppressPackageStartupMessages(library(pheatmap))
suppressPackageStartupMessages(library(cowplot))
suppressPackageStartupMessages(library(GGally))
suppressPackageStartupMessages(library(goseq))
suppressPackageStartupMessages(library(scales))
suppressPackageStartupMessages(library(statmod))
suppressPackageStartupMessages(library(limma))
suppressPackageStartupMessages(library(stringr))
suppressPackageStartupMessages(library(RColorBrewer))
suppressPackageStartupMessages(library(scatterplot3d))

source("~/Dropbox/R_sessions/GGMike/theme_mike.R")
source("~/Dropbox/R_sessions/Clustering/CosineKNN.R")
source("~/Dropbox/R_sessions/SingleCellFunctions/single_cell_functions.R")

# load in all data and meta data
adipose.nz <- read.table("~/Dropbox/Adipose/Adipose-SF_norm.tsv",
                         sep="\t", h=T, stringsAsFactors=F)
rownames(adipose.nz) <- adipose.nz$gene_id
adipose.nz <- adipose.nz[, -ncol(adipose.nz)]

meta.merge <- read.table("~/Dropbox/Adipose/Adipose-metaMerge.tsv",
                         sep="\t", h=T, stringsAsFactors=F)

# correct the cluster labels
meta.merge$Cluster[meta.merge$Cluster == 1] <- 5
meta.merge$Cluster[meta.merge$Cluster == 2] <- 1
meta.merge$Cluster[meta.merge$Cluster == 5] <- 2

HVG.df <- read.table("~/Dropbox/Adipose/Adipose-HVG.tsv",
                  sep="\t", h=T, stringsAsFactors=F)

# HVG.df <- find_hvg(adipose.nz, return.ranks=TRUE, p.threshold=1e-2)

adipose.hvg <- adipose.nz[HVG.df$HVG, ]
adipose.hvg[is.na(adipose.hvg)] <- 0

adipose.pca <- prcomp(t(adipose.hvg), scale=T, center=T)
var.exp <- adipose.pca$sdev**2/sum(adipose.pca$sdev**2)
sum.var <- 0
n.comps <- 0
for(x in 1:length(var.exp)){
  while(sum.var < 0.75){
    sum.var <- sum.var + var.exp[x]
    n.comps <- n.comps + 1
  }
}

adipose.pcs <- as.data.frame(t(adipose.pca$x[, 1:n.comps]))

ensembl <- useEnsembl(biomart='ensembl', dataset='mmusculus_gene_ensembl', GRCh=37)

gene_symbol <- getBM(attributes=c('ensembl_gene_id', 'external_gene_name'),
                     filters='ensembl_gene_id', mart=ensembl,
                     values=rownames(adipose.nz))
rownames(gene_symbol) <- gene_symbol$ensembl_gene_id

adipose.nz$gene_id <- rownames(adipose.nz)
adipose.merge <- merge(adipose.nz, gene_symbol, by.x="gene_id", by.y="ensembl_gene_id", all.x=TRUE)
adipose.merge$external_gene_name[adipose.merge$gene_id == "eGFP"] <- "GFP"

marker.genes <- c("GFP", "Pparg", "Cebpa", "Bmpr1a", "Bmpr1b", "Bmpr2", "Cd34", "Pecam1", "Ly6a",
                  "Dlk1", "Tcf21", "Prdm16", "Foxc2", "Ndn", "Wt1", "HoxC9", "Shox2", "En1", "Sfrp2",
                  "Hoxa5", "Hoxa4", "Nr2f1", "Thbd", "Pappa", "Sfrp1", "Agt", "Cd24a", "Itgb1", "Fabp4",
                  "Pdgfra", "Pdgfrb", "Cspg4", "Msln", "Upk3b")
marker.exprs <- adipose.merge[adipose.merge$external_gene_name %in% marker.genes, ]
marker.trans <- as.data.frame(apply(as.data.frame(t(marker.exprs)[2:931 ,]),
                                    2, as.numeric))
colnames(marker.trans) <- marker.exprs$external_gene_name
marker.trans$Sample <- colnames(marker.exprs)[2:931]
adipose.nz <- adipose.nz[, -dim(adipose.nz)[2]]

cycle.labels <- read.table("~/Dropbox/Adipose/Adipose-CellCycle_labels.tsv",
                           h=T, sep="\t", stringsAsFactors=F)
meta.merge <- merge(meta.merge, cycle.labels, by='Sample')
meta.merge$Cluster <- as.factor(meta.merge$Cluster)
#meta.merge
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
# read in the hallmarks and map the human gene IDs onto mouse homologs
# pull in the c4 and c6 gene sets
source("~/Dropbox/R_sessions/src/generic_functions.R")
hallmark.geneset <- read.table("~/Dropbox/Genesets/hallmark.c2_ensemblgeneIds.txt",
                         sep="\t", stringsAsFactors=FALSE, h=T)

ortho.gene_symbol <- getBM(attributes=c('ensembl_gene_id', 'external_gene_name', 'hsapiens_homolog_orthology_type',
                                  'hsapiens_homolog_ensembl_gene'),
                     filters='ensembl_gene_id', mart=ensembl,
                     values=rownames(adipose.nz))
# remove many 2 one orthologs and many2many that are duplicated

# map the human gene IDs to the mouse ones
hallmark.merge <- merge(hallmark.geneset, ortho.gene_symbol, by.x='ensembl_gene_id', by.y='hsapiens_homolog_ensembl_gene')
colnames(hallmark.merge) <- c("hsapiens_ensembl", "GeneSet", "ensembl_gene_id", "external_gene_name", "hsapiens_orthology_type")

# this might take a while!
hallmark.go <- map2go(hallmark.merge)
```


### Pseudotemporal ordering of cells
The first thing to say, is that is always possible to create a pseudotemporal ordering of cells, what matters is whether it is biologically meaningful!  In the absence of markers for early and late adipocyte precursors this might be a problem, and so needs to be kept in mind at all times.

```{r, echo=FALSE}
#set.seed(24021982)
set.seed(42)
# remove the mult-cell controls and turquoise cells 
multi.cell <- meta.merge$Sample[meta.merge$SampleType == "Multi-cell"]
# adipose.dm <- DiffusionMap(t(adipose.hvg[, !colnames(adipose.hvg) %in% multi.cell]), n_eigs=3, k=10,
#                            distance='cosine', sigma='local')
adipose.dm <- DiffusionMap(t(adipose.pcs[1:10, !colnames(adipose.hvg) %in% multi.cell]),
                           n_eigs=5, k=13, distance='cosine')
# how many diffusion components do we need?
plot(eigenvalues(adipose.dm), pch=20)

plot(adipose.dm)
```

This is the raw diffusion map constructed from the top 30 principal components of the HVG X cell matrix, using the Euclidean distance between cells.  It looks like a trajectory of cells from left to right, with some bifurcation.  This is just the first step in constructing a pseudotemporal ordering, but it does look like there is at least one trajectory, possible with two branches.  It is hard to discern exactly with this plot in 3 dimensions.

```{r, echo=FALSE, message=FALSE, warning=FALSE}
adipose.dpt <- DPT(adipose.dm)
adipose.dc <- do.call(cbind.data.frame,
                      list("DC1"=adipose.dpt$DC1,
                           "DC2"=adipose.dpt$DC2,
                           "DC3"=adipose.dpt$DC3,
                           "DPT1"=adipose.dpt$DPT1,
                           "DPT2"=adipose.dpt$DPT2,
                           "DPT3"=adipose.dpt$DPT3))
#colnames(adipose.dc) <- paste0("DC", 1:dim(adipose.dc)[2])
rownames(adipose.dc) <- colnames(adipose.pcs[, !colnames(adipose.hvg) %in% multi.cell])
adipose.dc$DPT <- adipose.dpt$DPT1
adipose.dc$Sample <- colnames(adipose.hvg)[!colnames(adipose.hvg) %in% multi.cell]
dc.merge <- merge(adipose.dc, marker.trans, by='Sample')
dc.merge <- merge(dc.merge, meta.merge, by='Sample')

write.table(dc.merge,
            "~/Dropbox/Adipose/Adipose-DiffusionPseudotime.tsv",
            sep="\t", quote=F, row.names=F)

col.map <- col2hcl(unique(dc.merge$Cluster))
#names(col.map) <- (unique(dc.merge$Cluster))
names(col.map) <- c(1, 3, 2, 4)
```


```{r}
plot(adipose.dpt, root=3, paths_to=c(1, 2), pch=20)
```

How many branches are there?  This looks like 2 separate, parallel, trajectories.  One goes from cluster 2 to cluster one (yellow to turquoise), and another bifurcates from cluster 2 into clusters 3 and 4.

```{r}
adi.ptime <- ggplot(dc.merge, aes(x=DC1, y=DC2, fill=DPT1)) +
  geom_point(size=3, shape=21) + theme_minimal() +
  scale_fill_gradient(low="#FAFF60", high="#8615B0") +
  labs(x="Diffusion Component 1", y="Diffusion Component 2") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=17, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold")) +
  guides(fill=guide_colourbar(title="Diffusion\nPseudotime"))

ggsave(adi.ptime,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Pseudotime_DCs-scatter.png",
       height=5.5, width=7.75, dpi=300)

adi.ptime
```

Create a 3d scatter plot to visualise the first 3 diffusion components together.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
# map individual cells to clusters with the relevant colour
col.df <- do.call(cbind.data.frame,
                  list("Cluster"=names(col.map),
                       "Colour"=col.map))
dc.merge <- merge(dc.merge, col.df, by='Cluster')
```


```{r}
ggplot(dc.merge, aes(x=DC1, y=DC2, fill=Cluster)) +
  geom_point(size=3, shape=21) + theme_minimal() +
  scale_fill_manual(values=col.map) +
  labs(x="Diffusion Component 1", y="Diffusion Component 2") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=17, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14)) +
  guides(fill=guide_legend(override.aes=aes(size=8)))
```



```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=7.25, fig.width=7.25}
png("~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/SupplementaryFigure_3D-diffusion.png",
    height=7.25, width=7.25, res=300, units="in")
scatterplot3d(x=dc.merge[, 4], y=dc.merge[, 5], z=dc.merge[,3], bg=col2hcl(dc.merge$Cluster), 
              pch=21, xlab="Diffusion Component 2", ylab="Diffusion Component 3",
              zlab="Diffusion Component 1")
legend("right", legend = names(col.map),
      pt.bg =  col.map, 
      pch = 21, ncol=1, title="Cluster",
      inset = 0, xpd = TRUE, horiz = FALSE)
dev.off()
```

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=7.25, fig.width=7.25}
scatterplot3d(x=dc.merge[, 4], y=dc.merge[, 5], z=dc.merge[,3], bg=col2hcl(dc.merge$Cluster), 
              pch=21, xlab="Diffusion Component 2", ylab="Diffusion Component 3",
              zlab="Diffusion Component 1")
legend("right", legend = names(col.map),
      pt.bg =  col.map, 
      pch = 21, ncol=1, title="Cluster",
      inset = 0, xpd = TRUE, horiz = FALSE)
```

```{r, echo=FALSE, warning=FALSE, message=FALSE}
gfp.cols <- c("#13CF00", "#CC8EF4")
names(gfp.cols) <- c("GFP+", "GFP-")
gfp.col.mat <- cbind.data.frame("GFP.Colour"=gfp.cols,
                                "Marker"=names(gfp.cols))

dc.merge <- merge(dc.merge, gfp.col.mat, by='Marker')
dc.merge$GFP.Colour <- as.character(dc.merge$GFP.Colour)
```

```{r, echo=FALSE, warning=FALSE, message=FALSE}
png("~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/SupplementaryFigure_3D-diffusion-byGFP.png",
    height=7.25, width=7.25, res=300, units="in")
scatterplot3d(x=dc.merge[, 5], y=dc.merge[, 6], z=dc.merge[, 4], bg=dc.merge$GFP.Colour, 
              pch=21, xlab="Diffusion Component 2", ylab="Diffusion Component 3",
              zlab="Diffusion Component 1")
legend("right", legend = names(gfp.cols),
      pt.bg =  gfp.cols, 
      pch = 21, ncol=1, title="Marker",
      inset = 0, xpd = TRUE, horiz = FALSE)
dev.off()
```

```{r, echo=FALSE, warning=FALSE, message=FALSE}
plate.cols <- c("#386cb0", "#fdb462","#7fc97f")
names(plate.cols) <- c(6, 7, 8)
plate.col.mat <- cbind.data.frame("Plate.Colour"=plate.cols,
                                  "Plate"=names(plate.cols))
dc.merge <- merge(dc.merge, plate.col.mat, by='Plate')
dc.merge$Plate.Colour <- as.character(dc.merge$Plate.Colour)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
png("~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/SupplementaryFigure_3D-diffusion-byPlate.png",
     height=7.25, width=7.25, res=300, units="in")
scatterplot3d(x=dc.merge[, 6], y=dc.merge[, 7], z=dc.merge[, 5], bg=dc.merge$Plate.Colour, 
              pch=21, xlab="Diffusion Component 2", ylab="Diffusion Component 3",
              zlab="Diffusion Component 1")
legend("right", legend = names(plate.cols),
      pt.bg =  plate.cols, 
      pch = 21, ncol=1, title="Plate",
      inset = 0, xpd = TRUE, horiz = FALSE)
dev.off()
```



```{r, echo=FALSE, fig.height=7.5, fig.width=7.5}
dc.p <- ggplot(dc.merge, aes(x=DC1, y=DC2, colour=Cluster)) +
  geom_point() + theme_mike() +
  scale_colour_manual(values=col.map)
  #scale_colour_Publication()
  #scale_colour_distiller(palette="RdYlBu")

dc1.p.dens <- ggplot(dc.merge, aes(DC1, colour=Cluster)) +
  geom_density(alpha=0.3) + theme_mike() +
  scale_colour_manual(values=col.map)
  #scale_colour_Publication()

dc2.p.dens <- ggplot(dc.merge, aes(DC2, colour=Cluster)) +
  geom_density(alpha=0.3) + theme_mike() +
  scale_colour_manual(values=col.map)
  #scale_colour_Publication()

dc.cc <-  ggplot(dc.merge, aes(x=DC1, colour=PhaseLabel)) +
  geom_density(alpha=0.3) + theme_mike() +
  scale_colour_Publication()
  
pseudo.plots <- plot_grid(dc.p, dc1.p.dens, dc2.p.dens, dc.cc,
                          ncol=2)

ggsave(pseudo.plots,
       filename="~/Dropbox/Adipose/plots.dir/Pseudotime-AllCells.png",
       height=7.5, width=9.5)

pseudo.plots
```

```{r}
dc.p <- ggplot(dc.merge, aes(x=DC1, y=DC2, fill=Cluster)) +
  geom_point(size=3, shape=21) + theme_minimal() +
  scale_fill_manual(values=col.map) +
  labs(x="Diffusion Component 1", y="Diffusion Component 2") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=17, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14)) +
  guides(fill=guide_legend(override.aes=aes(size=8)))

ggsave(dc.p,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Pseudotime_Clusters-scatter.png",
       width=7.75, height=5.5, dpi=300)

dc.p
```


```{r}
dc.gfp <- ggplot(dc.merge, aes(x=DC1, y=DC2, fill=Wt1)) +
  geom_point(size=3, shape=21) + theme_minimal() +
  scale_fill_gradient(low="#F4F6AC", high="#A40000") +
  labs(x="Diffusion Component 1", y="Diffusion Component 2") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=17, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14)) +
  guides(fill=guide_colorbar(override.aes=aes(size=4)))

dc.gfp
```

This looks a bit like the separation between the two lineages is primarily driven by Wt1.  Can I get a partitioning of this space based on the two diffusion 
components and Wt1 expression?

```{r, echo=FALSE, warning=FALSE, message=FALSE}
set.seed(6514354)
partition <- kmeans(dc.merge[, c("DC2")], centers=2)
# make sure the partition names are consistent
dc.merge$Kmeans <- partition$cluster
# dc.merge$Kmeans[dc.merge$Kmeans == 1] <- 5
# dc.merge$Kmeans[dc.merge$Kmeans == 2] <- 1
# dc.merge$Kmeans[dc.merge$Kmeans == 5] <- 2
dc.merge$Kmeans <- as.factor(dc.merge$Kmeans)

dc.kmeans <- ggplot(dc.merge, aes(x=DC1, y=DC2, fill=Cluster, shape=Kmeans, group=Kmeans)) +
  geom_point(size=3.5) + theme_minimal() +
  scale_fill_manual(values=col.map, labels=names(col.map)) +
  scale_shape_manual(values=c(21, 24), labels=c(1, 2)) +
  labs(x="Diffusion Component 1", y="Diffusion Component 2") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=17, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14))  +
  guides(fill=guide_legend(title="Cluster", override.aes=list(shape=c(21), size=5)),
         shape=guide_legend(title="Partition", override.aes=list(size=5)))

ggsave(dc.kmeans,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Pseudotime_Kmeans-Scatter.png",
       width=6.75, height=5.75, dpi=300)

dc.kmeans
```

Add Wt1 makes the partitioning worse.  Maybe construct different trajectories for these two partitions.  Let's compare this to the PCA for these cells using these genes.

```{r}
gfp.cols <- c("#13CF00", "#CC8EF4")
names(gfp.cols) <- c("GFP+", "GFP-")

dc.gfp <- ggplot(dc.merge, aes(x=DC1, y=DC2, fill=Marker, shape=Kmeans, group=Kmeans)) +
  geom_point(size=3) + theme_minimal() +
  scale_fill_manual(values=gfp.cols) +
  scale_shape_manual(values=c(21, 24), labels=c(1, 2)) +
  labs(x="Diffusion Component 1", y="Diffusion Component 2") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=17, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14)) +
    guides(fill=guide_legend(title="Marker", override.aes=list(shape=c(21), size=8)),
         shape=guide_legend(title="Partition", override.aes=list(size=5)))

ggsave(dc.gfp,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Pseudotime-GFP_marker-Scatter.png",
       height=5.75, width=6.75)

dc.gfp
```



```{r, echo=FALSE, warning=FALSE, message=FALSE}
ptime.pca <- as.data.frame(t(adipose.pcs[1:10, !colnames(adipose.hvg) %in% multi.cell]))
ptime.pca$Sample <- rownames(ptime.pca)

pca.merge <- merge(dc.merge, ptime.pca, by='Sample')

ggplot(pca.merge, aes(x=PC1, y=PC2, fill=Cluster, shape=Kmeans)) +
  geom_point(size=3) + theme_minimal() +
  scale_fill_manual(values=col.map) +
  scale_shape_manual(values=c(21, 24)) +
  #scale_fill_gradient(low="#F4F6AC", high="#A40000") +
  labs(x="Principal Component 1", y="Principal Component 2") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=17, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14)) +
  guides(fill=guide_colorbar(override.aes=aes(size=4)))
```

The diffusion map is required, as the first 2 PCs do not capture the separation between the two apparent trajectories to the same extent.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
# add cluster interpretation labels
# 1 & 3 = most mature, 2 = most naive, 4 = intermediate
meta.merge$ClusterLabel <- ""
meta.merge$ClusterLabel[meta.merge$Cluster %in% c(1, 3)] <- "Mature"
meta.merge$ClusterLabel[meta.merge$Cluster %in% c(2)] <- "Naive"
meta.merge$ClusterLabel[meta.merge$Cluster %in% c(4)] <- "Intermediate"
meta.merge$ClusterLabel <- factor(meta.merge$ClusterLabel,
                                  levels=c("Naive", "Intermediate", "Mature"),
                                  labels=c("Naive", "Intermediate", "Mature"))

# need different colours to the clustering to prevent confusion
#label.map <- col2hcl(unique(meta.merge$Cluster))
label.map <- c("#FF7800", "#FFB200", "#AB611E")

names(label.map) <- c("Naive", "Intermediate", "Mature")
```


### Partition 1 (yellow, red & purple) - diffusion pseudotime

```{r, echo=FALSE, message=FALSE, warning=FALSE}
# remove the mult-cell controls
multi.cell <- meta.merge$Sample[meta.merge$SampleType == "Multi-cell"]

part2.cells <- dc.merge$Sample[dc.merge$Kmeans == 2 & !dc.merge$Sample %in% multi.cell]
part1.cells <- dc.merge$Sample[dc.merge$Kmeans == 1 & !dc.merge$Sample %in% multi.cell]

part2.adipose <- adipose.nz[, colnames(adipose.nz) %in% part2.cells]
part1.adipose <- adipose.nz[, colnames(adipose.nz) %in% part1.cells]

# just use the intersection of gfp+ and gfp- HVGs to be able to compare them
# there are many more highly variable genes in the GFP+ cells, despite the same
# number of cells, why is this?

part2.genes <- rownames(part2.adipose)[find_hvg(part2.adipose, p.threshold=5e-3)]
part1.genes <- rownames(part1.adipose)[find_hvg(part1.adipose, p.threshold=5e-3)]

#part_both.genes <- union(part2.genes, part1.genes)
part_both.genes <- intersect(part2.genes, part1.genes)

part2.hvg <- part2.adipose[part2.genes, ]
#part2.hvg <- part2.hvg[, !duplicated(part2.hvg), ]

part1.hvg <- part1.adipose[part1.genes, ]
#part1.hvg <- part1.hvg[, !duplicated(part1.hvg)]
# remove duplicated cells
```


```{r, echo=FALSE, fig.width=9.5, fig.height=4.5}
set.seed(24021982)

part1.dm <- DiffusionMap(t(part1.hvg), n_eigs=5, k=20, distance='cosine', sigma='local')
# how many diffusion components do we need?

part1.dpt <- DPT(part1.dm)
part1.dc <- do.call(cbind.data.frame,
                      list("DC1"=part1.dpt$DC1,
                           "DC2"=part1.dpt$DC2,
                           "DC3"=part1.dpt$DPT3,
                           "DPT1"=part1.dpt$DPT1,
                           "DPT2"=part1.dpt$DPT2,
                           "DPT3"=part1.dpt$DPT3))

rownames(part1.dc) <- colnames(part1.hvg)
part1.dc$Sample <- rownames(part1.dc)
part1.dc.merge <- merge(part1.dc, marker.trans, by='Sample')
part1.dc.merge <- merge(part1.dc.merge, meta.merge, by='Sample')

write.table(part1.dc.merge,
            "~/Dropbox/Adipose/Adipose-Kmeans1-DiffusionPseudotime.tsv",
            sep="\t", quote=F, row.names=F)

pos.dc.p <- ggplot(part1.dc.merge, aes(x=DC1, y=DC2, colour=ClusterLabel)) +
  geom_point() + theme_mike() +
  scale_colour_manual(values=label.map)


pos.cc <- ggplot(part1.dc.merge, aes(x=DC1, y=DC2, colour=PhaseLabel)) +
  geom_point() + theme_mike() +
  scale_colour_Publication()

plot_grid(pos.dc.p, pos.cc, ncol=2)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
part1.ptime.p <- ggplot(part1.dc.merge, aes(x=DC1, y=DC2, fill=ClusterLabel)) +
  geom_point(size=3, shape=21) + theme_minimal() +
  scale_fill_manual(values=label.map) +
  labs(x="Diffusion Component 1", y="Diffusion Component 2") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=17, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14)) +
  guides(fill=guide_legend(override.aes=aes(size=8)))

ggsave(part1.ptime.p,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Pseudotime-Kmeans1-Scatter.png",
       height=5.75, width=6.75, dpi=300)

part1.ptime.p
```


```{r}
plot(part1.dpt, pch=20)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
ggplot(part1.dc.merge, aes(x=DPT1, fill=ClusterLabel)) +
  geom_density(alpha=0.5) + theme_mike() +
  scale_fill_manual(values=label.map)
```

It does look like there is a fate choice between 2 branches.  What then do these represent?  What about expression of Wt and the GFP sorting, is this 
different for these cells?

```{r}
gfp.cols <- c("#13CF00", "#CC8EF4")
names(gfp.cols) <- c("GFP+", "GFP-")

part1.dc.gfp <- ggplot(part1.dc.merge, aes(x=DC1, y=DC2, fill=Marker)) +
  geom_point(size=3, shape=21) + theme_minimal() +
  scale_fill_manual(values=gfp.cols) +
  labs(x="Diffusion Component 1", y="Diffusion Component 2") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=17, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14)) +
  guides(fill=guide_legend(override.aes=aes(size=8)))

part1.dc.gfp
```

It looks lke generally speaking there are more GFP+ cells towards the naive end of the pseudotime.  Let's double check that in a density plot.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
ggplot(part1.dc.merge, aes(x=DC1, fill=Marker)) +
  geom_density(alpha=0.5) + theme_mike() +
  scale_fill_manual(values=gfp.cols)
```

There are marginally more GFP+ cells towards the right hand end, or rather a depletion of the GFP- cells at least.


```{r}
part1.dc.gfp <- ggplot(part1.dc.merge, aes(x=DC1, y=DC2, fill=Wt1)) +
  geom_point(size=3, shape=21) + theme_minimal() +
  scale_fill_gradient(low="#F4F6AC", high="#A40000") +
  labs(x="Diffusion Component 1", y="Diffusion Component 2") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=17, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14)) +
  guides(fill=guide_colorbar(override.aes=aes(size=4)))

part1.dc.gfp
```

Wt1 expression seems quite dispersed across cells.


```{r}
part1.dc.gfp <- ggplot(part1.dc.merge, aes(x=DC1, y=DC2, fill=DPT1)) +
  geom_point(size=3, shape=21) + theme_minimal() +
  #scale_fill_manual(values=gfp.cols) +
  labs(x="Diffusion Component 1", y="Diffusion Component 2") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=17, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14)) +
  guides(fill=guide_legend(override.aes=aes(size=8)))

part1.dc.gfp
```

### Partition 2 (yellow & turquoise) - diffusion pseudotime

```{r, echo=FALSE, fig.width=9.5, fig.height=4.5}
set.seed(24021982)
part2.dm <- DiffusionMap(t(part2.hvg), n_eigs=2, k=20, distance='cosine')
# how many diffusion components do we need?

part2.dpt <- DPT(part2.dm)
part2.dc <- do.call(cbind.data.frame,
                      list("DC1"=part2.dpt$DC1,
                           "DC2"=part2.dpt$DC2,
                           "DPT1"=part2.dpt$DPT1,
                           "DPT2"=part2.dpt$DPT2))

rownames(part2.dc) <- colnames(part2.hvg)
part2.dc$Sample <- rownames(part2.dc)
part2.dc.merge <- merge(part2.dc, marker.trans, by='Sample')
part2.dc.merge <- merge(part2.dc.merge, meta.merge, by='Sample')

write.table(part2.dc.merge,
            "~/Dropbox/Adipose/Adipose-Kmeans2-DiffusionPseudotime.tsv",
            sep="\t", quote=F, row.names=F)

pos.dc.p <- ggplot(part2.dc.merge, aes(x=DC1, y=DC2, colour=ClusterLabel)) +
  geom_point() + theme_mike() +
  scale_colour_manual(values=label.map)
  #scale_colour_Publication()
  #scale_colour_distiller(palette="RdYlBu")

pos.cc <- ggplot(part2.dc.merge, aes(x=DC1, y=DC2, colour=PhaseLabel)) +
  geom_point() + theme_mike() +
  scale_colour_Publication()

plot_grid(pos.dc.p, pos.cc, ncol=2)
```

This looks like a single trajectory where the cells become really dispersed before collapsing back down to a particular fate.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
part2.ptime.p <- ggplot(part2.dc.merge, aes(x=DC1, y=DC2, fill=ClusterLabel)) +
  geom_point(size=3, shape=24) + theme_minimal() +
  scale_fill_manual(values=label.map) +
  labs(x="Diffusion Component 1", y="Diffusion Component 2") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=17, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14)) +
  guides(fill=guide_legend(override.aes=aes(size=8)))

ggsave(part2.ptime.p,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Pseudotime-Kmeans2-Scatter.png",
       height=5.75, width=6.75, dpi=300)

part2.ptime.p
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
plot(part2.dpt)
```

Although destiny calls 3 branches here, I'm not 100% convinced that is the case.  There is definitely a change in the cell-to-cell heterogeneity though.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
ggplot(part2.dc.merge, aes(x=DPT1, fill=ClusterLabel)) +
  geom_density(alpha=0.5) + theme_mike() +
  scale_fill_manual(values=label.map)
```

The cellular density is concentrated at the two ends of the first diffusion pseudotime ordering (yellow are naive so need to read from right to left). 
There are only 2 cells from cluster 4 in this trajectory, hence the strange red block shape in the middle of the density plot.

```{r}
part2.dc.gfp <- ggplot(part2.dc.merge, aes(x=DC1, y=DC2, fill=Marker)) +
  geom_point(size=3, shape=21) + theme_minimal() +
  scale_fill_manual(values=gfp.cols) +
  labs(x="Diffusion Component 1", y="Diffusion Component 2") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=17, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14)) +
  guides(fill=guide_legend(override.aes=aes(size=8)))

part2.dc.gfp
```

It doesn't really look like there is much distinction between the GFP+ and GFP- cells

```{r, echo=FALSE, warning=FALSE, message=FALSE}
ggplot(part2.dc.merge, aes(x=DC1, fill=Marker)) +
  geom_density(alpha=0.5) + theme_mike() +
  scale_fill_manual(values=gfp.cols)
```

Again, there are marginally more GFP+ cells towards the right hand end, or rather a depletion of the GFP- cells at least.


```{r, echo=FALSE, warning=FALSE, message=FALSE}
part2.dc.gfp <- ggplot(part2.dc.merge, aes(x=DC1, y=DC2, fill=Wt1)) +
  geom_point(size=3, shape=21) + theme_minimal() +
  scale_fill_gradient(low="#F4F6AC", high="#A40000") +
  labs(x="Diffusion Component 1", y="Diffusion Component 2") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=17, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14)) +
  guides(fill=guide_colorbar(override.aes=aes(size=4)))

part2.dc.gfp
```

Wt1 expression seems quite dispersed across cells, but also fairly highly expressed.

I think given these results, and the apparent emergence of two separate trajectories I will treat them as two completely separate data sets from here on.

## Partition 1 analysis

```{r, echo=FALSE}
part1.ptime.clust <- ggplot(part1.dc.merge, aes(x=Cluster, y=DPT1, fill=ClusterLabel)) +
  geom_boxplot() +
  geom_jitter(shape=21) +
  theme_bw() +
  scale_fill_manual(values=label.map) +
  labs(x="Cluster", y="Diffusion Pseudotime") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=17, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14), 
        panel.grid.major=element_blank(),
        panel.border=element_blank(),
        axis.line=element_line(colour="black")) +
  guides(fill=FALSE)

ggsave(part1.ptime.clust,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Pseudotime-Kmeans1_boxplot-clusters.png",
       height=6.5, width=3.25, dpi=300)

part1.ptime.clust
```

Show expression for a couple of interesting markers, such as Cd24 and Sca1.

```{r, echo=FALSE}
part1.cd34.clust <- ggplot(part1.dc.merge, aes(y=Cd34, x=Cluster, fill=ClusterLabel)) +
  geom_boxplot() +
  geom_jitter(shape=21) +
  theme_bw() +
  scale_fill_manual(values=label.map) +
  labs(y=expression(bold(paste("Cd34 log"[2], " Expression"))), x="Cluster") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=17, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14), 
        panel.grid.major=element_blank(),
        panel.border=element_blank(),
        axis.line=element_line(colour="black")) +
  guides(fill=FALSE) +
  lims(y=c(-1, 16))

ggsave(part1.cd34.clust,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Cd34_cluster-Kmeans1-boxplot.png",
       height=3.25, width=3.25, dpi=300)

part1.cd34.clust
```

```{r, echo=FALSE}
part1.cd29.clust <- ggplot(part1.dc.merge, aes(y=Itgb1, x=Cluster, fill=ClusterLabel)) +
  geom_boxplot() +
  geom_jitter(shape=21) +
  theme_bw() +
  scale_fill_manual(values=label.map) +
  labs(y=expression(bold(paste("Cd29 log"[2], " Expression"))), x="Cluster") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=17, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14), 
        panel.grid.major=element_blank(),
        panel.border=element_blank(),
        axis.line=element_line(colour="black")) +
  guides(fill=FALSE) +
  lims(y=c(-1, 16))

ggsave(part1.cd29.clust,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Cd29_cluster-Kmeans1-boxplot.png",
       height=3.25, width=3.25, dpi=300)

part1.cd29.clust
```

```{r, echo=FALSE}
part1.sca1.clust <- ggplot(part1.dc.merge, aes(y=Ly6a, x=Cluster, fill=ClusterLabel)) +
  geom_boxplot() +
  geom_jitter(shape=21) +
  theme_bw() +
  scale_fill_manual(values=label.map) +
  labs(y=expression(bold(paste("Sca1 log"[2], " Expression"))), x="Cluster") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=17, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14), 
        panel.grid.major=element_blank(),
        panel.border=element_blank(),
        axis.line=element_line(colour="black")) +
  guides(fill=FALSE) +
  lims(y=c(-1, 16))

ggsave(part1.sca1.clust,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Sca1_cluster-Kmeans1-boxplot.png",
       height=3.25, width=3.25, dpi=300)

part1.sca1.clust
```



```{r, echo=FALSE}
part1.cd34.clust <- ggplot(part1.dc.merge, aes(y=Cd34, x=Cluster, fill=ClusterLabel)) +
  geom_boxplot() +
  geom_jitter(shape=21) +
  theme_bw() +
  scale_fill_manual(values=label.map) +
  labs(y=expression(bold(paste("Cd34 log"[2], " Expression"))), x="Cluster") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=17, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14), 
        panel.grid.major=element_blank(),
        panel.border=element_blank(),
        axis.line=element_line(colour="black")) +
  guides(fill=FALSE) +
  lims(y=c(-1, 16))

ggsave(part1.cd34.clust,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Cd34_cluster-Kmeans1-boxplot.png",
       height=3.25, width=3.25, dpi=300)

part1.cd34.clust
```




```{r, echo=FALSE}
part1.wt1.clust <- ggplot(part1.dc.merge, aes(y=Wt1, x=Cluster, fill=ClusterLabel)) +
  geom_boxplot() +
  geom_jitter(shape=21) +
  theme_bw() +
  scale_fill_manual(values=label.map) +
  labs(y=expression(bold(paste("Wt1 log"[2], " Expression"))), x="Cluster") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=17, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14), 
        panel.grid.major=element_blank(),
        panel.border=element_blank(),
        axis.line=element_line(colour="black")) +
  guides(fill=FALSE) +
  lims(y=c(-1, 16))

ggsave(part1.wt1.clust,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Wt1_cluster-Kmeans1-boxplot.png",
       height=3.25, width=3.25, dpi=300)

part1.wt1.clust
```

```{r, echo=FALSE}
part1.pparg.clust <- ggplot(part1.dc.merge, aes(y=Pparg, x=Cluster, fill=ClusterLabel)) +
  geom_boxplot() +
  geom_jitter(shape=21) +
  theme_bw() +
  scale_fill_manual(values=label.map) +
  labs(y=expression(bold(paste("Pparg log"[2], " Expression"))), x="Cluster") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=17, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14), 
        panel.grid.major=element_blank(),
        panel.border=element_blank(),
        axis.line=element_line(colour="black")) +
  guides(fill=FALSE) +
  lims(y=c(-1, 16))

ggsave(part1.pparg.clust,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Pparg_cluster-Kmeans1-boxplot.png",
       height=3.25, width=3.25, dpi=300)

part1.pparg.clust
```


The diffusion map creation is quite sensitive to the magnitude of the number of nearest neighbours, `k`.  Using a small value for `k`, we can see a branch-like trajectory of cells moving from right to left (or _vice versa_).  Let's take a look to see what genes are expressed along this pseudotime trajectory, starting with _Wt1_ and _Tcf21_.  Perhaps we can also split the data into GFP+ and GFP- cells and see whether these explain the branches?

It doesn't look like cells in particular phases of the cell cycle cluster on particular branches, perhaps there is a slight skew towards the left hand end, but it is not clear cut.  I suspect this indicates that the cell cycle phase classification doesn't work so well on these cells.  It would be better to get that information from specific proliferation assays on adipose precursors.

### Is the adipose precursor pseudotime indicative of a hierarchy of cells?

These adipose precursors (AP) cells were extracted from the stromal vascular fraction (SVF) of white adipose tissue (WAT) epididymal depots as Lin-CD31-IgM-CD34+GFP+/-.  Matthew S Roderheffer's lab have published on APs, showing that CD34+CD29+Sca1+CD24- are derived from a CD24+ population of cells.  We do not observe any expression of Cd24 in these cells, indicating they are probably a slightly later populations of APs.  The question is, how are the sanity/marker genes expressed along the pseudotime?

```{r, echo=FALSE, fig.height=7.5, fig.width=9.5}
part1.cd29.dc <- ggplot(part1.dc.merge, aes(x=DPT1, y=Itgb1, colour=ClusterLabel)) +
  geom_point() + theme_mike() +
  scale_colour_manual(values=label.map) +
  geom_smooth(method="loess", span=0.5, colour="black")

part1.sca1.dc <- ggplot(part1.dc.merge, aes(x=DPT1, y=Ly6a, colour=ClusterLabel)) +
  geom_point() + theme_mike() +
  scale_colour_manual(values=label.map) +
  geom_smooth(method="loess", span=0.5, colour="black")

part1.wt1.dc <- ggplot(part1.dc.merge, aes(x=DPT1, y=Wt1, colour=ClusterLabel)) +
  geom_point() + theme_mike() +
  scale_colour_manual(values=label.map) +
  geom_smooth(method="loess", span=0.5, colour="black")

part1.tcf21.dc <- ggplot(part1.dc.merge, aes(x=DPT1, y=Tcf21, colour=ClusterLabel)) +
  geom_point() + theme_mike() +
  scale_colour_manual(values=label.map) +
  geom_smooth(method="loess", span=0.5, colour="black")

part1.pparg.dc <- ggplot(part1.dc.merge, aes(x=DPT1, y=Pparg, colour=ClusterLabel)) +
  geom_point() + theme_mike() +
  scale_colour_manual(values=label.map) +
  geom_smooth(method="loess", span=0.5, colour="black")

part1.cebpa.dc <- ggplot(part1.dc.merge, aes(x=DPT1, y=Cebpa, colour=ClusterLabel)) +
  geom_point() + theme_mike() +
  scale_colour_manual(values=label.map) +
  geom_smooth(method="loess", span=0.5, colour="black")


plot_grid(part1.cd29.dc, part1.sca1.dc, part1.wt1.dc, part1.tcf21.dc, part1.pparg.dc, part1.cebpa.dc,
          labels=c("Cd29", "Sca1", "Wt1", "Tcf21", "Pparg", "Cebpa"),
          ncol=3)
```


It looks like the expression of Cd29 (Itgb1), Sca1 (Ly6a) and Wt1 are across pretty much the entire pseudotime, without any obvious restriction to any aspect of it.  Pparg on the other hand is restricted to the left end of the trajectory, mostly in the brown, pink and turquoise clusters of cells.  It looks like Tcf21 expression is declining gradually along the pseudotime.  Cebpa is likewise restricted to the left end of the pseudotime, but with slightly more dispersed expression that includes some of the green, red and blue cells.  These expression patterns for Pparg and Cebpa suggest the left branch are the more mature adipocyte precursors, as these are markers of mature adipocytes.

### GFP+ve precursors - diffusion pseudotime

```{r, echo=FALSE, message=FALSE, warning=FALSE}
# remove the mult-cell controls
multi.cell <- meta.merge$Sample[meta.merge$SampleType == "Multi-cell"]

part1.gfp_pos.cells <- part1.dc.merge$Sample[part1.dc.merge$Marker == "GFP+" & !part1.dc.merge$Sample %in% multi.cell]
part1.gfp_neg.cells <- part1.dc.merge$Sample[part1.dc.merge$Marker == "GFP-" & !part1.dc.merge$Sample %in% multi.cell]

part1.gfp_pos.adipose <- adipose.nz[, colnames(adipose.nz) %in% part1.gfp_pos.cells]
part1.gfp_neg.adipose <- adipose.nz[, colnames(adipose.nz) %in% part1.gfp_neg.cells]

# just use the intersection of gfp+ and gfp- HVGs to be able to compare them
# there are many more highly variable genes in the GFP+ cells, despite the same
# number of cells, why is this?

part1.gfp_pos.genes <- rownames(part1.gfp_pos.adipose)[find_hvg(part1.gfp_pos.adipose)]
part1.gfp_neg.genes <- rownames(part1.gfp_neg.adipose)[find_hvg(part1.gfp_neg.adipose)]

part1.gfp_both.genes <- union(part1.gfp_pos.genes, part1.gfp_neg.genes)

part1.gfp_pos.hvg <- part1.gfp_pos.adipose[part1.gfp_both.genes, ]
#gfp_pos.hvg <- gfp_pos.hvg[, !duplicated(gfp_pos.hvg), ]

part1.gfp_neg.hvg <- part1.gfp_neg.adipose[part1.gfp_both.genes, ]
#gfp_neg.hvg <- gfp_neg.hvg[, !duplicated(gfp_neg.hvg)]
# remove duplicated cells
```



```{r, echo=FALSE}
set.seed(24021982)
#gfp_pos.dm <- DiffusionMap(t(gfp_pos.pcs[!colnames(gfp_pos.hvg) %in% multi.cell, 1:55]), n_eigs=5, k=10, distance='cosine')
part1.gfp_pos.dm <- DiffusionMap(t(part1.gfp_pos.hvg), n_eigs=3, k=8, distance='euclidean')
# how many diffusion components do we need?
#plot(eigenvalues(gfp_pos.dm), pch=20)

#plot(gfp_pos.dm)
```


```{r, echo=FALSE, fig.width=9.5, fig.height=4.5}
part1.gfp_pos.dpt <- DPT(part1.gfp_pos.dm)
part1.gfp_pos.dc <- do.call(cbind.data.frame,
                      list("DC1"=part1.gfp_pos.dpt$DC1,
                           "DC2"=part1.gfp_pos.dpt$DC2,
                           "DC3"=part1.gfp_pos.dpt$DPT3,
                           "DPT1"=part1.gfp_pos.dpt$DPT1,
                           "DPT2"=part1.gfp_pos.dpt$DPT2,
                           "DPT3"=part1.gfp_pos.dpt$DPT3))

rownames(part1.gfp_pos.dc) <- colnames(part1.gfp_pos.hvg)
part1.gfp_pos.dc$Sample <- rownames(part1.gfp_pos.dc)
gfp_pos.part1.dc.merge <- merge(part1.gfp_pos.dc, marker.trans, by='Sample')
gfp_pos.part1.dc.merge <- merge(gfp_pos.part1.dc.merge, meta.merge, by='Sample')

write.table(gfp_pos.part1.dc.merge,
            "~/Dropbox/Adipose/Adipose-GFPpos_Kmeans1-DiffusionPseudotime.tsv",
            sep="\t", quote=F, row.names=F)

part1.pos.dc.p <- ggplot(gfp_pos.part1.dc.merge, aes(x=DC1, y=DC2, colour=ClusterLabel)) +
  geom_point() + theme_mike() +
  scale_colour_manual(values=label.map)
  #scale_colour_Publication()
  #scale_colour_distiller(palette="RdYlBu")

part1.pos.cc <- ggplot(gfp_pos.part1.dc.merge, aes(x=DC1, y=DC2, colour=PhaseLabel)) +
  geom_point() + theme_mike() +
  scale_colour_Publication()


plot_grid(part1.pos.dc.p, part1.pos.cc, ncol=2)
#plot(gfp_pos.dpt, pch=20)
#plot(gfp_pos.dpt, root=2, paths_to=c(1, 3), col_by='branch', pch=20)
```

The cell cycle classification labels look pretty uniformative, they certainly don't cluster in a specific branch of the pseudotime.  This also doesn't 
really look any different from all cells together regardless of GFP presence/absence.

### Partition 1 GFP-ve precursors - diffusion pseudotime

```{r, echo=FALSE}
set.seed(24021982)
# remove the mult-cell controls
#gfp_pos.dm <- DiffusionMap(t(gfp_pos.pcs[!colnames(gfp_pos.hvg) %in% multi.cell, 1:20]), n_eigs=5, distance='euclidean')
part1.gfp_neg.dm <- DiffusionMap(t(part1.gfp_neg.hvg[, !colnames(part1.gfp_neg.hvg) %in% multi.cell]), n_eigs=3, k=8, distance='euclidean')
# how many diffusion components do we need?
#plot(eigenvalues(gfp_neg.dm), pch=20)

#plot(gfp_neg.dm)
```


```{r, echo=FALSE, fig.width=9.5, fig.height=4.5}
part1.gfp_neg.dpt <- DPT(part1.gfp_neg.dm)
part1.gfp_neg.dc <- do.call(cbind.data.frame,
                      list("DC1"=part1.gfp_neg.dpt$DC1,
                           "DC2"=part1.gfp_neg.dpt$DC2,
                           "DC3"=part1.gfp_neg.dpt$DPT3,
                           "DPT1"=part1.gfp_neg.dpt$DPT1,
                           "DPT2"=part1.gfp_neg.dpt$DPT2,
                           "DPT3"=part1.gfp_neg.dpt$DPT3))

rownames(part1.gfp_neg.dc) <- colnames(part1.gfp_neg.hvg[, !colnames(part1.gfp_neg.hvg) %in% multi.cell])
part1.gfp_neg.dc$Sample <- rownames(part1.gfp_neg.dc)
gfp_neg.part1.dc.merge <- merge(part1.gfp_neg.dc, marker.trans, by='Sample')
gfp_neg.part1.dc.merge <- merge(gfp_neg.part1.dc.merge, meta.merge, by='Sample')

write.table(gfp_neg.part1.dc.merge,
            "~/Dropbox/Adipose/Adipose-GFPneg-Kmeans1-DiffusionPseudotime.tsv",
            sep="\t", quote=F, row.names=F)


part1.neg.dc.p <- ggplot(gfp_neg.part1.dc.merge, aes(x=DC1, y=DC2, colour=ClusterLabel)) +
  geom_point() + theme_mike() +
  scale_colour_manual(values=label.map)
  #scale_colour_Publication()
  #scale_colour_distiller(palette="RdYlBu")

part1.neg.cc <- ggplot(gfp_neg.part1.dc.merge, aes(x=DC1, y=DC2, colour=PhaseLabel)) +
  geom_point() + theme_mike() +
  scale_colour_Publication()


plot_grid(part1.neg.dc.p, part1.neg.cc, ncol=2)
#plot(gfp_neg.dpt, pch=20)
#plot(gfp_neg.dpt, root=2, paths_to=c(1, 3), col_by='branch', pch=20)
```

It looks like the constructed pseudotime trajectories are largely consistent between the GFP+ and GFP- cells,  but rotated 90 degrees.  Maybe there are 
fewer GFP- cells in the most naive groups of cells (yellow).  Either way, splitting my sorting marker doesn't seem to be informative.

So for each of these trajectories, what genes are most highly correlated with the different branches? 
How much of these effects are driven by variation in cell cycle, and how much of this is a confounding effect?
How much does expression variation and cellular heterogeneity change over these different trajectories?  I need to also make sure that these changes are 
not driven by any QC artifacts, e.g. cells with many 0's in the middle or extreme ends of the trajectory.

There are some interesting complexities here.  Perhaps these can be summarised over several dimensions?  i.e. a PCA on either the diffusion components, or the diffusion pseudotime components.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
part1.dpt.pca <- prcomp(part1.dc.merge[, c("DC1", "DC2", "DC3")])
part1.dpt.pcs <- do.call(cbind.data.frame, list("PC.DC1"=part1.dpt.pca$x[, 1],
                                                "PC.DC2"=part1.dpt.pca$x[, 2],
                                                "PC.DC3"=part1.dpt.pca$x[, 3]))
part1.dpt.pcs$Sample <- part1.dc.merge$Sample
part1.dpt.pcs.merge <- merge(part1.dpt.pcs, part1.dc.merge, by='Sample')
ggplot(part1.dpt.pcs.merge,
       aes(x=PC.DC1, y=PC.DC2, fill=ClusterLabel)) +
  geom_point(shape=21, size=3) + theme_minimal() +
  scale_fill_manual(values=label.map)
```

This just reconstructs the observed diffusion map, so no surprises.  Doing an eigendecomposition on a matrix of eigenvectors, I guess will just return 
those eigen vectors that form the columns of the eignematrix.

After that little segue, let's look for genes that are correlated with the different branches of cells along the diffusion pseudotime.  I'll group cells into "meta-stable", states based on areas of high cellular density, a la Haghverdi _et al_.  Conveniently these also correspond roughly to the cellular clusters defined previously on the top 55 principal components (who'd have thought, huh?).  I'll test for differentially expressed genes between the centre group, and the two extreme ends of the diffusion pseudotime trajectories.  Perform a partitioning of cells on the first diffusion component should generate ~4 clusters looking at the density plots.

#### Differential Expression Analysis: 1 vs. 2

```{r, echo=FALSE}
part1.dc.merge <- part1.dc.merge[!part1.dc.merge$Sample %in% multi.cell, ]
part1.design.mat.1v2 <- model.matrix(~ 0 + ClusterLabel + Plate, data=part1.dc.merge)

# fit a linear model
part1.nz <- adipose.nz[, colnames(adipose.nz) %in% part1.dc.merge$Sample]
part1.fit <- lmFit(part1.nz[, !colnames(part1.nz) %in% multi.cell], part1.design.mat.1v2)

part1.contrast.mat <-  makeContrasts(ClusterLabelIntermediate-ClusterLabelNaive, ClusterLabelMature-ClusterLabelNaive,
                                     ClusterLabelMature-ClusterLabelIntermediate,
                                     levels=part1.design.mat.1v2)

part1.fit.1v2 <- contrasts.fit(part1.fit, contrasts=part1.contrast.mat)
part1.fit.1v2 <- eBayes(part1.fit.1v2)
part1.sum.res.1v2 <- summary(decideTests(part1.fit.1v2))

part1.de.res.1v2 <- topTable(part1.fit.1v2, coef=1, n=Inf, sort="p", p=1.0)
part1.de.res.1v2$Sig <- 0
part1.de.res.1v2$Sig[part1.de.res.1v2$adj.P.Val <= 0.01] <- 1
part1.de.res.1v2$Sig <- as.factor(part1.de.res.1v2$Sig)

part1.de.res.1v2$Diff <- 0
part1.de.res.1v2$Diff[part1.de.res.1v2$logFC < 0 & part1.de.res.1v2$Sig == 1] <- -1
part1.de.res.1v2$Diff[part1.de.res.1v2$logFC > 0 & part1.de.res.1v2$Sig == 1] <- 1

part1.de.res.1v2$gene_id <- rownames(part1.de.res.1v2)
part1.de.merge.1v2 <- merge(part1.de.res.1v2, gene_symbol,
                            by.x='gene_id', by.y='ensembl_gene_id', all.x=TRUE)

part1.gene.top.1v2 <- part1.de.merge.1v2$external_gene_name
part1.gene.top.1v2[part1.de.merge.1v2$logFC > -4.75 & part1.de.merge.1v2$logFC < 4] <- ""
part1.gene.top.1v2[is.na(part1.gene.top.1v2)] <- ""

write.table(part1.de.merge.1v2,
            file="~/Dropbox/Adipose/Adipose-DEgenes_1v2-Kmeans1.tsv",
            sep="\t", quote=F)

ggplot(part1.de.merge.1v2, aes(x=AveExpr, y=logFC, colour=Sig)) +
  geom_point(size=1, alpha=0.8) + theme_mike() +
  scale_colour_Publication() +
  labs(x=expression(paste("Mean log"[2], " Expression")),
       y=expression(paste("log"[2], " Fold Change"))) +
  guides('colour'=guide_legend(title="Statistically significant")) +
  ylim(c(-10, 10)) +
  geom_text_repel(aes(label=part1.gene.top.1v2),
                  colour='black',
                  nudge_x=0.15,
                  nudge_y=0.5)
```

The interpretation of the log$_2$ fold changes is that positive values represent higher expression towards the right hand side of the first diffusions component, i.e. up-regulated in k-2 relative to k-1.  Therefore, negative log fold changes represent higher down-regulation of genes moving along to the right of the first diffusion component.  For example, as we move to the right of the pseudotime, we have increased expression of _Dpp4_, _Efhd1_, _Cd55_, etc, and down-regulation of _Inmt_, _Steap4_, _C7_, etc.

#### Differential Expression Analysis: 1 vs. 3

```{r, echo=FALSE}
part1.de.res.3v1 <- topTable(part1.fit.1v2, coef=2, n=Inf, sort="p", p=1.0)
part1.de.res.3v1$Sig <- 0
part1.de.res.3v1$Sig[part1.de.res.3v1$adj.P.Val <= 0.01] <- 1
part1.de.res.3v1$Sig <- as.factor(part1.de.res.3v1$Sig)

part1.de.res.3v1$Diff <- 0
part1.de.res.3v1$Diff[part1.de.res.3v1$logFC < 0 & part1.de.res.3v1$Sig == 1] <- -1
part1.de.res.3v1$Diff[part1.de.res.3v1$logFC > 0 & part1.de.res.3v1$Sig == 1] <- 1

part1.de.res.3v1$gene_id <- rownames(part1.de.res.3v1)
part1.de.merge.3v1 <- merge(part1.de.res.3v1, gene_symbol,
                  by.x='gene_id', by.y='ensembl_gene_id', all.x=TRUE)
part1.gene.top.3v1 <- part1.de.merge.3v1$external_gene_name
part1.gene.top.3v1[part1.de.merge.3v1$logFC > -5 & part1.de.merge.3v1$logFC < 5] <- ""
part1.gene.top.3v1[is.na(part1.gene.top.3v1)] <- ""

write.table(part1.de.merge.3v1,
            file="~/Dropbox/Adipose/Adipose-DEgenes_1v3-Kmeans1.tsv",
            sep="\t", quote=F)

ggplot(part1.de.merge.3v1, aes(x=AveExpr, y=logFC, colour=Sig)) +
  geom_point(size=1, alpha=0.8) + theme_mike() +
  scale_colour_Publication() +
  labs(x=expression(paste("Mean log"[2], " Expression")),
       y=expression(paste("log"[2], " Fold Change"))) +
  guides('colour'=guide_legend(title="Statistically significant")) +
  ylim(c(-10, 10)) +
  geom_text_repel(aes(label=part1.gene.top.3v1),
                  colour='black',
                  nudge_x=0.15,
                  nudge_y=0.5)
```


#### Differential Expression Analysis: 1 vs. 4

```{r, echo=FALSE}
part1.de.res.4v1 <- topTable(part1.fit.1v2, coef=3, n=Inf, sort="p", p=1.0)
part1.de.res.4v1$Sig <- 0
part1.de.res.4v1$Sig[part1.de.res.4v1$adj.P.Val <= 0.01] <- 1
part1.de.res.4v1$Sig <- as.factor(part1.de.res.4v1$Sig)

part1.de.res.4v1$Diff <- 0
part1.de.res.4v1$Diff[part1.de.res.4v1$logFC < 0 & part1.de.res.4v1$Sig == 1] <- -1
part1.de.res.4v1$Diff[part1.de.res.4v1$logFC > 0 & part1.de.res.4v1$Sig == 1] <- 1

part1.de.res.4v1$gene_id <- rownames(part1.de.res.4v1)
part1.de.merge.4v1 <- merge(part1.de.res.4v1, gene_symbol,
                  by.x='gene_id', by.y='ensembl_gene_id', all.x=TRUE)
part1.gene.top.4v1 <- part1.de.merge.4v1$external_gene_name
part1.gene.top.4v1[part1.de.merge.4v1$logFC > -5 & part1.de.merge.4v1$logFC < 5] <- ""
part1.gene.top.4v1[is.na(part1.gene.top.4v1)] <- ""

write.table(part1.de.merge.4v1,
            file="~/Dropbox/Adipose/Adipose-DEgenes_1v4-Kmeans1.tsv",
            sep="\t", quote=F)

ggplot(part1.de.merge.4v1, aes(x=AveExpr, y=logFC, colour=Sig)) +
  geom_point(size=1, alpha=0.8) + theme_mike() +
  scale_colour_Publication() +
  labs(x=expression(paste("Mean log"[2], " Expression")),
       y=expression(paste("log"[2], " Fold Change"))) +
  guides('colour'=guide_legend(title="Statistically significant")) +
  ylim(c(-10, 10)) +
  geom_text_repel(aes(label=part1.gene.top.4v1),
                  colour='black',
                  nudge_x=0.15,
                  nudge_y=0.5)
```

#### Differential Expression Analysis: 2 vs. 3

```{r, echo=FALSE}
# part1.de.res.3v2 <- topTable(part1.fit.1v2, coef=4, n=Inf, sort="p", p=1.0)
# part1.de.res.3v2$Sig <- 0
# part1.de.res.3v2$Sig[part1.de.res.3v2$adj.P.Val <= 0.01] <- 1
# part1.de.res.3v2$Sig <- as.factor(part1.de.res.3v2$Sig)
# 
# part1.de.res.3v2$Diff <- 0
# part1.de.res.3v2$Diff[part1.de.res.3v2$logFC < 0 & part1.de.res.3v2$Sig == 1] <- -1
# part1.de.res.3v2$Diff[part1.de.res.3v2$logFC > 0 & part1.de.res.3v2$Sig == 1] <- 1
# 
# part1.de.res.3v2$gene_id <- rownames(part1.de.res.3v2)
# part1.de.merge.3v2 <- merge(part1.de.res.3v2, gene_symbol,
#                   by.x='gene_id', by.y='ensembl_gene_id', all.x=TRUE)
# part1.gene.top.3v2 <- part1.de.merge.3v2$external_gene_name
# part1.gene.top.3v2[part1.de.merge.3v2$logFC > -5 & part1.de.merge.3v2$logFC < 5] <- ""
# part1.gene.top.3v2[is.na(part1.gene.top.3v2)] <- ""
# 
# write.table(part1.de.merge.3v2,
#             file="~/Dropbox/Adipose/Adipose-DEgenes_2v3-Kmeans1.tsv",
#             sep="\t", quote=F)
# 
# ggplot(part1.de.merge.3v2, aes(x=AveExpr, y=logFC, colour=Sig)) +
#   geom_point(size=1, alpha=0.8) + theme_mike() +
#   scale_colour_Publication() +
#   labs(x=expression(paste("Mean log"[2], " Expression")),
#        y=expression(paste("log"[2], " Fold Change"))) +
#   guides('colour'=guide_legend(title="Statistically significant")) +
#   ylim(c(-10, 10)) +
#   geom_text_repel(aes(label=part1.gene.top.3v2),
#                   colour='black',
#                   nudge_x=0.15,
#                   nudge_y=0.5)
```


#### Differential Expression Analysis: 2 vs. 4

```{r, echo=FALSE}
# part1.de.res.4v2 <- topTable(part1.fit.1v2, coef=5, n=Inf, sort="p", p=1.0)
# part1.de.res.4v2$Sig <- 0
# part1.de.res.4v2$Sig[part1.de.res.4v2$adj.P.Val <= 0.01] <- 1
# part1.de.res.4v2$Sig <- as.factor(part1.de.res.4v2$Sig)
# 
# part1.de.res.4v2$Diff <- 0
# part1.de.res.4v2$Diff[part1.de.res.4v2$logFC < 0 & part1.de.res.4v2$Sig == 1] <- -1
# part1.de.res.4v2$Diff[part1.de.res.4v2$logFC > 0 & part1.de.res.4v2$Sig == 1] <- 1
# 
# part1.de.res.4v2$gene_id <- rownames(part1.de.res.4v2)
# part1.de.merge.4v2 <- merge(part1.de.res.4v2, gene_symbol,
#                   by.x='gene_id', by.y='ensembl_gene_id', all.x=TRUE)
# part1.gene.top.4v2 <- part1.de.merge.4v2$external_gene_name
# part1.gene.top.4v2[part1.de.merge.4v2$logFC > -5 & part1.de.merge.4v2$logFC < 5] <- ""
# part1.gene.top.4v2[is.na(part1.gene.top.4v2)] <- ""
# 
# write.table(part1.de.merge.4v2,
#             file="~/Dropbox/Adipose/Adipose-DEgenes_2v4-Kmeans1.tsv",
#             sep="\t", quote=F)
# 
# ggplot(part1.de.merge.4v2, aes(x=AveExpr, y=logFC, colour=Sig)) +
#   geom_point(size=1, alpha=0.8) + theme_mike() +
#   scale_colour_Publication() +
#   labs(x=expression(paste("Mean log"[2], " Expression")),
#        y=expression(paste("log"[2], " Fold Change"))) +
#   guides('colour'=guide_legend(title="Statistically significant")) +
#   ylim(c(-10, 10)) +
#   geom_text_repel(aes(label=part1.gene.top.4v2),
#                   colour='black',
#                   nudge_x=0.15,
#                   nudge_y=0.5)
```

#### Differential Expression Analysis: 3 vs. 4

```{r, echo=FALSE}
# part1.de.res.4v3 <- topTable(part1.fit.1v2, coef=6, n=Inf, sort="p", p=1.0)
# part1.de.res.4v3$Sig <- 0
# part1.de.res.4v3$Sig[part1.de.res.4v3$adj.P.Val <= 0.01] <- 1
# part1.de.res.4v3$Sig <- as.factor(part1.de.res.4v3$Sig)
# 
# part1.de.res.4v3$Diff <- 0
# part1.de.res.4v3$Diff[part1.de.res.4v3$logFC < 0 & part1.de.res.4v3$Sig == 1] <- -1
# part1.de.res.4v3$Diff[part1.de.res.4v3$logFC > 0 & part1.de.res.4v3$Sig == 1] <- 1
# 
# part1.de.res.4v3$gene_id <- rownames(part1.de.res.4v3)
# part1.de.merge.4v3 <- merge(part1.de.res.4v3, gene_symbol,
#                   by.x='gene_id', by.y='ensembl_gene_id', all.x=TRUE)
# part1.gene.top.4v3 <- part1.de.merge.4v3$external_gene_name
# part1.gene.top.4v3[part1.de.merge.4v3$logFC > -5 & part1.de.merge.4v3$logFC < 5] <- ""
# part1.gene.top.4v3[is.na(part1.gene.top.4v3)] <- ""
# 
# write.table(part1.de.merge.4v3,
#             file="~/Dropbox/Adipose/Adipose-DEgenes_3v4-Kmeans1.tsv",
#             sep="\t", quote=F)
# 
# ggplot(part1.de.merge.4v3, aes(x=AveExpr, y=logFC, colour=Sig)) +
#   geom_point(size=1, alpha=0.8) + theme_mike() +
#   scale_colour_Publication() +
#   labs(x=expression(paste("Mean log"[2], " Expression")),
#        y=expression(paste("log"[2], " Fold Change"))) +
#   guides('colour'=guide_legend(title="Statistically significant")) +
#   ylim(c(-10, 10)) +
#   geom_text_repel(aes(label=part1.gene.top.4v3),
#                   colour='black',
#                   nudge_x=0.15,
#                   nudge_y=0.5)
```


This looks very curious.  The interpretation is that expression is increased moving along the left of the pseudotime from the center.  This suggests that gene expression changes involve much greater magnitudes of up-regulation, than down-regulation, as we move in this direction along the pseudotime trajectory.  Either this is a separate branch entirely, or the middle of the diffusion component represents a "meta-stable" transition state characterised by co-ordinated down-regulation of many genes.  Unfortunately it is not abundantly clear whether either of these two explanations is correct.  Analagously we are looking at an unrooted tree, so we don't know what the ancestor/most immature cells are, or where they lie on the pseudotime.

```{r}
knitr::kable(part1.sum.res.1v2)
```

There are a large number of differentially expressed genes between the 3 groups, specifically between the middle portion of the first diffusion component, and the either of the two extreme ends.  In the comparison of groups 1 vs. 3, the strength of down regulation of genes is much smaller than for the same comparison with group 2.  This may not have any specific biological meaning, other than that they are more similar along diffusion component 1, than to group 2.

Let's take a look at how these genes are expressed across cells, ordered along the first diffusion component.

```{r, echo=FALSE, fig.height=6.5, fig.width=6.5}
part1.diff.genes <- unique(c(part1.de.merge.1v2$gene_id[part1.de.merge.1v2$Sig == 1 & (abs(part1.de.merge.1v2$logFC) > 3)],
                       part1.de.merge.3v1$gene_id[part1.de.merge.3v1$Sig == 1 & (abs(part1.de.merge.3v1$logFC) > 3)],
                       part1.de.merge.4v1$gene_id[part1.de.merge.4v1$Sig == 1 & (abs(part1.de.merge.4v1$logFC) > 3)]))

part1.clust1.up <- unique(c(part1.de.merge.1v2$gene_id[part1.de.merge.1v2$Sig == 1 & (part1.de.merge.1v2$logFC < -3)],
                      part1.de.merge.3v1$gene_id[part1.de.merge.3v1$Sig == 1 & (part1.de.merge.3v1$logFC < -3)]))

part1.clust2.up <- unique(c(part1.de.merge.1v2$gene_id[part1.de.merge.1v2$Sig == 1 & (part1.de.merge.1v2$logFC > 3)],
                      part1.de.merge.4v1$gene_id[part1.de.merge.4v1$Sig == 1 & (part1.de.merge.4v1$logFC < -3)]))

part1.clust3.up <- unique(c(part1.de.merge.3v1$gene_id[part1.de.merge.3v1$Sig == 1 & (part1.de.merge.3v1$logFC > 3)],
                      part1.de.merge.4v1$gene_id[part1.de.merge.4v1$Sig == 1 & (part1.de.merge.4v1$logFC > 3)]))

# part1.clust4.up <- unique(c(part1.de.merge.4v1$gene_id[part1.de.merge.4v1$Sig == 1 & (part1.de.merge.4v1$logFC > 3)],
#                       part1.de.merge.4v2$gene_id[part1.de.merge.4v2$Sig == 1 & (part1.de.merge.4v2$logFC > 3)],
#                       part1.de.merge.4v3$gene_id[part1.de.merge.4v3$Sig == 1 & (part1.de.merge.4v3$logFC > 3)]))

part1.clust1.genes <- setdiff(part1.clust1.up, unique(c(part1.clust2.up, part1.clust3.up)))
part1.clust2.genes <- setdiff(part1.clust2.up, unique(c(part1.clust1.up, part1.clust3.up)))
part1.clust3.genes <- setdiff(part1.clust3.up, unique(c(part1.clust2.up, part1.clust1.up)))
#part1.clust4.genes <- setdiff(part1.clust4.up, unique(c(part1.clust2.up, part1.clust3.up, part1.clust1.up)))

part1.cluster.genes <- unique(c(part1.clust1.genes, part1.clust2.genes, 
                                part1.clust3.genes))

# only plot the genes unique to each cluster
part1.diff.exprs <- part1.nz[part1.cluster.genes, part1.dc.merge$Sample[order(part1.dc.merge$DPT1, decreasing=TRUE)]]

part1.genes.df <- data.frame(cbind(part1.cluster.genes))
colnames(part1.genes.df) <- c("gene_id")
rownames(part1.genes.df) <- part1.genes.df[, 1]
part1.genes.df$GeneGroup <- "Multiple"
part1.genes.df[part1.genes.df$gene_id %in% part1.clust1.genes, ]$GeneGroup <- "Early"
part1.genes.df[part1.genes.df$gene_id %in% part1.clust2.genes, ]$GeneGroup <- "Intermediate"
part1.genes.df[part1.genes.df$gene_id %in% part1.clust3.genes, ]$GeneGroup <- "Late"
#part1.genes.df[part1.genes.df$gene_id %in% part1.clust4.genes, ]$GeneGroup <- "4"

part1.genes.df$GeneGroup <- as.factor(part1.genes.df$GeneGroup)

#genes.cols <- col.map
#genes.cols <- c("#386cb0", "#fdb462", "#7fc97f","#ef3b2c")
#names(genes.cols) <- c(1:4)
part1.genes.df <- as.data.frame(part1.genes.df[, -1])
rownames(part1.genes.df) <- part1.cluster.genes
colnames(part1.genes.df) <- "GeneGroup"

part1.annot.df <- part1.dc.merge[, c("Marker", "ClusterLabel", "DPT1")]
part1.annot.df$Marker <- as.factor(part1.annot.df$Marker)
part1.annot.df$ClusterLabel <- as.factor(part1.annot.df$ClusterLabel)
part1.annot.df$DPT1 <- as.numeric(part1.annot.df$DPT1)
rownames(part1.annot.df) <- part1.dc.merge$Sample

dc.cols <- colorRampPalette(c("#FAFF60", "#8615B0"))(dim(part1.diff.exprs)[2])
names(dc.cols) <- part1.annot.df$DPT1
  
marker.cols <- gfp.cols
names(marker.cols) <- as.factor(c("GFP+", "GFP-"))

cluster.cols <- label.map

part1.annot.cols <- list(Marker=marker.cols,
                         ClusterLabel=cluster.cols,
                         #GeneGroup=genes.cols,
                         DPT1=dc.cols)

# annot.cols <- list(Marker=marker.cols,
#                    Cluster=cluster.cols,
#                    DC1=dc.cols)
gene.order <- order(part1.genes.df$GeneGroup)

pheatmap(part1.diff.exprs[gene.order, ], show_rownames=FALSE, show_colnames=FALSE,
         cluster_rows=FALSE, annotation_col=part1.annot.df,
         annotation_row=part1.genes.df,
         annotation_colors=part1.annot.cols,
         cluster_cols=FALSE,
         filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Pseudotime_DE_genes-Kmeans1-heatmap.png",
         height=6.75, width=6.75, res=300)
```

There are clearly groups of genes that characterise cluster 2 along the DC quite well, whilst the majority of other genes form a spectrum of expression (which is expected).  We can also more clearly see which end of the pseudotime the different genes are differentially expressed in.  Three of the genes are quite interesting, as they are down in the root/middle of the pseudotime, and are increased in expression in both ends of the trajectory.  These three genes, _Gfpt2_ _Ptx3_ and _Ccl7_, might indicate the presence of shared biological processes in the two branches (Kmeans13).  They might also give a clue to the strange "bridging" population of cells across the two branches.

Let's see what biological processes are enriched in these differentially expressed genes.  I'll combine all differentially expressed genes for now, then look a the separate groups later.

Plot the top DE gene in each cluster on the diffusion map/pseudotime co-ordinate space.

```{r}
part1.c1.max <- unique(c(part1.de.merge.1v2$gene_id[part1.de.merge.1v2$Sig == 1 & (part1.de.merge.1v2$logFC == min(part1.de.merge.1v2$logFC))],
                   part1.de.merge.3v1$gene_id[part1.de.merge.3v1$Sig == 1 & (part1.de.merge.3v1$logFC == min(part1.de.merge.3v1$logFC))],
                   part1.de.merge.4v1$gene_id[part1.de.merge.4v1$Sig == 1 & (part1.de.merge.4v1$logFC == min(part1.de.merge.4v1$logFC))]))

part1.c2.max <- unique(c(part1.de.merge.1v2$gene_id[part1.de.merge.1v2$Sig == 1 & (part1.de.merge.1v2$logFC == max(part1.de.merge.1v2$logFC))],
                   part1.de.merge.3v2$gene_id[part1.de.merge.3v2$Sig == 1 & (part1.de.merge.3v2$logFC == min(part1.de.merge.3v1$logFC))],
                   part1.de.merge.4v1$gene_id[part1.de.merge.4v1$Sig == 1 & (part1.de.merge.4v1$logFC == min(part1.de.merge.4v1$logFC))]))

part1.c3.max <- unique(c(part1.de.merge.3v1$gene_id[part1.de.merge.3v1$Sig == 1 & (part1.de.merge.3v1$logFC == max(part1.de.merge.3v1$logFC))],
                   part1.de.merge.3v2$gene_id[part1.de.merge.3v2$Sig == 1 & (part1.de.merge.3v2$logFC == max(part1.de.merge.3v2$logFC))],
                   part1.de.merge.4v3$gene_id[part1.de.merge.4v3$Sig == 1 & (part1.de.merge.4v3$logFC == min(part1.de.merge.4v3$logFC))]))

part1.c4.max <- unique(c(part1.de.merge.4v1$gene_id[part1.de.merge.4v1$Sig == 1 & (part1.de.merge.4v1$logFC == max(part1.de.merge.4v1$logFC))],
                   part1.de.merge.4v2$gene_id[part1.de.merge.4v2$Sig == 1 & (part1.de.merge.4v2$logFC == max(part1.de.merge.4v2$logFC))],
                   part1.de.merge.4v3$gene_id[part1.de.merge.4v3$Sig == 1 & (part1.de.merge.4v3$logFC == max(part1.de.merge.4v3$logFC))]))

part1.max.exprs <-as.data.frame(t(part1.nz[unique(c(part1.c1.max, part1.c2.max, part1.c3.max, part1.c4.max, "ENSMUSG00000079105")),
                                           part1.dc.merge$Sample[order(part1.dc.merge$DPT1, decreasing=TRUE)]]))
colnames(part1.max.exprs) <- gene_symbol[unique(c(part1.c1.max, part1.c2.max, part1.c3.max, part1.c4.max, "ENSMUSG00000079105")), ]$external_gene_name
part1.max.exprs$Sample <- rownames(part1.max.exprs)

part1.max.merge <- merge(part1.dc.merge, part1.max.exprs, by='Sample')
```


```{r}
# part1.steap4.p <- ggplot(part1.max.merge, aes(x=DC1, y=DC2, fill=Steap4)) +
#   geom_point(size=2, shape=21) + theme_minimal() +
#   scale_fill_gradient(low="#F4F6AC", high="#FF0000", breaks=c(0, 4, 8, 12, 16)) +
#   labs(x="Diffusion Component 1", y="Diffusion Component 2") +
#   theme(panel.grid=element_blank(),
#         axis.text=element_text(size=14, colour="black"),
#         axis.title=element_text(size=17, colour="black", face="bold"),
#         legend.title=element_text(size=14, colour="black", face="bold"),
#         legend.text=element_text(size=14),
#         legend.key=element_rect(size=0.5),
#         legend.key.size=unit(0.5, "cm"),
#         legend.position="bottom")
# 
# ggsave(part1.steap4.p,
#        filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Steap4_pseudotime-Kmeans1.png",
#        height=3.95, width=4.75)
# 
# part1.myc.p <- ggplot(part1.max.merge, aes(x=DC1, y=DC2, fill=Myc)) +
#   geom_point(size=2, shape=21) + theme_minimal() +
#   scale_fill_gradient(low="#F4F6AC", high="#FF0000", breaks=c(0, 4, 8, 12, 16)) +
#   labs(x="Diffusion Component 1", y="Diffusion Component 2") +
#   theme(panel.grid=element_blank(),
#         axis.text=element_text(size=14, colour="black"),
#         axis.title=element_text(size=17, colour="black", face="bold"),
#         legend.title=element_text(size=14, colour="black", face="bold"),
#         legend.text=element_text(size=14),
#         legend.key=element_rect(size=0.5),
#         legend.key.size=unit(0.5, "cm"),
#         legend.position="bottom")
# 
# ggsave(part1.myc.p,
#        filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Myc_pseudotime.png",
#        height=3.95, width=4.75)
# 
# 
# part1.Efhd1.p <- ggplot(part1.max.merge, aes(x=DC1, y=DC2, fill=Efhd1)) +
#   geom_point(size=2, shape=21) + theme_minimal() +
#   scale_fill_gradient(low="#F4F6AC", high="#FF0000", breaks=c(0, 4, 8, 12, 16)) +
#   labs(x="Diffusion Component 1", y="Diffusion Component 2") +
#   theme(panel.grid=element_blank(),
#         axis.text=element_text(size=14, colour="black"),
#         axis.title=element_text(size=17, colour="black", face="bold"),
#         legend.title=element_text(size=14, colour="black", face="bold"),
#         legend.text=element_text(size=14),
#         legend.key=element_rect(size=0.5),
#         legend.key.size=unit(0.5, "cm"),
#         legend.position="bottom")
# 
# ggsave(part1.Efhd1.p,
#        filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Efhd1_pseudotime.png",
#        height=3.95, width=4.75)
# 
# 
# part1.Cilp.p <- ggplot(part1.max.merge, aes(x=DC1, y=DC2, fill=Cilp)) +
#   geom_point(size=2, shape=21) + theme_minimal() +
#   scale_fill_gradient(low="#F4F6AC", high="#FF0000", breaks=c(0, 4, 8, 12, 16)) +
#   labs(x="Diffusion Component 1", y="Diffusion Component 2") +
#   theme(panel.grid=element_blank(),
#         axis.text=element_text(size=14, colour="black"),
#         axis.title=element_text(size=17, colour="black", face="bold"),
#         legend.title=element_text(size=14, colour="black", face="bold"),
#         legend.text=element_text(size=14),
#         legend.key=element_rect(size=0.5),
#         legend.key.size=unit(0.5, "cm"),
#         legend.position="bottom")
# 
# ggsave(part1.Cilp.p,
#        filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Cilp_pseudotime.png",
#        height=3.95, width=4.75)
# 
# 
# part1.Vtn.p <- ggplot(part1.max.merge, aes(x=DC1, y=DC2, fill=Vtn)) +
#   geom_point(size=2, shape=21) + theme_minimal() +
#   scale_fill_gradient(low="#F4F6AC", high="#FF0000", breaks=c(0, 4, 8, 12, 16)) +
#   labs(x="Diffusion Component 1", y="Diffusion Component 2") +
#   theme(panel.grid=element_blank(),
#         axis.text=element_text(size=14, colour="black"),
#         axis.title=element_text(size=17, colour="black", face="bold"),
#         legend.title=element_text(size=14, colour="black", face="bold"),
#         legend.text=element_text(size=14),
#         legend.key=element_rect(size=0.5),
#         legend.key.size=unit(0.5, "cm"),
#         legend.position="bottom")
# 
# ggsave(part1.Vtn.p,
#        filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Vtn_pseudotime.png",
#        height=3.95, width=4.75)
# 
# part1.Mgp.p <- ggplot(part1.max.merge, aes(x=DC1, y=DC2, fill=Mgp)) +
#   geom_point(size=2, shape=21) + theme_minimal() +
#   scale_fill_gradient(low="#F4F6AC", high="#FF0000", breaks=c(0, 4, 8, 12, 16)) +
#   labs(x="Diffusion Component 1", y="Diffusion Component 2") +
#   theme(panel.grid=element_blank(),
#         axis.text=element_text(size=14, colour="black"),
#         axis.title=element_text(size=17, colour="black", face="bold"),
#         legend.title=element_text(size=14, colour="black", face="bold"),
#         legend.text=element_text(size=14),
#         legend.key=element_rect(size=0.5),
#         legend.key.size=unit(0.5, "cm"),
#         legend.position="bottom")
# 
# ggsave(part1.Mgp.p,
#        filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Mgp_pseudotime.png",
#        height=3.95, width=4.75)
# 
# part1.C7.p <- ggplot(part1.max.merge, aes(x=DC1, y=DC2, fill=C7)) +
#   geom_point(size=2, shape=21) + theme_minimal() +
#   scale_fill_gradient(low="#F4F6AC", high="#FF0000", breaks=c(0, 4, 8, 12, 16)) +
#   labs(x="Diffusion Component 1", y="Diffusion Component 2") +
#   theme(panel.grid=element_blank(),
#         axis.text=element_text(size=14, colour="black"),
#         axis.title=element_text(size=17, colour="black", face="bold"),
#         legend.title=element_text(size=14, colour="black", face="bold"),
#         legend.text=element_text(size=14),
#         legend.key=element_rect(size=0.5),
#         legend.key.size=unit(0.5, "cm"),
#         legend.position="bottom")
# 
# ggsave(part1.C7.p,
#        filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/C7_pseudotime.png",
#        height=3.95, width=4.75)
```

We want to test functional enrichments specific to each cluster.

```{r, echo=FALSE}
# use goseq
# use GO enrichment to test for specific pathways and function
# if there are enriched terms, this would at least imply some
# similar functional connection
part1.de.genes.all <- unique(c(part1.de.merge.1v2$gene_id[part1.de.merge.1v2$Sig == 1],
                         part1.de.merge.3v1$gene_id[part1.de.merge.3v1$Sig == 1],
                         part1.de.merge.4v1$gene_id[part1.de.merge.4v1$Sig == 1],
                         part1.de.merge.3v2$gene_id[part1.de.merge.3v2$Sig == 1],
                         part1.de.merge.4v2$gene_id[part1.de.merge.4v2$Sig == 1],
                         part1.de.merge.4v3$gene_id[part1.de.merge.4v3$Sig == 1]))

gene.tpm <- read.table("~/Dropbox/Thymus/mTEC_1_01.quant", sep="\t",
                       h=T, stringsAsFactors=F)
part1.gene.length <- gene.tpm$Length[unique(gene.tpm$Name) %in% unique(union(part1.de.merge.1v2$gene_id, part1.de.merge.3v2$gene_id))]
names(part1.gene.length) <- intersect(unique(gene.tpm$Name) , unique(union(part1.de.merge.1v2$gene_id, part1.de.merge.3v2$gene_id)))

part1.all.gene.vector <- as.integer(unique(union(part1.de.merge.1v2$gene_id, part1.de.merge.3v2$gene_id)) %in% unique(part1.de.genes.all))
names(part1.all.gene.vector) <- unique(union(part1.de.merge.1v2$gene_id, part1.de.merge.3v2$gene_id))

part1.all.gene.vector <- part1.all.gene.vector[names(part1.gene.length)]

# react <- mapGene2Reactome()
pwf.all <- nullp(part1.all.gene.vector, "mm10", "ensGene", bias.data=part1.gene.length, plot.fit=FALSE)
# GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=react,
#                      method="Wallenius", use_genes_without_cat=T)
# part1.GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
#                            method="Wallenius", use_genes_without_cat=T)
part1.GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
                           method="Wallenius", use_genes_without_cat=T)

part1.GO.wall.all$padjust <- p.adjust(part1.GO.wall.all$over_represented_pvalue, method="BH")
part1.GO.wall.all$foldEnrich <- ((part1.GO.wall.all$numDEInCat/length(unique(part1.de.genes.all)))/(part1.GO.wall.all$numInCat/length(part1.gene.length)))

# id2name <- as.list(reactomePATHID2NAME)
# GO.wall.all$description <- do.call(rbind, id2name[GO.wall.all$category])

# select on p-value and foldEnrich
part1.GO.sig.all <- part1.GO.wall.all[(part1.GO.wall.all$padjust <= 0.01), ]
part1.go_cats.all <- data.frame(part1.GO.sig.all$category, part1.GO.sig.all$foldEnrich)

write.table(part1.go_cats.all,
            file="~/Dropbox/Adipose/REVIGO/Adipose-All_4REVIGO-Kmeans1.tsv",
            sep="\t", col.names=F, row.names=F, quote=F)

ggplot(part1.GO.sig.all,
       aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
           size=foldEnrich)) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C")
```

```{r, echo=FALSE, fig.width=7.5, fig.height=9.5}
part1.GO.sig.all <- part1.GO.sig.all[order(part1.GO.sig.all$padjust, decreasing=FALSE), ]

ggplot(part1.GO.sig.all,
       aes(x=reorder(category, 
                     foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")
```

A lot of these gene ontology terms refer to cell migration and motion, which suggests these cells are physically moving during their differentiation.  The enrichment of a wound healing process, coupled with inflammatory response, suggests these are particularly important processes in the development of visceral adipocytes.  BMP stimulus and response to BMP stimulus also appears in the enrichment list.  This is a good validation, given that _Bmp4_ has been shown to stimulate differentiation of MSCs to adipocytes.  It will be interesting to see if and how the BMP genes form a gradient along the pseudotime trajectory.  There is also considerably strong enrichment for gene ontology terms relating to Wnt and $\beta$-catenin signalling, which repress adipocyte development.  Let's start to look at how these genes are distributed over the pseudotime trajectory.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=9.5, fig.width=9.7}
# pull out BMP, Wnt and beta-catenin signalling genes
wnt_table <- read.table("~/Dropbox/Genesets/GO_wnt_catenin-mouse.tsv",
                        sep="\t", h=T, stringsAsFactors=F)
bmp_table <- read.table("~/Dropbox/Genesets/GO_BMP-mouse.tsv",
                        sep="\t", h=T, stringsAsFactors=F)

# genes need to be in gene list and differentially expressed
part1.wnt.exprs <- part1.nz[(rownames(part1.nz) %in% wnt_table$ensembl_gene_id) & (rownames(part1.nz) %in% part1.de.genes.all),
                            part1.dc.merge$Sample[order(part1.dc.merge$DPT1, decreasing=TRUE)]]
part1.wnt.exprs$gene_id <- rownames(part1.wnt.exprs)
part1.wnt.merge <-  merge(part1.wnt.exprs, wnt_table, by.x='gene_id', by.y='ensembl_gene_id', all.x=TRUE)
rownames(part1.wnt.merge) <- part1.wnt.merge$external_gene_name
part1.wnt.merge <- part1.wnt.merge[, 2:(dim(part1.wnt.merge)[2]-1)]

part1.bmp.exprs <- part1.nz[(rownames(part1.nz) %in% bmp_table$ensembl_gene_id) & (rownames(part1.nz) %in% part1.de.genes.all),
                        part1.dc.merge$Sample[order(part1.dc.merge$DPT1, decreasing=TRUE)]]
part1.bmp.exprs$gene_id <- rownames(part1.bmp.exprs)
part1.bmp.merge <- merge(part1.bmp.exprs, bmp_table, by.x='gene_id', by.y='ensembl_gene_id', x.all=TRUE)
rownames(part1.bmp.merge) <- part1.bmp.merge$external_gene_name
part1.bmp.merge <- part1.bmp.merge[, 2:(dim(part1.bmp.merge)[2]-1)]

# heatmap?
part1.diff.exprs <- rbind(part1.wnt.merge, part1.bmp.merge)
part1.diff.exprs <- part1.diff.exprs[!duplicated(part1.diff.exprs), ]

part1.genes.df <- data.frame(cbind(rownames(part1.diff.exprs)))
rownames(part1.genes.df) <- part1.genes.df[, 1]
part1.genes.df$Pathway[rownames(part1.genes.df) %in% rownames(part1.bmp.merge)] <- "BMP"
part1.genes.df$Pathway[rownames(part1.genes.df) %in% rownames(part1.wnt.merge)] <- "Wnt"
part1.genes.df$Pathway <- as.factor(part1.genes.df$Pathway)

genes.cols <- c("#6a3d9a", "#ffff99")
names(genes.cols) <- levels(part1.genes.df$Pathway)
part1.genes.df <- as.data.frame(part1.genes.df[, -1])
rownames(part1.genes.df) <- rownames(part1.diff.exprs)
colnames(part1.genes.df) <- "Pathway"


#annot.df <- part1.dc.merge[, c("Cluster", "Kmeans")]
part1.annot.df <- part1.dc.merge[, c("Marker", "ClusterLabel", "DPT1")]
part1.annot.df$ClusterLabel <- as.factor(part1.annot.df$ClusterLabel)
#annot.df$Kmeans <- as.factor(annot.df$Kmeans)
part1.annot.df$DPT1 <- as.numeric(part1.annot.df$DPT1)
rownames(part1.annot.df) <- part1.dc.merge$Sample

dc.cols <- colorRampPalette(brewer.pal(10, "PRGn"))(dim(part1.diff.exprs)[2])
names(dc.cols) <- part1.annot.df$DPT1
  
marker.cols <- c("#ef3b2c","#662506")
names(marker.cols) <- as.factor(c("GFP+", "GFP-"))

cluster.cols <- label.map

annot.cols <- list(ClusterLabel=label.map,
                   DPT=dc.cols,
                   Marker=marker.cols,
                   Pathway=genes.cols)

pheatmap(part1.diff.exprs, show_rownames=TRUE, show_colnames=FALSE,
         cluster_rows=TRUE, annotation_col=part1.annot.df,
         annotation_row=part1.genes.df,
         annotation_colors=annot.cols,
         cluster_cols=FALSE)
```

The cells are ordered according to their position along the first diffusion component.  There is a noticeable switch-like behaviour for some genes, in that they are almost entirely absent, or expression is restricted to, the K-2 group.  These include genes involved in both BMP signalling and response, and Wnt signalling.  We might expect to see less Wnt signalling and more BMP stimulus in the more mature cells, as Wnt repressess adipogenesis in MSCs and BMPs generate promote adipocyte differentiation.  The genes off in K-2 are _Ndrg2_, _Ror1_, _Rspo3_, _Fzd7_, _Rgma_, _Snai1_, _Sdc1_, _Cpe_, _Bambi_.  The genes that are on are _Sfrp2_, _Sfrp4_ and _Bmp4_, all other other genes form a more continuous gradient of expression.  The most striking genes are _Cpe_ and _Bmp4_.  _Bmp4_ is probably expressed in more mature pre-adipocytes, which suggests this end of the pseudotime represents the more mature cells.  The cells on the left end are highest in _Myc_ and _Egr1_, so I wonder if they are actively proliferating cells.  If so, then I need to look at the impact of cell cycle as (a) a confounder, and (b) as a source of actively proliferating cells (can I also classify cells into particular parts of the cell cycle <- is that an informative analysis?).

Some of these genes look like they are quite strongly anti-correlated.  Is it possible to identify possible antagnostic gradients along this trajectory?

```{r, echo=FALSE, fig.height=9.5, fig.width=9.5}
part1.diff.cor <- cor(t(part1.diff.exprs), method='spearman')

pheatmap(part1.diff.cor,
         annotation_col=part1.genes.df,
         annotation_row=part1.genes.df,
         annotation_colors=annot.cols)
```

There are clearly some interesting behaviours amongst these genes, also indicated by the different groups of correlated genes.  Is it possible to identify genes that are associated with cell fate decisions?  For instance, _Cpe_ and _Bmp4_ have a switch-like behaviour that might suggest they are involved in making, or reinforcing a cell fate decision.  Can we also use this to determine at what point along a pseudotime that a decision is made?

Let's take a look at the wider correlation network of differentially expressed genes.  What patterns/clusters of co-expression are there?

This looks kind of cool.  It doesn't look so much like there are many discrete modules of gene expression _per se_, but gradients of gene expression changes.  This might be slightly artificial as these genes are selected for their differential expression between the middle of the pseudotime and the two arms.  Nonetheless it reinforces the idea that there is a continuum of development in these cells, and not necessarily discrete cell states.  It might be quite interesting to see how these genes compare to the ones with rapid onset/reduction in expression, and where they reach their peak/nadir along the diffusion component.  What functions do each of these clusters represent?  The blue, yellow and turquoise should be merged, but the others can stay as they are.  The background genes should be all of the DE genes, rather than all expressed genes in this case; I want to know what is specifically enriched in each cluster above and beyond what is enriched across all clusters.

## Cluster 1 up-regulated genes functional enrichment

```{r, echo=FALSE, fig.height=5.5, fig.width=12.5}
# use goseq
# use GO enrichment to test for specific pathways and function
# if there are enriched terms, this would at least imply some
# similar functional connection
# big cluster first
part1.all.genes <- unique(rownames(part1.nz))

part1.clust1.up <- unique(c(part1.de.merge.1v2$gene_id[part1.de.merge.1v2$Sig == 1 & (part1.de.merge.1v2$logFC < -1)],
                      part1.de.merge.3v1$gene_id[part1.de.merge.3v1$Sig == 1 & (part1.de.merge.3v1$logFC < -1)],
                      part1.de.merge.4v1$gene_id[part1.de.merge.4v1$Sig == 1 & (part1.de.merge.4v1$logFC < -1)]))

part1.clust2.up <- unique(c(part1.de.merge.1v2$gene_id[part1.de.merge.1v2$Sig == 1 & (part1.de.merge.1v2$logFC > 1)],
                      part1.de.merge.3v2$gene_id[part1.de.merge.3v2$Sig == 1 & (part1.de.merge.3v2$logFC < -1)],
                      part1.de.merge.4v2$gene_id[part1.de.merge.4v2$Sig == 1 & (part1.de.merge.4v2$logFC < -1)]))

part1.clust3.up <- unique(c(part1.de.merge.3v1$gene_id[part1.de.merge.1v2$Sig == 1 & (part1.de.merge.1v2$logFC > 1)],
                      part1.de.merge.3v2$gene_id[part1.de.merge.3v2$Sig == 1 & (part1.de.merge.3v2$logFC > 1)],
                      part1.de.merge.4v3$gene_id[part1.de.merge.4v3$Sig == 1 & (part1.de.merge.4v3$logFC < -1)]))

part1.clust4.up <- unique(c(part1.de.merge.4v1$gene_id[part1.de.merge.4v1$Sig == 1 & (part1.de.merge.4v1$logFC > 1)],
                            part1.de.merge.4v2$gene_id[part1.de.merge.4v2$Sig == 1 & (part1.de.merge.4v2$logFC > 1)],
                            part1.de.merge.4v3$gene_id[part1.de.merge.4v3$Sig == 1 & (part1.de.merge.4v3$logFC > 1)]))

part1.clust1.genes <- setdiff(part1.clust1.up, unique(c(part1.clust2.up, part1.clust3.up, part1.clust4.up)))
part1.clust2.genes <- setdiff(part1.clust2.up, unique(c(part1.clust1.up, part1.clust3.up, part1.clust4.up)))
part1.clust3.genes <- setdiff(part1.clust3.up, unique(c(part1.clust2.up, part1.clust1.up, part1.clust4.up)))
part1.clust4.genes <- setdiff(part1.clust4.up, unique(c(part1.clust2.up, part1.clust3.up, part1.clust1.up)))

# blue.genes <- rownames(annot.df)[annot.df$Cluster %in% c("blue")]
# yellow.genes <- rownames(annot.df)[annot.df$Cluster %in% c("yellow")]
# turquoise.genes <- rownames(annot.df)[annot.df$Cluster %in% c("turquoise")]
# big.cluster.genes <- rownames(annot.df)[annot.df$Cluster %in% c("blue", "turquoise", "black")]
# brown.genes <- rownames(annot.df)[annot.df$Cluster %in% c("brown")]
# pink.genes <- rownames(annot.df)[annot.df$Cluster %in% c("pink")]
# red.genes <- rownames(annot.df)[annot.df$Cluster %in% c("red")]
# black.genes <- rownames(annot.df)[annot.df$Cluster %in% c("black")]
# #magenta.genes <- rownames(annot.df)[annot.df$Cluster %in% c("magenta")]
# green.genes <- rownames(annot.df)[annot.df$Cluster %in% c("green")]
# #grey.genes <- rownames(annot.df)[annot.df$Cluster %in% c("grey")]

de.big.vector <- as.integer(unique(part1.all.genes) %in% unique(part1.clust1.up))
names(de.big.vector) <- unique(part1.all.genes)

de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part1.gene.length))]

pwf.all <- nullp(de.big.vector, "mm10", "ensGene",
                 bias.data=part1.gene.length[intersect(names(de.big.vector), names(part1.gene.length))],
                 plot.fit=FALSE)
part1.GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
                           method="Wallenius", use_genes_without_cat=T)
  
# part1.GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
#                            method="Wallenius", use_genes_without_cat=T)

part1.GO.wall.all$padjust <- p.adjust(part1.GO.wall.all$over_represented_pvalue, method="BH")
part1.GO.wall.all$foldEnrich <- ((part1.GO.wall.all$numDEInCat/length(unique(part1.clust1.up)))/(part1.GO.wall.all$numInCat/length(part1.gene.length)))

# select on p-value and foldEnrich
part1.GO.sig.all <- part1.GO.wall.all[(part1.GO.wall.all$padjust <= 0.01), ]
part1.go_cats.all <- data.frame(part1.GO.sig.all$category, part1.GO.sig.all$foldEnrich)

part1.p.enrich <- ggplot(part1.GO.sig.all,
                   aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
                       size=foldEnrich)) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C")

part1.GO.all <- part1.GO.sig.all[order(part1.GO.sig.all$padjust, decreasing=FALSE), ]

part1.p.terms <- ggplot(na.omit(part1.GO.all[1:25, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=60)) +
  coord_flip() + labs(x="Gene Ontology category", y="Fold Enrichment")

plot_grid(part1.p.enrich, part1.p.terms, ncol=2, rel_widths=c(0.4, 0.6))
```

Mostly enriched for inflammatory response, proliferation, development and adhesion.


```{r}
part1.c1.pterms <- ggplot(na.omit(part1.GO.all[1:5, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1, breaks=c(min(round(-log10(part1.GO.all[1:5, ]$padjust), 1)),
                                                             max(round(-log10(part1.GO.all[1:5, ]$padjust), 1)))) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=45)) +
  coord_flip() + labs(x="Gene Ontology category", y="Fold Enrichment") +
  theme(axis.text=element_text(size=18))

ggsave(part1.c1.pterms,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Cluster1_GOenrich-bar-Kmeans1.png",
       height=4.75, width=6.75, dpi=300)
```

The genes up-regulated in cluster 1 are largely enriched for cytokine and immunne response genes.

```{r, echo=FALSE, fig.height=4.5, fig.width=5.5}
psorder <- part1.dc.merge$Sample[order(part1.dc.merge$DPT1, decreasing=TRUE)]
part1.annot.df <- cbind.data.frame(part1.dc.merge[, c("ClusterLabel")])
colnames(part1.annot.df) <- c("ClusterLabel")
part1.annot.df$ClusterLabel <- as.factor(part1.annot.df$ClusterLabel)
rownames(part1.annot.df) <- part1.dc.merge$Sample

pheatmap(part1.nz[part1.clust1.up, psorder],
         show_rownames=FALSE, show_colnames=FALSE,
         cluster_cols=FALSE, annotation_col=part1.annot.df,
         annotation_colors=list(ClusterLabel=label.map))
```


#### Cluster 2 genes functional enrichment

```{r, echo=FALSE, fig.height=5.5, fig.width=12.5}
de.big.vector <- as.integer(unique(part1.all.genes) %in% unique(part1.clust2.up))
names(de.big.vector) <- unique(part1.all.genes)

de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part1.gene.length))]

pwf.all <- nullp(de.big.vector, "mm10", "ensGene", 
                 bias.data=part1.gene.length[intersect(names(de.big.vector), names(part1.gene.length))],
                 plot.fit=FALSE)
part1.GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
                           method="Wallenius", use_genes_without_cat=T)
# part1.GO.wall.all<- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
#                       method="Wallenius", use_genes_without_cat=T)

part1.GO.wall.all$padjust <- p.adjust(part1.GO.wall.all$over_represented_pvalue, method="BH")
part1.GO.wall.all$foldEnrich <- ((part1.GO.wall.all$numDEInCat/length(unique(part1.clust2.up)))/(part1.GO.wall.all$numInCat/length(part1.gene.length)))

# select on p-value and foldEnrich
part1.GO.sig.all <- part1.GO.wall.all[(part1.GO.wall.all$padjust <= 0.01), ]
part1.go_cats.all <- data.frame(part1.GO.sig.all$category, part1.GO.sig.all$foldEnrich)

part1.p.enrich <- ggplot(part1.GO.sig.all,
                   aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
                       size=foldEnrich)) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C")

part1.GO.all <- part1.GO.sig.all[order(part1.GO.sig.all$padjust, decreasing=FALSE), ]

part1.p.terms <- ggplot(na.omit(part1.GO.all[1:20, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")

plot_grid(part1.p.enrich, part1.p.terms, ncol=2)
```

Strong enrichments for cell adhesion and signalling in cluster 2.

```{r}
part1.c2.pterms <- ggplot(na.omit(part1.GO.all[1:5, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1, breaks=c(min(round(-log10(part1.GO.all[1:5, ]$padjust), 1)),
                                                             max(round(-log10(part1.GO.all[1:5, ]$padjust), 1)))) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=45)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment") +
  theme(axis.text=element_text(size=18))

ggsave(part1.c2.pterms,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Cluster2_GOenrich-bar-Kmeans1.png",
       height=4.75, width=6.75, dpi=300)
```


```{r, echo=FALSE, fig.height=4.5, fig.width=5.5}
pheatmap(part1.nz[part1.clust2.up, psorder],
         show_rownames=FALSE, show_colnames=FALSE,
         cluster_cols=FALSE, annotation_col=part1.annot.df,
         annotation_colors=list(ClusterLabel=label.map))
```

#### Cluster 3 functional enrichment

```{r, echo=FALSE, fig.height=5.5, fig.width=12.5}
de.big.vector <- as.integer(unique(part1.all.genes) %in% unique(part1.clust3.up))
names(de.big.vector) <- unique(part1.all.genes)

de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part1.gene.length))]

pwf.all <- nullp(de.big.vector, "mm10", "ensGene", 
                 bias.data=part1.gene.length[intersect(names(de.big.vector), names(part1.gene.length))],
                 plot.fit=FALSE)
part1.GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
                           method="Wallenius", use_genes_without_cat=T)
# part1.GO.wall.all<- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
#                       method="Wallenius", use_genes_without_cat=T)

part1.GO.wall.all$padjust <- p.adjust(part1.GO.wall.all$over_represented_pvalue, method="BH")
part1.GO.wall.all$foldEnrich <- ((part1.GO.wall.all$numDEInCat/length(unique(part1.clust3.up)))/(part1.GO.wall.all$numInCat/length(part1.gene.length)))

# select on p-value and foldEnrich
part1.GO.sig.all <- part1.GO.wall.all[(part1.GO.wall.all$padjust <= 0.01), ]
part1.go_cats.all <- data.frame(part1.GO.sig.all$category, part1.GO.sig.all$foldEnrich)

part1.p.enrich <- ggplot(part1.GO.sig.all,
                   aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
                       size=foldEnrich)) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C")

part1.GO.all <- part1.GO.sig.all[order(part1.GO.sig.all$padjust, decreasing=FALSE), ]

part1.p.terms <- ggplot(na.omit(part1.GO.all[1:20, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")

plot_grid(part1.p.enrich, part1.p.terms, ncol=2)
```

There is a strong enrichment for developmental processes here.

```{r}
part1.c3.pterms <- ggplot(na.omit(part1.GO.all[1:5, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1, breaks=c(min(round(-log10(part1.GO.all[1:5, ]$padjust), 1)),
                                                             max(round(-log10(part1.GO.all[1:5, ]$padjust), 1)))) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=45)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment") +
  theme(axis.text=element_text(size=18))

ggsave(part1.c3.pterms,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Cluster3_GOenrich-bar-Kmeans1.png",
       height=4.75, width=6.75, dpi=300)
```



```{r, echo=FALSE, fig.height=4.5, fig.width=5.5}
pheatmap(part1.nz[part1.clust3.up, psorder],
         show_rownames=FALSE, show_colnames=FALSE,
         cluster_cols=FALSE, annotation_col=part1.annot.df,
         annotation_colors=list(ClusterLabel=label.map))
```

These are all quite strongly expressed and broadly expressed; it is hard to see where the differences arise.

#### Cluster 4 functional enrichment

```{r, echo=FALSE, fig.height=5.5, fig.width=12.5}
de.big.vector <- as.integer(unique(part1.all.genes) %in% unique(part1.clust4.up))
names(de.big.vector) <- unique(part1.all.genes)

de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part1.gene.length))]

pwf.all <- nullp(de.big.vector, "mm10", "ensGene", 
                 bias.data=part1.gene.length[intersect(names(de.big.vector), names(part1.gene.length))],
                 plot.fit=FALSE)
part1.GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
                           method="Wallenius", use_genes_without_cat=T)

# part1.GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
#                       method="Wallenius", use_genes_without_cat=T)

part1.GO.wall.all$padjust <- p.adjust(part1.GO.wall.all$over_represented_pvalue, method="BH")
part1.GO.wall.all$foldEnrich <- ((part1.GO.wall.all$numDEInCat/length(unique(part1.clust4.up)))/(part1.GO.wall.all$numInCat/length(part1.gene.length)))

# select on p-value and foldEnrich
part1.GO.sig.all <- part1.GO.wall.all[(part1.GO.wall.all$padjust <= 0.01), ]
part1.go_cats.all <- data.frame(part1.GO.sig.all$category, part1.GO.sig.all$foldEnrich)

part1.p.enrich <- ggplot(part1.GO.sig.all,
                   aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
                       size=foldEnrich)) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C")

part1.GO.all <- part1.GO.sig.all[order(part1.GO.sig.all$padjust, decreasing=FALSE), ]

part1.p.terms <- ggplot(na.omit(part1.GO.all[1:20, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")

plot_grid(part1.p.enrich, part1.p.terms, ncol=2)
```

Cell adhesions seems to dominated quite a lot.

```{r}
part1.c4.pterms <- ggplot(na.omit(part1.GO.all[1:5, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1, breaks=c(min(round(-log10(part1.GO.all[1:5, ]$padjust), 1)),
                                                             max(round(-log10(part1.GO.all[1:5, ]$padjust), 1)))) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=45)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment") +
  theme(axis.text=element_text(size=18))

ggsave(part1.c4.pterms,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Cluster4_GOenrich-bar-Kmeans1.png",
       height=4.75, width=6.75, dpi=300)
```



```{r, echo=FALSE, fig.height=4.5, fig.width=5.5}
pheatmap(part1.nz[part1.clust4.up, psorder],
         show_rownames=FALSE, show_colnames=FALSE,
         cluster_cols=FALSE, annotation_col=part1.annot.df,
         annotation_colors=list(ClusterLabel=label.map))
```

Just looking at the DE genes between clusters isn't necessarily that informative.  Perhaps finds genes that are DE along each branch separately.

Firstly I want to consider each partition as a single trajectory, then I'll split into 2 branches.

## Differential Expression Analysis: Partition 1 pseudotime

```{r, echo=FALSE}
all.ptime <- part1.dc.merge$Sample
design.mat.alltime <- model.matrix(~ DPT1 + Plate, data=part1.dc.merge[part1.dc.merge$Sample %in% all.ptime, ])

# fit a linear model
fit.alltime <- lmFit(part1.nz[, colnames(part1.nz) %in% all.ptime], design.mat.alltime)
fit.alltime <- eBayes(fit.alltime)
sum.res.alltime <- summary(decideTests(fit.alltime))

de.res.alltime <- topTable(fit.alltime, coef=2, n=Inf, sort="p", p=1.0)
de.res.alltime$Sig <- 0
de.res.alltime$Sig[de.res.alltime$adj.P.Val <= 0.01] <- 1
de.res.alltime$Sig <- as.factor(de.res.alltime$Sig)

de.res.alltime$Diff <- 0
de.res.alltime$Diff[de.res.alltime$logFC < 0 & de.res.alltime$Sig == 1] <- -1
de.res.alltime$Diff[de.res.alltime$logFC > 0 & de.res.alltime$Sig == 1] <- 1

de.res.alltime$gene_id <- rownames(de.res.alltime)
de.merge.alltime <- merge(de.res.alltime, gene_symbol,
                  by.x='gene_id', by.y='ensembl_gene_id', all.x=TRUE)
gene.top.alltime <- de.merge.alltime$external_gene_name

# plot the top 6 and bottom 6 gene symbols
top6 <- de.merge.alltime$external_gene_name[order(de.merge.alltime$logFC, decreasing=TRUE)][1:6]
bottom6 <- de.merge.alltime$external_gene_name[order(de.merge.alltime$logFC, decreasing=FALSE)][1:6]
gene.top.alltime[!gene.top.alltime %in% c(top6, bottom6)] <- ""

write.table(de.merge.alltime,
            file="~/Dropbox/Adipose/Adipose-DEgenes_Allptime.tsv",
            sep="\t", quote=F)

all.time.ma <- ggplot(de.merge.alltime, aes(x=AveExpr, y=logFC, colour=Sig)) +
  geom_point(size=1, alpha=0.8) + theme_mike() +
  scale_colour_manual(values=c("#A3A3A3", "#FF0000")) +
  labs(x=expression(paste("Mean log"[2], " Expression")),
       y=expression(paste("log"[2], " Fold Change"))) +
  guides('colour'=FALSE) +
  ylim(c(-25, 25)) +
  geom_label_repel(aes(label=gene.top.alltime),
                  colour='black',
                  nudge_x=0.01,
                  nudge_y=0.01,
                  size=3)

ggsave(all.time.ma,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/MAplot-Kmeans1-Pseudotime.png",
       height=4.25, width=6.75, dpi=300)

all.time.ma
```


```{r}
sum.res.alltime
```


#### Partition 1 pseudotime functional enrichment

#### Up-regulated processes

```{r, echo=FALSE, fig.height=5.5, fig.width=12.5}
alltime.up <- de.merge.alltime$gene_id[de.merge.alltime$Sig == 1 & de.merge.alltime$logFC > 0]
alltime.down <- de.merge.alltime$gene_id[de.merge.alltime$Sig == 1 & de.merge.alltime$logFC < 0]

de.big.vector <- as.integer(unique(part1.all.genes) %in% unique(alltime.up))
names(de.big.vector) <- unique(part1.all.genes)

de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part1.gene.length))]

pwf.all <- nullp(de.big.vector, "mm10", "ensGene", bias.data=part1.gene.length[intersect(names(de.big.vector), names(part1.gene.length))],
                 plot.fit=FALSE)
GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
                           method="Wallenius", use_genes_without_cat=T)
# GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
#                       method="Wallenius", use_genes_without_cat=T)

GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(alltime.up)))/(GO.wall.all$numInCat/length(part1.gene.length)))

# select on p-value and foldEnrich
GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)

p.enrich <- ggplot(GO.sig.all,
                   aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
                       size=foldEnrich)) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C")

GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(GO.all[1:20, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")

plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.2, 0.8))
```


#### Down-regulated processes

```{r, echo=FALSE, fig.height=5.5, fig.width=12.5}
de.big.vector <- as.integer(unique(part1.all.genes) %in% unique(alltime.down))
names(de.big.vector) <- unique(part1.all.genes)

de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part1.gene.length))]

pwf.all <- nullp(de.big.vector, "mm10", "ensGene", bias.data=part1.gene.length[intersect(names(de.big.vector), names(part1.gene.length))],
                 plot.fit=FALSE)
GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
                           method="Wallenius", use_genes_without_cat=T)

# GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
#                       method="Wallenius", use_genes_without_cat=T)

GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(alltime.down)))/(GO.wall.all$numInCat/length(part1.gene.length)))

# select on p-value and foldEnrich
GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)

p.enrich <- ggplot(GO.sig.all,
                   aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
                       size=foldEnrich)) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C")

GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(GO.all[1:20, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")

plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.2, 0.8))
```

There are quite a lot of enriched terms in each of the up and down-regulated processes along the left branch of the trajectory.  If these are co-ordinated changes then we should be able to cluster the genes together, and find modules of genes that share functional annotations.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
all.diff.exprs <- part1.nz[unique(c(alltime.down, alltime.up)),
                              !colnames(part1.nz) %in% multi.cell]
all.de.cor <- cor(t(all.diff.exprs), method='spearman')

all.de.dist <- as.dist(abs(all.de.cor - 1))
all.de.clust <- flashClust(all.de.dist, method="average")
all.de.cut <- cutreeDynamicTree(all.de.clust, deepSplit=TRUE, minModuleSize=30)
all.de.cols <- labels2colors(all.de.cut)

all.annot.df <- data.frame(cbind(all.de.cols))
rownames(all.annot.df) <- rownames(all.diff.exprs)
colnames(all.annot.df) <- "Cluster"
all.gene.clust.cols <- col2hcl(unique(all.de.cols))
names(all.gene.clust.cols) <- unique(all.de.cols)
all.annot.cols <- list("Cluster"=all.gene.clust.cols)

pheatmap(all.de.cor,
         show_rownames=FALSE, show_colnames=FALSE,
         annotation_col=all.annot.df, annotation_colors=all.annot.cols,
         clustering_distance_cols=all.de.dist,
         clustering_distance_rows=all.de.dist,
         clustering_method="average",
         filename="~/Dropbox/Adipose/Adipose-Kmeans1-Pseudotime_DEgenes-heatmap.png",
         height=5.75, width=5.75, res=300)
```

For the most part there are genes that go up and genes that go down.


```{r, echo=FALSE, warning=FALSE, message=FALSE}
all.annot.df$gene_id <- rownames(all.annot.df)

write.table(all.annot.df,
            file="~/Dropbox/Adipose/Adipose_Kmeans1-Pseudotime_DEgenes-clusters.tsv",
            sep="\t", quote=FALSE, row.names=FALSE)

all.blue.genes <- rownames(all.annot.df)[all.annot.df$Cluster == 'blue']
all.brown.genes <- rownames(all.annot.df)[all.annot.df$Cluster == 'brown']
all.turquoise.genes <- rownames(all.annot.df)[all.annot.df$Cluster == 'turquoise']
all.yellow.genes <- rownames(all.annot.df)[all.annot.df$Cluster == 'yellow']

all.black.genes <- rownames(all.annot.df)[all.annot.df$Cluster == 'black']
all.green.genes <- rownames(all.annot.df)[all.annot.df$Cluster == 'green']
all.red.genes <- rownames(all.annot.df)[all.annot.df$Cluster == 'red']
```


#### Functional enrichment test: blue cluster

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.5, fig.width=12.5}
de.big.vector <- as.integer(unique(part1.all.genes) %in% unique(all.blue.genes))
names(de.big.vector) <- unique(part1.all.genes)

de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part1.gene.length))]

pwf.all <- nullp(de.big.vector, "mm10", "ensGene", bias.data=part1.gene.length[intersect(names(de.big.vector), names(part1.gene.length))],
                 plot.fit=FALSE)
GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
                           method="Wallenius", use_genes_without_cat=T)

# GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
#                       method="Wallenius", use_genes_without_cat=T)

GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(all.blue.genes)))/(GO.wall.all$numInCat/length(part1.gene.length)))

# select on p-value and foldEnrich
GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)

p.enrich <- ggplot(GO.sig.all,
                   aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
                       size=foldEnrich)) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C")

GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(GO.all[1:20, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")

write.table(GO.all,
            file="~/Dropbox/Adipose/Adipose-Kmeans1-Pseudotime_GOenrich-blueCluster.tsv",
            sep="\t", row.names=FALSE, quote=FALSE)

plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.3, 0.7))
```

#### Functional enrichment test: yellow cluster

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.5, fig.width=12.5}
de.big.vector <- as.integer(unique(part1.all.genes) %in% unique(all.yellow.genes))
names(de.big.vector) <- unique(part1.all.genes)

de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part1.gene.length))]

pwf.all <- nullp(de.big.vector, "mm10", "ensGene", bias.data=part1.gene.length[intersect(names(de.big.vector), names(part1.gene.length))],
                 plot.fit=FALSE)
GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
                           method="Wallenius", use_genes_without_cat=T)

# GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
#                       method="Wallenius", use_genes_without_cat=T)

GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(all.yellow.genes)))/(GO.wall.all$numInCat/length(part1.gene.length)))

# select on p-value and foldEnrich
GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)

p.enrich <- ggplot(GO.sig.all,
                   aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
                       size=foldEnrich)) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C")

GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(GO.all[1:20, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")

write.table(GO.all,
            file="~/Dropbox/Adipose/Adipose-Kmeans1-Pseudotime_GOenrich-yellowCluster.tsv",
            sep="\t", row.names=FALSE, quote=FALSE)

plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.3, 0.7))
```

#### Functional enrichment test: brown cluster

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.5, fig.width=12.5}
de.big.vector <- as.integer(unique(part1.all.genes) %in% unique(all.brown.genes))
names(de.big.vector) <- unique(part1.all.genes)

de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part1.gene.length))]

pwf.all <- nullp(de.big.vector, "mm10", "ensGene", bias.data=part1.gene.length[intersect(names(de.big.vector), names(part1.gene.length))],
                 plot.fit=FALSE)
GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
                           method="Wallenius", use_genes_without_cat=T)

# GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
#                       method="Wallenius", use_genes_without_cat=T)

GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(all.brown.genes)))/(GO.wall.all$numInCat/length(part1.gene.length)))

# select on p-value and foldEnrich
GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)

p.enrich <- ggplot(GO.sig.all,
                   aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
                       size=foldEnrich)) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C")

GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(GO.all[1:20, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")

write.table(GO.all,
            file="~/Dropbox/Adipose/Adipose-Kmeans1-Pseudotime_GOenrich-brownCluster.tsv",
            sep="\t", row.names=FALSE, quote=FALSE)

plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.3, 0.7))
```

#### Functional enrichment test: turquoise cluster

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.5, fig.width=12.5}
de.big.vector <- as.integer(unique(part1.all.genes) %in% unique(all.turquoise.genes))
names(de.big.vector) <- unique(part1.all.genes)

de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part1.gene.length))]

pwf.all <- nullp(de.big.vector, "mm10", "ensGene", bias.data=part1.gene.length[intersect(names(de.big.vector), names(part1.gene.length))],
                 plot.fit=FALSE)
GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
                           method="Wallenius", use_genes_without_cat=T)

# GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
#                       method="Wallenius", use_genes_without_cat=T)

GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(all.turquoise.genes)))/(GO.wall.all$numInCat/length(part1.gene.length)))

# select on p-value and foldEnrich
GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)

p.enrich <- ggplot(GO.sig.all,
                   aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
                       size=foldEnrich)) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C")

GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(GO.all[1:20, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=35)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")

write.table(GO.all,
            file="~/Dropbox/Adipose/Adipose-Kmeans1-Pseudotime_GOenrich-turquoiseCluster.tsv",
            sep="\t", row.names=FALSE, quote=FALSE)

plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.2, 0.8))
```

#### Functional enrichment test: black cluster

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.5, fig.width=12.5}
de.big.vector <- as.integer(unique(part1.all.genes) %in% unique(all.black.genes))
names(de.big.vector) <- unique(part1.all.genes)

de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part1.gene.length))]

pwf.all <- nullp(de.big.vector, "mm10", "ensGene", bias.data=part1.gene.length[intersect(names(de.big.vector), names(part1.gene.length))],
                 plot.fit=FALSE)
GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
                           method="Wallenius", use_genes_without_cat=T)

# GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
#                       method="Wallenius", use_genes_without_cat=T)

GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(all.black.genes)))/(GO.wall.all$numInCat/length(part1.gene.length)))

# select on p-value and foldEnrich
GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)

p.enrich <- ggplot(GO.sig.all,
                   aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
                       size=foldEnrich)) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C")

GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(GO.all[1:20, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")

write.table(GO.all,
            file="~/Dropbox/Adipose/Adipose-Kmeans1-Pseudotime_GOenrich-blackCluster.tsv",
            sep="\t", row.names=FALSE, quote=FALSE)

plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.3, 0.7))
```

#### Functional enrichment test: green cluster

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.5, fig.width=12.5}
de.big.vector <- as.integer(unique(part1.all.genes) %in% unique(all.green.genes))
names(de.big.vector) <- unique(part1.all.genes)

de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part1.gene.length))]

pwf.all <- nullp(de.big.vector, "mm10", "ensGene", bias.data=part1.gene.length[intersect(names(de.big.vector), names(part1.gene.length))],
                 plot.fit=FALSE)
GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
                           method="Wallenius", use_genes_without_cat=T)

# GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
#                       method="Wallenius", use_genes_without_cat=T)

GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(all.green.genes)))/(GO.wall.all$numInCat/length(part1.gene.length)))

# select on p-value and foldEnrich
GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)

p.enrich <- ggplot(GO.sig.all,
                   aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
                       size=foldEnrich)) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C")

GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(GO.all[1:20, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")

write.table(GO.all,
            file="~/Dropbox/Adipose/Adipose-Kmeans1-Pseudotime_GOenrich-greenCluster.tsv",
            sep="\t", row.names=FALSE, quote=FALSE)

plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.3, 0.7))
```

#### Functional enrichment test: red cluster

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.5, fig.width=12.5}
de.big.vector <- as.integer(unique(part1.all.genes) %in% unique(all.red.genes))
names(de.big.vector) <- unique(part1.all.genes)

de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part1.gene.length))]

pwf.all <- nullp(de.big.vector, "mm10", "ensGene", bias.data=part1.gene.length[intersect(names(de.big.vector), names(part1.gene.length))],
                 plot.fit=FALSE)
GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
                           method="Wallenius", use_genes_without_cat=T)

# GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
#                       method="Wallenius", use_genes_without_cat=T)

GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(all.red.genes)))/(GO.wall.all$numInCat/length(part1.gene.length)))

# select on p-value and foldEnrich
GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)

p.enrich <- ggplot(GO.sig.all,
                   aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
                       size=foldEnrich)) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C")

GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(GO.all[1:20, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")

write.table(GO.all,
            file="~/Dropbox/Adipose/Adipose-Kmeans1-Pseudotime_GOenrich-greenCluster.tsv",
            sep="\t", row.names=FALSE, quote=FALSE)

plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.3, 0.7))
```

I want cluster the pseudotime-dependent genes into whether they are highest in the naive, intermediate or most mature 
cells.  I could do this by selecting genes that fit a particular shape curve over the diffusion pseudotime.

```{r, echo=FALSE, fig.height=6.5, fig.width=6.5}
# heatmap of partition 1 pseudotime-dependent DE genes
part1.diff.genes <- unique(de.merge.alltime$gene_id[de.merge.alltime$Sig == 1 & (abs(de.merge.alltime$logFC) > 3)])

part1.blue.genes <- intersect(unique(part1.diff.genes), all.blue.genes)
part1.yellow.genes <- intersect(unique(part1.diff.genes), all.yellow.genes)
part1.turq.genes <- intersect(unique(part1.diff.genes), all.turquoise.genes)
part1.brown.genes <- intersect(unique(part1.diff.genes), all.brown.genes)
part1.red.genes <- intersect(unique(part1.diff.genes), all.red.genes)
part1.green.genes <- intersect(unique(part1.diff.genes), all.green.genes)
part1.black.genes <- intersect(unique(part1.diff.genes), all.black.genes)

# only plot the genes unique to each cluster
part1.diff.exprs <- part1.nz[part1.diff.genes, part1.dc.merge$Sample[order(part1.dc.merge$DPT1, decreasing=TRUE)]]

part1.genes.df <- data.frame(cbind(part1.diff.genes))
colnames(part1.genes.df) <- c("gene_id")
rownames(part1.genes.df) <- part1.genes.df[, 1]
part1.genes.df$GeneGroup <- "Late"
part1.genes.df[part1.genes.df$gene_id %in% c(part1.blue.genes, part1.green.genes, part1.red.genes), ]$GeneGroup <- "Late"
part1.genes.df[part1.genes.df$gene_id %in% c(part1.turq.genes, part1.black.genes, part1.yellow.genes), ]$GeneGroup <- "Early"
part1.genes.df[part1.genes.df$gene_id %in% c(part1.brown.genes), ]$GeneGroup <- "Intermediate"

part1.genes.df$GeneGroup <- factor(part1.genes.df$GeneGroup,
                                   levels=c("Early", "Intermediate", "Late"),
                                   labels=c("Early", "Intermediate", "Late"))

#genes.cols <- 
genes.cols <- c("#F700FF", "#00F7FF", "#FF9E00")
names(genes.cols) <- c("Late", "Early", "Intermediate")
part1.genes.df <- as.data.frame(part1.genes.df[, -1])
rownames(part1.genes.df) <- part1.diff.genes
colnames(part1.genes.df) <- "GeneGroup"

part1.annot.df <- part1.dc.merge[, c("ClusterLabel", "DPT1")]
part1.annot.df$ClusterLabel <- as.factor(part1.annot.df$ClusterLabel)
part1.annot.df$DPT1 <- as.numeric(part1.annot.df$DPT1)
rownames(part1.annot.df) <- part1.dc.merge$Sample

dc.cols <- colorRampPalette(c("#FAFF60", "#8615B0"))(dim(part1.diff.exprs)[2])
names(dc.cols) <- part1.annot.df$DPT1
  
marker.cols <- gfp.cols
names(marker.cols) <- as.factor(c("GFP+", "GFP-"))

cluster.cols <- label.map

part1.annot.cols <- list(ClusterLabel=cluster.cols,
                         GeneGroup=genes.cols,
                         DPT1=dc.cols)

gene.order <- order(part1.genes.df$GeneGroup)

pheatmap(part1.diff.exprs[gene.order, ], show_rownames=FALSE, show_colnames=FALSE,
         cluster_rows=FALSE, annotation_col=part1.annot.df,
         annotation_row=part1.genes.df,
         annotation_colors=part1.annot.cols,
         cluster_cols=FALSE,
         filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Pseudotime_DE_genes-Kmeans1-heatmap.png",
         height=6.75, width=6.75, res=300)
```


#### Differential Expression Analysis: Left branch pseudotime with clusters 2 & 4

```{r, echo=FALSE}
# lptime <- part1.dc.merge$Sample[part1.dc.merge$Cluster %in% c(2, 4)]
# design.mat.lptime <- model.matrix(~ DPT1 + Plate, data=part1.dc.merge[part1.dc.merge$Sample %in% lptime, ])
# 
# # fit a linear model
# fit.lptime <- lmFit(part1.nz[, colnames(part1.nz) %in% lptime], design.mat.lptime)
# fit.lptime <- eBayes(fit.lptime)
# sum.res.lptime <- summary(decideTests(fit.lptime))
# 
# de.res.lptime <- topTable(fit.lptime, coef=2, n=Inf, sort="p", p=1.0)
# de.res.lptime$Sig <- 0
# de.res.lptime$Sig[de.res.lptime$adj.P.Val <= 0.01] <- 1
# de.res.lptime$Sig <- as.factor(de.res.lptime$Sig)
# 
# de.res.lptime$Diff <- 0
# de.res.lptime$Diff[de.res.lptime$logFC < 0 & de.res.lptime$Sig == 1] <- -1
# de.res.lptime$Diff[de.res.lptime$logFC > 0 & de.res.lptime$Sig == 1] <- 1
# 
# de.res.lptime$gene_id <- rownames(de.res.lptime)
# de.merge.lptime <- merge(de.res.lptime, gene_symbol,
#                   by.x='gene_id', by.y='ensembl_gene_id', all.x=TRUE)
# gene.top.lptime <- de.merge.lptime$external_gene_name
# 
# # plot the top 6 and bottom 6 gene symbols
# top6 <- de.merge.lptime$external_gene_name[order(de.merge.lptime$logFC, decreasing=TRUE)][1:6]
# bottom6 <- de.merge.lptime$external_gene_name[order(de.merge.lptime$logFC, decreasing=FALSE)][1:6]
# gene.top.lptime[!gene.top.lptime %in% c(top6, bottom6)] <- ""
# 
# # gene.top.lptime[de.merge.lptime$logFC > -10 & de.merge.lptime$logFC < 10] <- ""
# # gene.top.lptime[is.na(gene.top.lptime)] <- ""
# 
# write.table(de.merge.lptime,
#             file="~/Dropbox/Adipose/Adipose-DEgenes_Leftptime.tsv",
#             sep="\t", quote=F)
# 
# ltime.ma <- ggplot(de.merge.lptime, aes(x=AveExpr, y=logFC, colour=Sig)) +
#   geom_point(size=1, alpha=0.8) + theme_mike() +
#   scale_colour_manual(values=c("#A3A3A3", "#FF0000")) +
#   labs(x=expression(paste("Mean log"[2], " Expression")),
#        y=expression(paste("log"[2], " Fold Change"))) +
#   guides('colour'=FALSE) +
#   ylim(c(-25, 25)) +
#   geom_label_repel(aes(label=gene.top.lptime),
#                   colour='black',
#                   nudge_x=0.01,
#                   nudge_y=0.01,
#                   size=3)
# 
# ggsave(ltime.ma,
#        filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/MAplot-LeftPseudotime.png",
#        height=4.25, width=6.75, dpi=300)
# 
# ltime.ma
```


```{r}
# sum.res.lptime
```


There are a fairly decent number of genes DE along the pseudotime (~linearly).

#### Left pseudotime functional enrichment

#### Up-regulated processes

```{r, echo=FALSE, fig.height=5.5, fig.width=12.5}
# ltime.up <- de.merge.lptime$gene_id[de.merge.lptime$Sig == 1 & de.merge.lptime$logFC > 0]
# ltime.down <- de.merge.lptime$gene_id[de.merge.lptime$Sig == 1 & de.merge.lptime$logFC < 0]
# 
# de.big.vector <- as.integer(unique(part1.all.genes) %in% unique(ltime.up))
# names(de.big.vector) <- unique(part1.all.genes)
# 
# de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part1.gene.length))]
# 
# pwf.all <- nullp(de.big.vector, "mm10", "ensGene", bias.data=part1.gene.length[intersect(names(de.big.vector), names(part1.gene.length))],
#                  plot.fit=FALSE)
# GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
#                            method="Wallenius", use_genes_without_cat=T)
# # GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
# #                       method="Wallenius", use_genes_without_cat=T)
# 
# GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
# GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(ltime.up)))/(GO.wall.all$numInCat/length(part1.gene.length)))
# 
# # select on p-value and foldEnrich
# GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
# go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)
# 
# p.enrich <- ggplot(GO.sig.all,
#                    aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
#                        size=foldEnrich)) +
#   geom_point() + theme_mike() +
#   scale_colour_gradient(low="#000000", high="#DC143C")
# 
# GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]
# 
# p.terms <- ggplot(na.omit(GO.all[1:20, ]),
#                   aes(x=reorder(category, 
#                                 foldEnrich),
#            y=foldEnrich,
#            fill=-log10(padjust))) +
#   scale_fill_distiller(palette="Reds", direction=1) +
#   geom_bar(stat='identity') + theme_mike() +
#   scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
#   coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")
# 
# plot_grid(p.enrich, p.terms, ncol=2)
```


#### Down-regulated processes

```{r, echo=FALSE, fig.height=5.5, fig.width=12.5}
# de.big.vector <- as.integer(unique(part1.all.genes) %in% unique(ltime.down))
# names(de.big.vector) <- unique(part1.all.genes)
#
# de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part1.gene.length))]
#
# pwf.all <- nullp(de.big.vector, "mm10", "ensGene", bias.data=part1.gene.length[intersect(names(de.big.vector), names(part1.gene.length))],
#                  plot.fit=FALSE)
# GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
#                            method="Wallenius", use_genes_without_cat=T)
#
# # GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
# #                       method="Wallenius", use_genes_without_cat=T)
#
# GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
# GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(ltime.down)))/(GO.wall.all$numInCat/length(part1.gene.length)))
#
# # select on p-value and foldEnrich
# GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
# go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)
#
# p.enrich <- ggplot(GO.sig.all,
#                    aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
#                        size=foldEnrich)) +
#   geom_point() + theme_mike() +
#   scale_colour_gradient(low="#000000", high="#DC143C")
#
# GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]
#
# p.terms <- ggplot(na.omit(GO.all[1:20, ]),
#                   aes(x=reorder(category,
#                                 foldEnrich),
#            y=foldEnrich,
#            fill=-log10(padjust))) +
#   scale_fill_distiller(palette="Reds", direction=1) +
#   geom_bar(stat='identity') + theme_mike() +
#   scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
#   coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")
#
# plot_grid(p.enrich, p.terms, ncol=2)
```

There are quite a lot of enriched terms in each of the up and down-regulated processes along the left branch of the trajectory.  If these are co-ordinated changes then we should be able to cluster the genes together, and find modules of genes that share functional annotations.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
# left.diff.exprs <- part1.nz[unique(c(ltime.down, ltime.up)),
#                               !colnames(part1.nz) %in% multi.cell]
# left.de.cor <- cor(t(left.diff.exprs), method='spearman')
# 
# left.de.dist <- as.dist(abs(left.de.cor - 1))
# left.de.clust <- flashClust(left.de.dist, method="average")
# left.de.cut <- cutreeDynamicTree(left.de.clust, deepSplit=TRUE, minModuleSize=50)
# left.de.cols <- labels2colors(left.de.cut)
# 
# left.annot.df <- data.frame(cbind(left.de.cols))
# rownames(left.annot.df) <- rownames(left.diff.exprs)
# colnames(left.annot.df) <- "Cluster"
# left.gene.clust.cols <- col2hcl(unique(left.de.cols))
# names(left.gene.clust.cols) <- unique(left.de.cols)
# left.annot.cols <- list("Cluster"=left.gene.clust.cols)
# 
# pheatmap(left.de.cor,
#          show_rownames=FALSE, show_colnames=FALSE,
#          annotation_col=left.annot.df, annotation_colors=left.annot.cols,
#          clustering_distance_cols=left.de.dist,
#          clustering_distance_rows=left.de.dist,
#          clustering_method="average",
#          filename="~/Dropbox/Adipose/Adipose-LPseudotime_DEgenes-heatmap.png",
#          height=5.75, width=5.75, res=300)
```

For the most part there are genes that go up and genes that go down.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
# left.annot.df$gene_id <- rownames(left.annot.df)
# 
# write.table(left.annot.df,
#             file="~/Dropbox/Adipose/Adipose_LeftPseudotime_DEgenes-clusters.tsv",
#             sep="\t", quote=FALSE, row.names=FALSE)
# 
# left.blue.genes <- rownames(left.annot.df)[left.annot.df$Cluster == 'blue']
# left.brown.genes <- rownames(left.annot.df)[left.annot.df$Cluster == 'brown']
# left.turquoise.genes <- rownames(left.annot.df)[left.annot.df$Cluster == 'turquoise']
```


#### Functional enrichment test: blue cluster

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.5, fig.width=12.5}
# de.big.vector <- as.integer(unique(part1.all.genes) %in% unique(left.blue.genes))
# names(de.big.vector) <- unique(part1.all.genes)
# 
# de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part1.gene.length))]
# 
# pwf.all <- nullp(de.big.vector, "mm10", "ensGene", bias.data=part1.gene.length[intersect(names(de.big.vector), names(part1.gene.length))],
#                  plot.fit=FALSE)
# GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
#                            method="Wallenius", use_genes_without_cat=T)
# 
# # GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
# #                       method="Wallenius", use_genes_without_cat=T)
# 
# GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
# GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(left.blue.genes)))/(GO.wall.all$numInCat/length(part1.gene.length)))
# 
# # select on p-value and foldEnrich
# GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
# go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)
# 
# p.enrich <- ggplot(GO.sig.all,
#                    aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
#                        size=foldEnrich)) +
#   geom_point() + theme_mike() +
#   scale_colour_gradient(low="#000000", high="#DC143C")
# 
# GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]
# 
# p.terms <- ggplot(na.omit(GO.all[1:20, ]),
#                   aes(x=reorder(category, 
#                                 foldEnrich),
#            y=foldEnrich,
#            fill=-log10(padjust))) +
#   scale_fill_distiller(palette="Reds", direction=1) +
#   geom_bar(stat='identity') + theme_mike() +
#   scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
#   coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")
# 
# write.table(GO.all,
#             file="~/Dropbox/Adipose/Adipose-LeftPseudotime_GOenrich-blueCluster.tsv",
#             sep="\t", row.names=FALSE, quote=FALSE)
# 
# plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.3, 0.7))
```

There are still `r dim(GO.all)[1]` enriched categories within this cluster of genes, many of which are quite generic and uninformative.

#### Functional enrichment test: brown cluster

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.5, fig.width=12.5}
# de.big.vector <- as.integer(unique(part1.all.genes) %in% unique(left.brown.genes))
# names(de.big.vector) <- unique(part1.all.genes)
# 
# de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part1.gene.length))]
# 
# pwf.all <- nullp(de.big.vector, "mm10", "ensGene", bias.data=part1.gene.length[intersect(names(de.big.vector), names(part1.gene.length))],
#                  plot.fit=FALSE)
# GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
#                            method="Wallenius", use_genes_without_cat=T)
# # GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
# #                       method="Wallenius", use_genes_without_cat=T)
# 
# GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
# GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(left.brown.genes)))/(GO.wall.all$numInCat/length(part1.gene.length)))
# 
# # select on p-value and foldEnrich
# GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
# go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)
# 
# p.enrich <- ggplot(GO.sig.all,
#                    aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
#                        size=foldEnrich)) +
#   geom_point() + theme_mike() +
#   scale_colour_gradient(low="#000000", high="#DC143C")
# 
# GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]
# 
# p.terms <- ggplot(na.omit(GO.all[1:20, ]),
#                   aes(x=reorder(category, 
#                                 foldEnrich),
#            y=foldEnrich,
#            fill=-log10(padjust))) +
#   scale_fill_distiller(palette="Reds", direction=1) +
#   geom_bar(stat='identity') + theme_mike() +
#   scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
#   coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")
# 
# write.table(GO.all,
#             file="~/Dropbox/Adipose/Adipose-LeftPseudotime_GOenrich-brownCluster.tsv",
#             sep="\t", row.names=FALSE, quote=FALSE)
# 
# plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.3, 0.7))
```


Why would there be an enrichment for terms related to innate immune response when there are no macrophages in amongst these cells??  Is this an indication of a response to a pro-inflammatory signal?

#### Functional enrichment test: turquoise cluster

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.5, fig.width=12.5}
# de.big.vector <- as.integer(unique(part1.all.genes) %in% unique(left.turquoise.genes))
# names(de.big.vector) <- unique(part1.all.genes)
# 
# de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part1.gene.length))]
# 
# pwf.all <- nullp(de.big.vector, "mm10", "ensGene", bias.data=part1.gene.length[intersect(names(de.big.vector), names(part1.gene.length))],
#                  plot.fit=FALSE)
# GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
#                            method="Wallenius", use_genes_without_cat=T)
# # GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
# #                       method="Wallenius", use_genes_without_cat=T)
# 
# GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
# GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(left.turquoise.genes)))/(GO.wall.all$numInCat/length(part1.gene.length)))
# 
# # select on p-value and foldEnrich
# GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
# go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)
# 
# p.enrich <- ggplot(GO.sig.all,
#                    aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
#                        size=foldEnrich)) +
#   geom_point() + theme_mike() +
#   scale_colour_gradient(low="#000000", high="#DC143C")
# 
# GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]
# 
# p.terms <- ggplot(na.omit(GO.all[1:20, ]),
#                   aes(x=reorder(category, 
#                                 foldEnrich),
#            y=foldEnrich,
#            fill=-log10(padjust))) +
#   scale_fill_distiller(palette="Reds", direction=1) +
#   geom_bar(stat='identity') + theme_mike() +
#   scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
#   coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")
# 
# write.table(GO.all,
#             file="~/Dropbox/Adipose/Adipose-LeftPseudotime_GOenrich-turquoiseCluster.tsv",
#             sep="\t", row.names=FALSE, quote=FALSE)
# 
# plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.3, 0.7))
```

Adhesion, locomotion and differentiation imply a mesenchymal to epithelial transition.

#### Differential Expression Analysis: Pseudotime with clusters 2, & 3

```{r, echo=FALSE}
# rptime <- part1.dc.merge$Sample[part1.dc.merge$Cluster %in% c(2, 3)]
# design.mat.rptime <- model.matrix(~ DPT1 + Plate, data=part1.dc.merge[part1.dc.merge$Sample %in% rptime, ])
# 
# # fit a linear model
# fit.rptime <- lmFit(part1.nz[, colnames(part1.nz) %in% rptime], design.mat.rptime)
# fit.rptime <- eBayes(fit.rptime)
# sum.res.rptime <- summary(decideTests(fit.rptime))
# 
# de.res.rptime <- topTable(fit.rptime, coef=2, n=Inf, sort="p", p=1.0)
# de.res.rptime$Sig <- 0
# de.res.rptime$Sig[de.res.rptime$adj.P.Val <= 0.01] <- 1
# de.res.rptime$Sig <- as.factor(de.res.rptime$Sig)
# 
# de.res.rptime$Diff <- 0
# de.res.rptime$Diff[de.res.rptime$logFC < 0 & de.res.rptime$Sig == 1] <- -1
# de.res.rptime$Diff[de.res.rptime$logFC > 0 & de.res.rptime$Sig == 1] <- 1
# 
# de.res.rptime$gene_id <- rownames(de.res.rptime)
# de.merge.rptime <- merge(de.res.rptime, gene_symbol,
#                   by.x='gene_id', by.y='ensembl_gene_id', all.x=TRUE)
# gene.top.rptime <- de.merge.rptime$external_gene_name
# 
# # plot the top 6 and bottom 6 gene symbols
# top6 <- de.merge.rptime$external_gene_name[order(de.merge.rptime$logFC, decreasing=TRUE)][1:6]
# bottom6 <- de.merge.rptime$external_gene_name[order(de.merge.rptime$logFC, decreasing=FALSE)][1:6]
# gene.top.rptime[!gene.top.rptime %in% c(top6, bottom6)] <- ""
# 
# # gene.top.rptime[de.merge.rptime$logFC > -9 & de.merge.rptime$logFC < 9] <- ""
# # gene.top.rptime[is.na(gene.top.rptime)] <- ""
# 
# write.table(de.merge.rptime,
#             file="~/Dropbox/Adipose/Adipose-DEgenes_Rightptime.tsv",
#             sep="\t", quote=F)
# 
# rtime.ma <- ggplot(de.merge.rptime, aes(x=AveExpr, y=logFC, colour=Sig)) +
#   geom_point(size=1, alpha=0.8) + theme_mike() +
#   scale_colour_manual(values=c("#A3A3A3", "#FF0000")) +
#   labs(x=expression(paste("Mean log"[2], " Expression")),
#        y=expression(paste("log"[2], " Fold Change"))) +
#   guides('colour'=FALSE) +
#   ylim(c(-25, 25)) +
#   geom_label_repel(aes(label=gene.top.rptime),
#                   colour='black',
#                   nudge_x=0.1,
#                   nudge_y=0.1,
#                   size=3)
# 
# ggsave(rtime.ma,
#        filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/MAplot-RightPseudotime.png",
#        height=4.25, width=6.75, dpi=300)
# 
# rtime.ma
```


```{r}
# sum.res.rptime
```


#### Right pseudotime functional enrichment

#### Up-regulated processes

```{r, echo=FALSE, fig.height=5.5, fig.width=12.5}
# rtime.up <- de.merge.rptime$gene_id[de.merge.rptime$Sig == 1 & de.merge.rptime$logFC > 0]
# rtime.down <- de.merge.rptime$gene_id[de.merge.rptime$Sig == 1 & de.merge.rptime$logFC < 0]
# 
# de.big.vector <- as.integer(unique(part1.all.genes) %in% unique(rtime.up))
# names(de.big.vector) <- unique(part1.all.genes)
# 
# de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part1.gene.length))]
# 
# pwf.all <- nullp(de.big.vector, "mm10", "ensGene", bias.data=part1.gene.length[intersect(names(de.big.vector), names(part1.gene.length))],
#                  plot.fit=FALSE)
# GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
#                            method="Wallenius", use_genes_without_cat=T)
# # GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
# #                       method="Wallenius", use_genes_without_cat=T)
# 
# GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
# GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(rtime.up)))/(GO.wall.all$numInCat/length(part1.gene.length)))
# 
# # select on p-value and foldEnrich
# GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
# go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)
# 
# p.enrich <- ggplot(GO.sig.all,
#                    aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
#                        size=foldEnrich)) +
#   geom_point() + theme_mike() +
#   scale_colour_gradient(low="#000000", high="#DC143C")
# 
# GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]
# 
# p.terms <- ggplot(na.omit(GO.all[1:20, ]),
#                   aes(x=reorder(category, 
#                                 foldEnrich),
#            y=foldEnrich,
#            fill=-log10(padjust))) +
#   scale_fill_distiller(palette="Reds", direction=1) +
#   geom_bar(stat='identity') + theme_mike() +
#   scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
#   coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")
# 
# plot_grid(p.enrich, p.terms, ncol=2)
```


#### Down-regulated processes

```{r, echo=FALSE, fig.height=5.5, fig.width=12.5}
# de.big.vector <- as.integer(unique(part1.all.genes) %in% unique(rtime.down))
# names(de.big.vector) <- unique(part1.all.genes)
# 
# de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part1.gene.length))]
# 
# pwf.all <- nullp(de.big.vector, "mm10", "ensGene", bias.data=part1.gene.length[intersect(names(de.big.vector), names(part1.gene.length))],
#                  plot.fit=FALSE)
# GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
#                            method="Wallenius", use_genes_without_cat=T)
# # GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
# #                       method="Wallenius", use_genes_without_cat=T)
# 
# GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
# GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(rtime.down)))/(GO.wall.all$numInCat/length(part1.gene.length)))
# 
# # select on p-value and foldEnrich
# GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
# go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)
# 
# p.enrich <- ggplot(GO.sig.all,
#                    aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
#                        size=foldEnrich)) +
#   geom_point() + theme_mike() +
#   scale_colour_gradient(low="#000000", high="#DC143C")
# 
# GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]
# 
# p.terms <- ggplot(na.omit(GO.all[1:20, ]),
#                   aes(x=reorder(category, 
#                                 foldEnrich),
#            y=foldEnrich,
#            fill=-log10(padjust))) +
#   scale_fill_distiller(palette="Reds", direction=1) +
#   geom_bar(stat='identity') + theme_mike() +
#   scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
#   coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")
# 
# plot_grid(p.enrich, p.terms, ncol=2)
```

```{r, echo=FALSE, warning=FALSE, message=FALSE}
# right.diff.exprs <- part1.nz[unique(c(rtime.down, rtime.up)),
#                               !colnames(part1.nz) %in% multi.cell]
# right.de.cor <- cor(t(right.diff.exprs), method='spearman')
# 
# right.de.dist <- as.dist(abs(right.de.cor - 1))
# right.de.clust <- flashClust(right.de.dist, method="average")
# right.de.cut <- cutreeDynamicTree(right.de.clust, deepSplit=TRUE, minModuleSize=50)
# right.de.cols <- labels2colors(right.de.cut)
# 
# right.annot.df <- data.frame(cbind(right.de.cols))
# rownames(right.annot.df) <- rownames(right.diff.exprs)
# colnames(right.annot.df) <- "Cluster"
# right.gene.clust.cols <- col2hcl(unique(right.de.cols))
# names(right.gene.clust.cols) <- unique(right.de.cols)
# right.annot.cols <- list("Cluster"=right.gene.clust.cols)
# 
# pheatmap(right.de.cor,
#          show_rownames=FALSE, show_colnames=FALSE,
#          annotation_col=right.annot.df, annotation_colors=right.annot.cols,
#          clustering_distance_cols=right.de.dist,
#          clustering_distance_rows=right.de.dist,
#          clustering_method="average",
#          filename="~/Dropbox/Adipose/Adipose-RPseudotime_DEgenes-heatmap.png",
#          height=5.75, width=5.75, res=300)
```

For the most part there are genes that go up and genes that go down.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
# right.annot.df$gene_id <- rownames(right.annot.df)
# 
# write.table(right.annot.df,
#             file="~/Dropbox/Adipose/Adipose_RightPseudotime_DEgenes-clusters.tsv",
#             sep="\t", quote=FALSE, row.names=FALSE)
# 
# right.blue.genes <- rownames(right.annot.df)[right.annot.df$Cluster == 'blue']
# right.brown.genes <- rownames(right.annot.df)[right.annot.df$Cluster == 'brown']
# right.turquoise.genes <- rownames(right.annot.df)[right.annot.df$Cluster == 'turquoise']
# right.yellow.genes <- rownames(right.annot.df)[right.annot.df$Cluster == 'yellow']
```

#### Functional enrichment test: blue cluster

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.5, fig.width=12.5}
# de.big.vector <- as.integer(unique(part1.all.genes) %in% unique(right.blue.genes))
# names(de.big.vector) <- unique(part1.all.genes)
# 
# de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part1.gene.length))]
# 
# pwf.all <- nullp(de.big.vector, "mm10", "ensGene", bias.data=part1.gene.length[intersect(names(de.big.vector), names(part1.gene.length))],
#                  plot.fit=FALSE)
# GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
#                            method="Wallenius", use_genes_without_cat=T)
# # GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
# #                       method="Wallenius", use_genes_without_cat=T)
# 
# GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
# GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(right.blue.genes)))/(GO.wall.all$numInCat/length(part1.gene.length)))
# 
# # select on p-value and foldEnrich
# GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
# go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)
# 
# p.enrich <- ggplot(GO.sig.all,
#                    aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
#                        size=foldEnrich)) +
#   geom_point() + theme_mike() +
#   scale_colour_gradient(low="#000000", high="#DC143C")
# 
# GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]
# 
# p.terms <- ggplot(na.omit(GO.all[1:20, ]),
#                   aes(x=reorder(category, 
#                                 foldEnrich),
#            y=foldEnrich,
#            fill=-log10(padjust))) +
#   scale_fill_distiller(palette="Reds", direction=1) +
#   geom_bar(stat='identity') + theme_mike() +
#   scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
#   coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")
# 
# write.table(GO.all,
#             file="~/Dropbox/Adipose/Adipose-RightPseudotime_GOenrich-blueCluster.tsv",
#             sep="\t", row.names=FALSE, quote=FALSE)
# 
# ggsave(p.terms,
#        filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/GOenrich_example.png",
#        height=5.5, width=5.75, dpi=300)
# plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.3, 0.7))
```

These are fairly generic terms, mostly they seem to relate to protein synthesis and translation.


#### Functional enrichment test: brown cluster

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.5, fig.width=12.5}
# de.big.vector <- as.integer(unique(part1.all.genes) %in% unique(right.brown.genes))
# names(de.big.vector) <- unique(part1.all.genes)
# 
# de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part1.gene.length))]
# 
# pwf.all <- nullp(de.big.vector, "mm10", "ensGene", bias.data=part1.gene.length[intersect(names(de.big.vector), names(part1.gene.length))],
#                  plot.fit=FALSE)
# GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
#                            method="Wallenius", use_genes_without_cat=T)
# # GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
# #                       method="Wallenius", use_genes_without_cat=T)
# 
# GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
# GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(right.brown.genes)))/(GO.wall.all$numInCat/length(part1.gene.length)))
# 
# # select on p-value and foldEnrich
# GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
# go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)
# 
# p.enrich <- ggplot(GO.sig.all,
#                    aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
#                        size=foldEnrich)) +
#   geom_point() + theme_mike() +
#   scale_colour_gradient(low="#000000", high="#DC143C")
# 
# GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]
# 
# p.terms <- ggplot(na.omit(GO.all[1:20, ]),
#                   aes(x=reorder(category, 
#                                 foldEnrich),
#            y=foldEnrich,
#            fill=-log10(padjust))) +
#   scale_fill_distiller(palette="Reds", direction=1) +
#   geom_bar(stat='identity') + theme_mike() +
#   scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
#   coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")
# 
# write.table(GO.all,
#             file="~/Dropbox/Adipose/Adipose-RightPseudotime_GOenrich-brownCluster.tsv",
#             sep="\t", row.names=FALSE, quote=FALSE)
# 
# plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.3, 0.7))
```

The ERK1/2 cascade is quite sontrly enriched here, possibly linked to the overall cytokine signalling seen in other gene modules.

#### Functional enrichment test: turquoise cluster

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.5, fig.width=12.5}
# de.big.vector <- as.integer(unique(part1.all.genes) %in% unique(right.turquoise.genes))
# names(de.big.vector) <- unique(part1.all.genes)
# 
# de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part1.gene.length))]
# 
# pwf.all <- nullp(de.big.vector, "mm10", "ensGene", bias.data=part1.gene.length[intersect(names(de.big.vector), names(part1.gene.length))],
#                  plot.fit=FALSE)
# GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
#                            method="Wallenius", use_genes_without_cat=T)
# # GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
# #                       method="Wallenius", use_genes_without_cat=T)
# 
# GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
# GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(right.turquoise.genes)))/(GO.wall.all$numInCat/length(part1.gene.length)))
# 
# # select on p-value and foldEnrich
# GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
# go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)
# 
# p.enrich <- ggplot(GO.sig.all,
#                    aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
#                        size=foldEnrich)) +
#   geom_point() + theme_mike() +
#   scale_colour_gradient(low="#000000", high="#DC143C")
# 
# GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]
# 
# p.terms <- ggplot(na.omit(GO.all[1:20, ]),
#                   aes(x=reorder(category, 
#                                 foldEnrich),
#            y=foldEnrich,
#            fill=-log10(padjust))) +
#   scale_fill_distiller(palette="Reds", direction=1) +
#   geom_bar(stat='identity') + theme_mike() +
#   scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
#   coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")
# 
# write.table(GO.all,
#             file="~/Dropbox/Adipose/Adipose-RightPseudotime_GOenrich-turquoiseCluster.tsv",
#             sep="\t", row.names=FALSE, quote=FALSE)
# 
# plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.3, 0.7))
```

Lot's of motility and migration enrichment.  A lot of these gene cluster enrichments are quite similar between the left and right branches of the pseudotime.  Is this just because of the shared reference population of cells?

#### Functional enrichment test: yellow cluster

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.5, fig.width=12.5}
# de.big.vector <- as.integer(unique(part1.all.genes) %in% unique(right.yellow.genes))
# names(de.big.vector) <- unique(part1.all.genes)
# 
# de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part1.gene.length))]
# 
# pwf.all <- nullp(de.big.vector, "mm10", "ensGene", bias.data=part1.gene.length[intersect(names(de.big.vector), names(part1.gene.length))],
#                  plot.fit=FALSE)
# GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
#                            method="Wallenius", use_genes_without_cat=T)
# # GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
# #                       method="Wallenius", use_genes_without_cat=T)
# 
# GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
# GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(right.yellow.genes)))/(GO.wall.all$numInCat/length(part1.gene.length)))
# 
# # select on p-value and foldEnrich
# GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
# go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)
# 
# p.enrich <- ggplot(GO.sig.all,
#                    aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
#                        size=foldEnrich)) +
#   geom_point() + theme_mike() +
#   scale_colour_gradient(low="#000000", high="#DC143C")
# 
# GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]
# 
# p.terms <- ggplot(na.omit(GO.all[1:20, ]),
#                   aes(x=reorder(category, 
#                                 foldEnrich),
#            y=foldEnrich,
#            fill=-log10(padjust))) +
#   scale_fill_distiller(palette="Reds", direction=1) +
#   geom_bar(stat='identity') + theme_mike() +
#   scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
#   coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")
# 
# write.table(GO.all,
#             file="~/Dropbox/Adipose/Adipose-RightPseudotime_GOenrich-yellowCluster.tsv",
#             sep="\t", row.names=FALSE, quote=FALSE)
# 
# plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.3, 0.7))
```

Interesting, that's a really specific set of enrichment related to retinoid/diterpenoid metabolism.  That is a very strange and specific 
enrichment.  I wonder if it is driven by a single, or small number, of associations in the literature.

The locomotion and migration are a _really_ strong component of these gene expression changes.

There might also be a central trajectory involving cells from cluster 1, let's see what differences exist.

Most of the differences between the left and right branches relate to morphological changes and response to some external stimulus.

```{r, echo=FALSE, fig.height=6.5, fig.width=6.5}
# # heatmap of partition 1 pseudotime-dependent DE genes
# 
# part1.diff.genes <- unique(c(de.merge.rptime$gene_id[de.merge.rptime$Sig == 1 & (abs(de.merge.rptime$logFC) > 3)],
#                              de.merge.lptime$gene_id[de.merge.lptime$Sig == 1 & (abs(de.merge.lptime$logFC) > 3)]))
# 
# part1.ltime.up <- unique(c(de.merge.lptime$gene_id[de.merge.lptime$Sig == 1 & (de.merge.lptime$logFC > 3)]))
# part1.ltime.down <- unique(c(de.merge.lptime$gene_id[de.merge.lptime$Sig == 1 & (de.merge.lptime$logFC < -3)]))
# 
# part1.rtime.up <- unique(c(de.merge.rptime$gene_id[de.merge.rptime$Sig == 1 & (de.merge.rptime$logFC > 3)]))
# part1.rtime.down <- unique(c(de.merge.rptime$gene_id[de.merge.rptime$Sig == 1 & (de.merge.rptime$logFC < -3)]))
# 
# part1.ltime.genes <- setdiff(unique(part1.ltime.up), unique(c(part1.rtime.up, part1.ltime.down, part1.rtime.down)))
# part1.rtime.genes <- setdiff(unique(part1.rtime.up), unique(c(part1.ltime.up, part1.ltime.down, part1.rtime.down)))
# part1.early.genes <- setdiff(unique(c(part1.ltime.down, part1.rtime.down)), unique(c(part1.ltime.up, part1.rtime.up)))
# 
# # only plot the genes unique to each cluster
# part1.diff.exprs <- part1.nz[part1.diff.genes, part1.dc.merge$Sample[order(part1.dc.merge$DPT1, decreasing=FALSE)]]
# 
# part1.genes.df <- data.frame(cbind(part1.diff.genes))
# colnames(part1.genes.df) <- c("gene_id")
# rownames(part1.genes.df) <- part1.genes.df[, 1]
# part1.genes.df$GeneGroup <- "Late"
# part1.genes.df[part1.genes.df$gene_id %in% part1.ltime.genes, ]$GeneGroup <- "LeftBranch"
# part1.genes.df[part1.genes.df$gene_id %in% part1.rtime.genes, ]$GeneGroup <- "RightBranch"
# part1.genes.df[part1.genes.df$gene_id %in% part1.early.genes, ]$GeneGroup <- "Early"
# 
# part1.genes.df$GeneGroup <- as.factor(part1.genes.df$GeneGroup)
# 
# #genes.cols <- 
# genes.cols <- c("#F700FF", "#00FF46", "#00F7FF", "#FF9E00")
# names(genes.cols) <- c("LeftBranch", "RightBranch", "Early", "Late")
# part1.genes.df <- as.data.frame(part1.genes.df[, -1])
# rownames(part1.genes.df) <- part1.diff.genes
# colnames(part1.genes.df) <- "GeneGroup"
# 
# part1.annot.df <- part1.dc.merge[, c("ClusterLabel", "DPT1")]
# part1.annot.df$ClusterLabel <- as.factor(part1.annot.df$ClusterLabel)
# part1.annot.df$DPT1 <- as.numeric(part1.annot.df$DPT1)
# rownames(part1.annot.df) <- part1.dc.merge$Sample
# 
# dc.cols <- colorRampPalette(c("#FAFF60", "#8615B0"))(dim(part1.diff.exprs)[2])
# names(dc.cols) <- part1.annot.df$DPT1
#   
# marker.cols <- gfp.cols
# names(marker.cols) <- as.factor(c("GFP+", "GFP-"))
# 
# cluster.cols <- label.map
# 
# part1.annot.cols <- list(ClusterLabel=cluster.cols,
#                          GeneGroup=genes.cols,
#                          DPT1=dc.cols)
# 
# gene.order <- order(part1.genes.df$GeneGroup)
# 
# pheatmap(part1.diff.exprs[gene.order, ], show_rownames=FALSE, show_colnames=FALSE,
#          cluster_rows=FALSE, annotation_col=part1.annot.df,
#          annotation_row=part1.genes.df,
#          annotation_colors=part1.annot.cols,
#          cluster_cols=FALSE,
#          filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Pseudotime_DE_genes-Kmeans1-heatmap.png",
#          height=6.75, width=6.75, res=300)
```


## Partition 2 analysis

```{r}
part2.ptime.clust <- ggplot(part2.dc.merge, aes(x=Cluster, y=DPT1, fill=ClusterLabel)) +
  geom_boxplot() +
  geom_jitter(shape=24) +
  theme_bw() +
  scale_fill_manual(values=label.map) +
  labs(x="Cluster", y="Diffusion Pseudotime") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=17, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14), 
        panel.grid.major=element_blank(),
        panel.border=element_blank(),
        axis.line=element_line(colour="black")) +
  guides(fill=FALSE)

ggsave(part2.ptime.clust,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Pseudotime-Kmeans2_boxplot-clusters.png",
       height=6.5, width=3.25, dpi=300)

part2.ptime.clust
```


```{r}
part2.cd34.clust <- ggplot(part2.dc.merge, aes(y=Cd34, x=Cluster, fill=ClusterLabel)) +
  geom_boxplot() +
  geom_jitter(shape=21) +
  theme_bw() +
  scale_fill_manual(values=label.map) +
  labs(y=expression(bold(paste("Cd34 log"[2], " Expression"))), x="Cluster") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=17, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14), 
        panel.grid.major=element_blank(),
        panel.border=element_blank(),
        axis.line=element_line(colour="black")) +
  guides(fill=FALSE) +
  lims(y=c(-1, 16))

ggsave(part2.cd34.clust,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Cd34_cluster-Kmeans2-boxplot.png",
       height=3.25, width=3.25, dpi=300)

part2.cd34.clust
```

```{r, echo=FALSE}
part2.wt1.clust <- ggplot(part2.dc.merge, aes(y=Wt1, x=Cluster, fill=ClusterLabel)) +
  geom_boxplot() +
  geom_jitter(shape=21) +
  theme_bw() +
  scale_fill_manual(values=label.map) +
  labs(y=expression(bold(paste("Wt1 log"[2], " Expression"))), x="Cluster") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=17, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14), 
        panel.grid.major=element_blank(),
        panel.border=element_blank(),
        axis.line=element_line(colour="black")) +
  guides(fill=FALSE) +
  lims(y=c(-1, 16))

ggsave(part2.wt1.clust,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Wt1_cluster-Kmeans2-boxplot.png",
       height=3.25, width=3.25, dpi=300)

part2.wt1.clust
```

```{r, echo=FALSE}
part2.pparg.clust <- ggplot(part2.dc.merge, aes(y=Pparg, x=Cluster, fill=ClusterLabel)) +
  geom_boxplot() +
  geom_jitter(shape=21) +
  theme_bw() +
  scale_fill_manual(values=label.map) +
  labs(y=expression(bold(paste("Pparg log"[2], " Expression"))), x="Cluster") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=17, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14), 
        panel.grid.major=element_blank(),
        panel.border=element_blank(),
        axis.line=element_line(colour="black")) +
  guides(fill=FALSE)  +
  lims(y=c(-1, 16))

ggsave(part2.pparg.clust,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Pparg_cluster-Kmeans2-boxplot.png",
       height=3.25, width=3.25, dpi=300)

part2.pparg.clust
```


The diffusion map creation is quite sensitive to the magnitude of the number of nearest neighbours, `k`.  Using a small value for `k`, we can see a branch-like trajectory of cells moving from right to left (or _vice versa_).  Let's take a look to see what genes are expressed along this pseudotime trajectory, starting with _Wt1_ and _Tcf21_.  Perhaps we can also split the data into GFP+ and GFP- cells and see whether these explain the branches?

It doesn't look like cells in particular phases of the cell cycle cluster on particular branches, perhaps there is a slight skew towards the left hand end, but it is not clear cut.  I suspect this indicates that the cell cycle phase classification doesn't work so well on these cells.  It would be better to get that information from specific proliferation assays on adipose precursors.

### Is the adipose precursor pseudotime indicative of a hierarchy of cells?

These adipose precursors (AP) cells were extracted from the stromal vascular fraction (SVF) of white adipose tissue (WAT) epididymal depots as Lin-CD31-IgM-CD34+GFP+/-.  Matthew S Roderheffer's lab have published on APs, showing that CD34+CD29+Sca1+CD24- are derived from a CD24+ population of cells.  We do not observe any expression of Cd24 in these cells, indicating they are probably a slightly later populations of APs.  The question is, how are the sanity/marker genes expressed along the pseudotime?

```{r, echo=FALSE, fig.height=7.5, fig.width=9.5}
part2.cd29.dc <- ggplot(part2.dc.merge, aes(x=DPT1, y=Itgb1, colour=ClusterLabel)) +
  geom_point() + theme_mike() +
  scale_colour_manual(values=label.map) +
  geom_smooth(method="loess", span=0.5, colour="black")

part2.sca1.dc <- ggplot(part2.dc.merge, aes(x=DPT1, y=Ly6a, colour=ClusterLabel)) +
  geom_point() + theme_mike() +
  scale_colour_manual(values=label.map) +
  geom_smooth(method="loess", span=0.5, colour="black")

part2.wt1.dc <- ggplot(part2.dc.merge, aes(x=DPT1, y=Wt1, colour=ClusterLabel)) +
  geom_point() + theme_mike() +
  scale_colour_manual(values=label.map) +
  geom_smooth(method="loess", span=0.5, colour="black")

part2.tcf21.dc <- ggplot(part2.dc.merge, aes(x=DPT1, y=Tcf21, colour=ClusterLabel)) +
  geom_point() + theme_mike() +
  scale_colour_manual(values=label.map) +
  geom_smooth(method="loess", span=0.5, colour="black")

part2.pparg.dc <- ggplot(part2.dc.merge, aes(x=DPT1, y=Pparg, colour=ClusterLabel)) +
  geom_point() + theme_mike() +
  scale_colour_manual(values=label.map) +
  geom_smooth(method="loess", span=0.5, colour="black")

part2.cebpa.dc <- ggplot(part2.dc.merge, aes(x=DPT1, y=Cebpa, colour=ClusterLabel)) +
  geom_point() + theme_mike() +
  scale_colour_manual(values=label.map) +
  geom_smooth(method="loess", span=0.5, colour="black")


plot_grid(part2.cd29.dc, part2.sca1.dc, part2.wt1.dc, part2.tcf21.dc, part2.pparg.dc, part2.cebpa.dc,
          labels=c("Cd29", "Sca1", "Wt1", "Tcf21", "Pparg", "Cebpa"),
          ncol=3)
```


It looks like the expression of Cd29 (Itgb1), Sca1 (Ly6a) and Wt1 are across pretty much the entire pseudotime, without any obvious restriction to any aspect of it.  Pparg on the other hand is restricted to the left end of the trajectory, mostly in the brown, pink and turquoise clusters of cells.  It looks like Tcf21 expression is declining gradually along the pseudotime.  Cebpa is likewise restricted to the left end of the pseudotime, but with slightly more dispersed expression that includes some of the green, red and blue cells.  These expression patterns for Pparg and Cebpa suggest the left branch are the more mature adipocyte precursors, as these are markers of mature adipocytes.

There doesn't seem to be much gained from trying to split this non-bifurcating trajectory based on marker sorting.

```{r}
part2.dpt.pca <- prcomp(part2.dc.merge[, c("DC1", "DC2")])
part2.dpt.pcs <- do.call(cbind.data.frame, list("PC.DC1"=part2.dpt.pca$x[, 1],
                                                "PC.DC2"=part2.dpt.pca$x[, 2]))
part2.dpt.pcs$Sample <- part2.dc.merge$Sample
part2.dpt.pcs.merge <- merge(part2.dpt.pcs, part2.dc.merge, by='Sample')
ggplot(part2.dpt.pcs.merge,
       aes(x=PC.DC1, y=PC.DC2, fill=ClusterLabel)) +
  geom_point(shape=21, size=3) + theme_minimal() +
  scale_fill_manual(values=label.map)
```

This just reconstructs the observed diffusion map, so no surprises.  Doing an eigendecomposition on a matrix of eigenvectors, I guess will just return 
those eigen vectors that form the columns of the eignematrix.

After that little segue, let's look for genes that are correlated with the different branches of cells along the diffusion pseudotime.  I'll group cells into "meta-stable", states based on areas of high cellular density, a la Haghverdi _et al_.  Conveniently these also correspond roughly to the cellular clusters defined previously on the top 55 principal components (who'd have thought, huh?).  I'll test for differentially expressed genes between the centre group, and the two extreme ends of the diffusion pseudotime trajectories.  Perform a partitioning of cells on the first diffusion component should generate ~4 clusters looking at the density plots.

#### Differential Expression Analysis: 1 vs. 2

```{r, echo=FALSE}
part2.dc.merge <- part2.dc.merge[!part2.dc.merge$Sample %in% multi.cell, ]
part2.design.mat.1v2 <- model.matrix(~ 0 + ClusterLabel + Plate, data=part2.dc.merge)

# fit a linear model
part2.nz <- adipose.nz[, colnames(adipose.nz) %in% part2.dc.merge$Sample]
part2.fit <- lmFit(part2.nz[, !colnames(part2.nz) %in% multi.cell], part2.design.mat.1v2)

part2.contrast.mat <-  makeContrasts(ClusterLabelIntermediate-ClusterLabelNaive, 
                                     ClusterLabelMature-ClusterLabelNaive, 
                                     ClusterLabelMature-ClusterLabelIntermediate,
                                     levels=part2.design.mat.1v2)

part2.fit.1v2 <- contrasts.fit(part2.fit, contrasts=part2.contrast.mat)
part2.fit.1v2 <- eBayes(part2.fit.1v2)
part2.sum.res.1v2 <- summary(decideTests(part2.fit.1v2))

part2.de.res.1v2 <- topTable(part2.fit.1v2, coef=1, n=Inf, sort="p", p=1.0)
part2.de.res.1v2$Sig <- 0
part2.de.res.1v2$Sig[part2.de.res.1v2$adj.P.Val <= 0.01] <- 1
part2.de.res.1v2$Sig <- as.factor(part2.de.res.1v2$Sig)

part2.de.res.1v2$Diff <- 0
part2.de.res.1v2$Diff[part2.de.res.1v2$logFC < 0 & part2.de.res.1v2$Sig == 1] <- -1
part2.de.res.1v2$Diff[part2.de.res.1v2$logFC > 0 & part2.de.res.1v2$Sig == 1] <- 1

part2.de.res.1v2$gene_id <- rownames(part2.de.res.1v2)
part2.de.merge.1v2 <- merge(part2.de.res.1v2, gene_symbol,
                            by.x='gene_id', by.y='ensembl_gene_id', all.x=TRUE)

part2.gene.top.1v2 <- part2.de.merge.1v2$external_gene_name
part2.gene.top.1v2[part2.de.merge.1v2$logFC > -10 & part2.de.merge.1v2$logFC < 10] <- ""
part2.gene.top.1v2[is.na(part2.gene.top.1v2)] <- ""

write.table(part2.de.merge.1v2,
            file="~/Dropbox/Adipose/Adipose-DEgenes_1v2-Kmeans2.tsv",
            sep="\t", quote=F)

ggplot(part2.de.merge.1v2, aes(x=AveExpr, y=logFC, colour=Sig)) +
  geom_point(size=1, alpha=0.8) + theme_mike() +
  scale_colour_Publication() +
  labs(x=expression(paste("Mean log"[2], " Expression")),
       y=expression(paste("log"[2], " Fold Change"))) +
  guides('colour'=guide_legend(title="Statistically significant")) +
  ylim(c(-10, 10)) +
  geom_text_repel(aes(label=part2.gene.top.1v2),
                  colour='black',
                  nudge_x=0.15,
                  nudge_y=0.5)
```

The interpretation of the log$_2$ fold changes is that positive values represent higher expression towards the right hand side of the first diffusions component, i.e. up-regulated in k-2 relative to k-1.  Therefore, negative log fold changes represent higher down-regulation of genes moving along to the right of the first diffusion component.  For example, as we move to the right of the pseudotime, we have increased expression of _Dpp4_, _Efhd1_, _Cd55_, etc, and down-regulation of _Inmt_, _Steap4_, _C7_, etc.

#### Differential Expression Analysis: 1 vs. 3

```{r, echo=FALSE}
part2.de.res.3v1 <- topTable(part2.fit.1v2, coef=2, n=Inf, sort="p", p=1.0)
part2.de.res.3v1$Sig <- 0
part2.de.res.3v1$Sig[part2.de.res.3v1$adj.P.Val <= 0.01] <- 1
part2.de.res.3v1$Sig <- as.factor(part2.de.res.3v1$Sig)

part2.de.res.3v1$Diff <- 0
part2.de.res.3v1$Diff[part2.de.res.3v1$logFC < 0 & part2.de.res.3v1$Sig == 1] <- -1
part2.de.res.3v1$Diff[part2.de.res.3v1$logFC > 0 & part2.de.res.3v1$Sig == 1] <- 1

part2.de.res.3v1$gene_id <- rownames(part2.de.res.3v1)
part2.de.merge.3v1 <- merge(part2.de.res.3v1, gene_symbol,
                  by.x='gene_id', by.y='ensembl_gene_id', all.x=TRUE)
part2.gene.top.3v1 <- part2.de.merge.3v1$external_gene_name
part2.gene.top.3v1[part2.de.merge.3v1$logFC > -5 & part2.de.merge.3v1$logFC < 4] <- ""
part2.gene.top.3v1[is.na(part2.gene.top.3v1)] <- ""

write.table(part2.de.merge.3v1,
            file="~/Dropbox/Adipose/Adipose-DEgenes_1v3-Kmeans2.tsv",
            sep="\t", quote=F)

ggplot(part2.de.merge.3v1, aes(x=AveExpr, y=logFC, colour=Sig)) +
  geom_point(size=1, alpha=0.8) + theme_mike() +
  scale_colour_Publication() +
  labs(x=expression(paste("Mean log"[2], " Expression")),
       y=expression(paste("log"[2], " Fold Change"))) +
  guides('colour'=guide_legend(title="Statistically significant")) +
  ylim(c(-10, 10)) +
  geom_text_repel(aes(label=part2.gene.top.3v1),
                  colour='black',
                  nudge_x=0.15,
                  nudge_y=0.5)
```

Almost no differences between clusters 1 and 3 _within_ this set of cells.

#### Differential Expression Analysis: 2 vs. 3

```{r, echo=FALSE}
part2.de.res.3v2 <- topTable(part2.fit.1v2, coef=3, n=Inf, sort="p", p=1.0)
part2.de.res.3v2$Sig <- 0
part2.de.res.3v2$Sig[part2.de.res.3v2$adj.P.Val <= 0.01] <- 1
part2.de.res.3v2$Sig <- as.factor(part2.de.res.3v2$Sig)

part2.de.res.3v2$Diff <- 0
part2.de.res.3v2$Diff[part2.de.res.3v2$logFC < 0 & part2.de.res.3v2$Sig == 1] <- -1
part2.de.res.3v2$Diff[part2.de.res.3v2$logFC > 0 & part2.de.res.3v2$Sig == 1] <- 1

part2.de.res.3v2$gene_id <- rownames(part2.de.res.3v2)
part2.de.merge.3v2 <- merge(part2.de.res.3v2, gene_symbol,
                  by.x='gene_id', by.y='ensembl_gene_id', all.x=TRUE)
part2.gene.top.3v2 <- part2.de.merge.3v2$external_gene_name
part2.gene.top.3v2[part2.de.merge.3v2$logFC > -7 & part2.de.merge.3v2$logFC < 7] <- ""
part2.gene.top.3v2[is.na(part2.gene.top.3v2)] <- ""

write.table(part2.de.merge.3v2,
            file="~/Dropbox/Adipose/Adipose-DEgenes_2v3-Kmeans2.tsv",
            sep="\t", quote=F)

ggplot(part2.de.merge.3v2, aes(x=AveExpr, y=logFC, colour=Sig)) +
  geom_point(size=1, alpha=0.8) + theme_mike() +
  scale_colour_Publication() +
  labs(x=expression(paste("Mean log"[2], " Expression")),
       y=expression(paste("log"[2], " Fold Change"))) +
  guides('colour'=guide_legend(title="Statistically significant")) +
  ylim(c(-15, 15)) +
  geom_text_repel(aes(label=part2.gene.top.3v2),
                  colour='black',
                  nudge_x=0.15,
                  nudge_y=0.5)
```

```{r}
knitr::kable(part2.sum.res.1v2)
```

Let's take a look at how these genes are expressed across cells, ordered along the first diffusion component.

```{r, echo=FALSE, fig.height=6.5, fig.width=6.5}
part2.diff.genes <- unique(c(part2.de.merge.1v2$gene_id[part2.de.merge.1v2$Sig == 1 & (abs(part2.de.merge.1v2$logFC) > 3)],
                       part2.de.merge.3v1$gene_id[part2.de.merge.3v1$Sig == 1 & (abs(part2.de.merge.3v1$logFC) > 3)],
                       part2.de.merge.3v2$gene_id[part2.de.merge.3v2$Sig == 1 & (abs(part2.de.merge.3v2$logFC) > 3)]))

part2.clust1.up <- unique(c(part2.de.merge.1v2$gene_id[part2.de.merge.1v2$Sig == 1 & (part2.de.merge.1v2$logFC < -3)],
                      part2.de.merge.3v1$gene_id[part2.de.merge.3v1$Sig == 1 & (part2.de.merge.3v1$logFC < -3)]))

part2.clust2.up <- unique(c(part2.de.merge.1v2$gene_id[part2.de.merge.1v2$Sig == 1 & (part2.de.merge.1v2$logFC > 3)],
                      part2.de.merge.3v2$gene_id[part2.de.merge.3v2$Sig == 1 & (part2.de.merge.3v2$logFC < -3)]))

part2.clust3.up <- unique(c(part2.de.merge.3v1$gene_id[part2.de.merge.1v2$Sig == 1 & (part2.de.merge.1v2$logFC > 3)],
                      part2.de.merge.3v2$gene_id[part2.de.merge.3v2$Sig == 1 & (part2.de.merge.3v2$logFC > 3)]))

part2.clust1.genes <- setdiff(part2.clust1.up, unique(c(part2.clust2.up, part2.clust3.up)))
part2.clust2.genes <- setdiff(part2.clust2.up, unique(c(part2.clust1.up, part2.clust3.up)))
part2.clust3.genes <- setdiff(part2.clust3.up, unique(c(part2.clust2.up, part2.clust1.up)))

part2.cluster.genes <- unique(c(part2.clust1.genes, part2.clust2.genes, 
                                part2.clust3.genes))

# only plot the genes unique to each cluster
part2.diff.exprs <- part2.nz[part2.cluster.genes, part2.dc.merge$Sample[order(part2.dc.merge$DPT1, decreasing=TRUE)]]

part2.genes.df <- data.frame(cbind(part2.cluster.genes))
colnames(part2.genes.df) <- c("gene_id")
rownames(part2.genes.df) <- part2.genes.df[, 1]
part2.genes.df$GeneGroup <- "Multiple"
part2.genes.df[part2.genes.df$gene_id %in% part2.clust1.genes, ]$GeneGroup <- "1"
part2.genes.df[part2.genes.df$gene_id %in% part2.clust2.genes, ]$GeneGroup <- "2"
part2.genes.df[part2.genes.df$gene_id %in% part2.clust3.genes, ]$GeneGroup <- "3"

part2.genes.df$GeneGroup <- as.factor(part2.genes.df$GeneGroup)

genes.cols <- col.map
#genes.cols <- c("#386cb0", "#fdb462", "#7fc97f","#ef3b2c")
#names(genes.cols) <- c("1", "Cluster2", "Cluster3", "Cluster4")
part2.genes.df <- as.data.frame(part2.genes.df[, -1])
rownames(part2.genes.df) <- part2.cluster.genes
colnames(part2.genes.df) <- "GeneGroup"


part2.annot.df <- part2.dc.merge[, c("Marker", "ClusterLabel", "DPT1")]
part2.annot.df$Marker <- as.factor(part2.annot.df$Marker)
part2.annot.df$ClusterLabel <- as.factor(part2.annot.df$ClusterLabel)
part2.annot.df$DPT1 <- as.numeric(part2.annot.df$DPT1)
rownames(part2.annot.df) <- part2.dc.merge$Sample

dc.cols <- colorRampPalette(c("#FAFF60", "#8615B0"))(dim(part2.diff.exprs)[2])
names(dc.cols) <- part2.annot.df$DPT1
  
marker.cols <- gfp.cols
names(marker.cols) <- as.factor(c("GFP+", "GFP-"))

cluster.cols <- label.map

part2.annot.cols <- list(Marker=marker.cols,
                         ClusterLabel=cluster.cols,
                         GeneGroup=genes.cols,
                         DPT1=dc.cols)

# annot.cols <- list(Marker=marker.cols,
#                    Cluster=cluster.cols,
#                    DC1=dc.cols)
gene.order <- order(part2.genes.df$GeneGroup)

pheatmap(part2.diff.exprs[gene.order, ], show_rownames=FALSE, show_colnames=FALSE,
         cluster_rows=FALSE, annotation_col=part2.annot.df,
         annotation_row=part2.genes.df,
         annotation_colors=part2.annot.cols,
         cluster_cols=FALSE,
         filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Pseudotime_DE_genes-Kmeans2-heatmap.png",
         height=6.75, width=6.75, res=300)
```

There are clearly groups of genes that characterise cluster 2 along the DC quite well, whilst the majority of other genes form a spectrum of expression (which is expected).  We can also more clearly see which end of the pseudotime the different genes are differentially expressed in.

Let's see what biological processes are enriched in these differentially expressed genes.  I'll combine all differentially expressed genes for now, then look a the separate groups later.

Plot the top DE gene in each cluster on the diffusion map/pseudotime co-ordinate space.
```{r}
part2.c1.max <- unique(c(part2.de.merge.1v2$gene_id[part2.de.merge.1v2$Sig == 1 & (part2.de.merge.1v2$logFC == min(part2.de.merge.1v2$logFC))],
                         part2.de.merge.3v1$gene_id[part2.de.merge.3v1$Sig == 1 & (part2.de.merge.3v1$logFC == min(part2.de.merge.3v1$logFC))]))

part2.c2.max <- unique(c(part2.de.merge.1v2$gene_id[part2.de.merge.1v2$Sig == 1 & (part2.de.merge.1v2$logFC == max(part2.de.merge.1v2$logFC))],
                         part2.de.merge.3v2$gene_id[part2.de.merge.3v2$Sig == 1 & (part2.de.merge.3v2$logFC == min(part2.de.merge.3v1$logFC))]))

part2.c3.max <- unique(c(part2.de.merge.3v1$gene_id[part2.de.merge.3v1$Sig == 1 & (part2.de.merge.3v1$logFC == max(part2.de.merge.3v1$logFC))],
                         part2.de.merge.3v2$gene_id[part2.de.merge.3v2$Sig == 1 & (part2.de.merge.3v2$logFC == max(part2.de.merge.3v2$logFC))]))

part2.max.exprs <-as.data.frame(t(part2.nz[unique(c(part2.c1.max, part2.c2.max, part2.c3.max,
                                                    "ENSMUSG00000079105", "ENSMUSG00000024011", "ENSMUSG00000012428")),
                                           part2.dc.merge$Sample[order(part2.dc.merge$DPT1, decreasing=TRUE)]]))
colnames(part2.max.exprs) <- gene_symbol[unique(c(part2.c1.max, part2.c2.max, part2.c3.max, 
                                                  "ENSMUSG00000079105", "ENSMUSG00000024011", "ENSMUSG00000012428")), ]$external_gene_name
part2.max.exprs$Sample <- rownames(part2.max.exprs)

part2.max.merge <- merge(part2.dc.merge, part2.max.exprs, by='Sample')
```


```{r}
# part2.fos.p <- ggplot(part2.max.merge, aes(x=DC1, y=DC2, fill=Fos)) +
#   geom_point(size=2, shape=21) + theme_minimal() +
#   scale_fill_gradient(low="#F4F6AC", high="#FF0000", breaks=c(0, 4, 8, 12, 16)) +
#   labs(x="Diffusion Component 1", y="Diffusion Component 2") +
#   theme(panel.grid=element_blank(),
#         axis.text=element_text(size=14, colour="black"),
#         axis.title=element_text(size=17, colour="black", face="bold"),
#         legend.title=element_text(size=14, colour="black", face="bold"),
#         legend.text=element_text(size=14),
#         legend.key=element_rect(size=0.5),
#         legend.key.size=unit(0.5, "cm"),
#         legend.position="bottom")
# 
# ggsave(part2.fos.p,
#        filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Fos_pseudotime-Kmeans2.png",
#        height=3.95, width=4.75)
# 
# 
# part2.Efhd1.p <- ggplot(part2.max.merge, aes(x=DC1, y=DC2, fill=Efhd1)) +
#   geom_point(size=2, shape=21) + theme_minimal() +
#   scale_fill_gradient(low="#F4F6AC", high="#FF0000", breaks=c(0, 4, 8, 12, 16)) +
#   labs(x="Diffusion Component 1", y="Diffusion Component 2") +
#   theme(panel.grid=element_blank(),
#         axis.text=element_text(size=14, colour="black"),
#         axis.title=element_text(size=17, colour="black", face="bold"),
#         legend.title=element_text(size=14, colour="black", face="bold"),
#         legend.text=element_text(size=14),
#         legend.key=element_rect(size=0.5),
#         legend.key.size=unit(0.5, "cm"),
#         legend.position="bottom")
# 
# ggsave(part2.Efhd1.p,
#        filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Efhd1_pseudotime-Kmeans2.png",
#        height=3.95, width=4.75)
# 
# 
# part2.Inmt.p <- ggplot(part2.max.merge, aes(x=DC1, y=DC2, fill=Inmt)) +
#   geom_point(size=2, shape=21) + theme_minimal() +
#   scale_fill_gradient(low="#F4F6AC", high="#FF0000", breaks=c(0, 4, 8, 12, 16)) +
#   labs(x="Diffusion Component 1", y="Diffusion Component 2") +
#   theme(panel.grid=element_blank(),
#         axis.text=element_text(size=14, colour="black"),
#         axis.title=element_text(size=17, colour="black", face="bold"),
#         legend.title=element_text(size=14, colour="black", face="bold"),
#         legend.text=element_text(size=14),
#         legend.key=element_rect(size=0.5),
#         legend.key.size=unit(0.5, "cm"),
#         legend.position="bottom")
# 
# ggsave(part2.Inmt.p,
#        filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Inmt_pseudotime-Kmeans2.png",
#        height=3.95, width=4.75)
# 
# 
# part2.Has1.p <- ggplot(part2.max.merge, aes(x=DC1, y=DC2, fill=Has1)) +
#   geom_point(size=2, shape=21) + theme_minimal() +
#   scale_fill_gradient(low="#F4F6AC", high="#FF0000", breaks=c(0, 4, 8, 12, 16)) +
#   labs(x="Diffusion Component 1", y="Diffusion Component 2") +
#   theme(panel.grid=element_blank(),
#         axis.text=element_text(size=14, colour="black"),
#         axis.title=element_text(size=17, colour="black", face="bold"),
#         legend.title=element_text(size=14, colour="black", face="bold"),
#         legend.text=element_text(size=14),
#         legend.key=element_rect(size=0.5),
#         legend.key.size=unit(0.5, "cm"),
#         legend.position="bottom")
# 
# ggsave(part2.Has1.p,
#        filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Has1_pseudotime-Kmeans2.png",
#        height=3.95, width=4.75)
# 
# part2.steap4.p <- ggplot(part2.max.merge, aes(x=DC1, y=DC2, fill=Steap4)) +
#   geom_point(size=2, shape=21) + theme_minimal() +
#   scale_fill_gradient(low="#F4F6AC", high="#FF0000", breaks=c(0, 4, 8, 12, 16)) +
#   labs(x="Diffusion Component 1", y="Diffusion Component 2") +
#   theme(panel.grid=element_blank(),
#         axis.text=element_text(size=14, colour="black"),
#         axis.title=element_text(size=17, colour="black", face="bold"),
#         legend.title=element_text(size=14, colour="black", face="bold"),
#         legend.text=element_text(size=14),
#         legend.key=element_rect(size=0.5),
#         legend.key.size=unit(0.5, "cm"),
#         legend.position="bottom")
# 
# ggsave(part2.steap4.p,
#        filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Steap4_pseudotime-Kmeans2.png",
#        height=3.95, width=4.75)
# 
# 
# part2.pi16.p <- ggplot(part2.max.merge, aes(x=DC1, y=DC2, fill=Pi16)) +
#   geom_point(size=2, shape=21) + theme_minimal() +
#   scale_fill_gradient(low="#F4F6AC", high="#FF0000", breaks=c(0, 4, 8, 12, 16)) +
#   labs(x="Diffusion Component 1", y="Diffusion Component 2") +
#   theme(panel.grid=element_blank(),
#         axis.text=element_text(size=14, colour="black"),
#         axis.title=element_text(size=17, colour="black", face="bold"),
#         legend.title=element_text(size=14, colour="black", face="bold"),
#         legend.text=element_text(size=14),
#         legend.key=element_rect(size=0.5),
#         legend.key.size=unit(0.5, "cm"),
#         legend.position="bottom")
# 
# ggsave(part2.pi16.p,
#        filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Pi16_pseudotime-Kmeans2.png",
#        height=3.95, width=4.75)
# 
# 
# part2.C7.p <- ggplot(part2.max.merge, aes(x=DC1, y=DC2, fill=C7)) +
#   geom_point(size=2, shape=21) + theme_minimal() +
#   scale_fill_gradient(low="#F4F6AC", high="#FF0000", breaks=c(0, 4, 8, 12, 16)) +
#   labs(x="Diffusion Component 1", y="Diffusion Component 2") +
#   theme(panel.grid=element_blank(),
#         axis.text=element_text(size=14, colour="black"),
#         axis.title=element_text(size=17, colour="black", face="bold"),
#         legend.title=element_text(size=14, colour="black", face="bold"),
#         legend.text=element_text(size=14),
#         legend.key=element_rect(size=0.5),
#         legend.key.size=unit(0.5, "cm"),
#         legend.position="bottom")
# 
# ggsave(part2.C7.p,
#        filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/C7_pseudotime-Kmeans2.png",
#        height=3.95, width=4.75)
```


We want to test functional enrichments specific to each cluster.


```{r, echo=FALSE}
# use goseq
# use GO enrichment to test for specific pathways and function
# if there are enriched terms, this would at least imply some
# similar functional connection
part2.de.genes.all <- unique(c(part2.de.merge.1v2$gene_id[part2.de.merge.1v2$Sig == 1],
                         part2.de.merge.3v1$gene_id[part2.de.merge.3v1$Sig == 1],
                         part2.de.merge.3v2$gene_id[part2.de.merge.3v2$Sig == 1]))

gene.tpm <- read.table("~/Dropbox/Thymus/mTEC_1_01.quant", sep="\t",
                       h=T, stringsAsFactors=F)
part2.gene.length <- gene.tpm$Length[unique(gene.tpm$Name) %in% unique(union(part2.de.merge.1v2$gene_id, part2.de.merge.3v2$gene_id))]
names(part2.gene.length) <- intersect(unique(gene.tpm$Name) , unique(union(part2.de.merge.1v2$gene_id, part2.de.merge.3v2$gene_id)))

part2.all.gene.vector <- as.integer(unique(union(part2.de.merge.1v2$gene_id, part2.de.merge.3v2$gene_id)) %in% unique(part2.de.genes.all))
names(part2.all.gene.vector) <- unique(union(part2.de.merge.1v2$gene_id, part2.de.merge.3v2$gene_id))

part2.all.gene.vector <- part2.all.gene.vector[names(part2.gene.length)]

# react <- mapGene2Reactome()
pwf.all <- nullp(part2.all.gene.vector, "mm10", "ensGene", bias.data=part2.gene.length, plot.fit=FALSE)
part2.GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
                      method="Wallenius", use_genes_without_cat=T)
# part2.GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
#                            method="Wallenius", use_genes_without_cat=T)

part2.GO.wall.all$padjust <- p.adjust(part2.GO.wall.all$over_represented_pvalue, method="BH")
part2.GO.wall.all$foldEnrich <- ((part2.GO.wall.all$numDEInCat/length(unique(part2.de.genes.all)))/(part2.GO.wall.all$numInCat/length(part2.gene.length)))

# id2name <- as.list(reactomePATHID2NAME)
# GO.wall.all$description <- do.call(rbind, id2name[GO.wall.all$category])

# select on p-value and foldEnrich
part2.GO.sig.all <- part2.GO.wall.all[(part2.GO.wall.all$padjust <= 0.01), ]
part2.go_cats.all <- data.frame(part2.GO.sig.all$category, part2.GO.sig.all$foldEnrich)

write.table(part2.go_cats.all,
            file="~/Dropbox/Adipose/REVIGO/Adipose-All_4REVIGO-Kmeans2.tsv",
            sep="\t", col.names=F, row.names=F, quote=F)

ggplot(part2.GO.sig.all,
       aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
           size=foldEnrich)) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C")
```

```{r, echo=FALSE, fig.width=7.5, fig.height=9.5}
part2.GO.sig.all <- part2.GO.sig.all[order(part2.GO.sig.all$padjust, decreasing=FALSE), ]

ggplot(part2.GO.sig.all,
       aes(x=reorder(category, 
                     foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")
```

A lot of these gene ontology terms refer to cell migration and motion, which suggests these cells are physically moving during their differentiation.  The enrichment of a wound healing process, coupled with inflammatory response, suggests these are particularly important processes in the development of visceral adipocytes.  BMP stimulus and response to BMP stimulus also appears in the enrichment list.  This is a good validation, given that _Bmp4_ has been shown to stimulate differentiation of MSCs to adipocytes.  It will be interesting to see if and how the BMP genes form a gradient along the pseudotime trajectory.  There is also considerably strong enrichment for gene ontology terms relating to Wnt and $\beta$-catenin signalling, which repress adipocyte development.  Let's start to look at how these genes are distributed over the pseudotime trajectory.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=9.5, fig.width=9.7}
# pull out BMP, Wnt and beta-catenin signalling genes
wnt_table <- read.table("~/Dropbox/Genesets/GO_wnt_catenin-mouse.tsv",
                        sep="\t", h=T, stringsAsFactors=F)
bmp_table <- read.table("~/Dropbox/Genesets/GO_BMP-mouse.tsv",
                        sep="\t", h=T, stringsAsFactors=F)

# genes need to be in gene list and differentially expressed
part2.wnt.exprs <- part2.nz[(rownames(part2.nz) %in% wnt_table$ensembl_gene_id) & (rownames(part2.nz) %in% part2.de.genes.all),
                            part2.dc.merge$Sample[order(part2.dc.merge$DPT1, decreasing=TRUE)]]
part2.wnt.exprs$gene_id <- rownames(part2.wnt.exprs)
part2.wnt.merge <-  merge(part2.wnt.exprs, wnt_table, by.x='gene_id', by.y='ensembl_gene_id', all.x=TRUE)
rownames(part2.wnt.merge) <- part2.wnt.merge$external_gene_name
part2.wnt.merge <- part2.wnt.merge[, 2:(dim(part2.wnt.merge)[2]-1)]

part2.bmp.exprs <- part2.nz[(rownames(part2.nz) %in% bmp_table$ensembl_gene_id) & (rownames(part2.nz) %in% part2.de.genes.all),
                        part2.dc.merge$Sample[order(part2.dc.merge$DPT1, decreasing=TRUE)]]
part2.bmp.exprs$gene_id <- rownames(part2.bmp.exprs)
part2.bmp.merge <- merge(part2.bmp.exprs, bmp_table, by.x='gene_id', by.y='ensembl_gene_id', x.all=TRUE)
rownames(part2.bmp.merge) <- part2.bmp.merge$external_gene_name
part2.bmp.merge <- part2.bmp.merge[, 2:(dim(part2.bmp.merge)[2]-1)]

# heatmap?
part2.diff.exprs <- rbind(part2.wnt.merge, part2.bmp.merge)
part2.diff.exprs <- part2.diff.exprs[!duplicated(part2.diff.exprs), ]

part2.genes.df <- data.frame(cbind(rownames(part2.diff.exprs)))
rownames(part2.genes.df) <- part2.genes.df[, 1]
part2.genes.df$Pathway[rownames(part2.genes.df) %in% rownames(part2.bmp.merge)] <- "BMP"
part2.genes.df$Pathway[rownames(part2.genes.df) %in% rownames(part2.wnt.merge)] <- "Wnt"
part2.genes.df$Pathway <- as.factor(part2.genes.df$Pathway)

genes.cols <- c("#6a3d9a", "#ffff99")
names(genes.cols) <- levels(part2.genes.df$Pathway)
part2.genes.df <- as.data.frame(part2.genes.df[, -1])
rownames(part2.genes.df) <- rownames(part2.diff.exprs)
colnames(part2.genes.df) <- "Pathway"


#annot.df <- part2.dc.merge[, c("Cluster", "Kmeans")]
part2.annot.df <- part2.dc.merge[, c("Marker", "ClusterLabel", "DPT1")]
part2.annot.df$ClusterLabel <- as.factor(part2.annot.df$ClusterLabel)
#annot.df$Kmeans <- as.factor(annot.df$Kmeans)
part2.annot.df$DPT1 <- as.numeric(part2.annot.df$DPT1)
rownames(part2.annot.df) <- part2.dc.merge$Sample

dc.cols <- colorRampPalette(brewer.pal(10, "PRGn"))(dim(part2.diff.exprs)[2])
names(dc.cols) <- part2.annot.df$DPT1
  
marker.cols <- c("#ef3b2c","#662506")
names(marker.cols) <- as.factor(c("GFP+", "GFP-"))

cluster.cols <- label.map

annot.cols <- list(ClusterLabel=cluster.cols,
                   DPT=dc.cols,
                   Marker=marker.cols,
                   Pathway=genes.cols)

pheatmap(part2.diff.exprs, show_rownames=TRUE, show_colnames=FALSE,
         cluster_rows=TRUE, annotation_col=part2.annot.df,
         annotation_row=part2.genes.df,
         annotation_colors=annot.cols,
         cluster_cols=FALSE)
```

The cells are ordered according to their position along the first diffusion component.  There is a noticeable switch-like behaviour for some genes, in that they are almost entirely absent, or expression is restricted to, the K-2 group.  These include genes involved in both BMP signalling and response, and Wnt signalling.  We might expect to see less Wnt signalling and more BMP stimulus in the more mature cells, as Wnt repressess adipogenesis in MSCs and BMPs generate promote adipocyte differentiation.  The genes off in K-2 are _Ndrg2_, _Ror1_, _Rspo3_, _Fzd7_, _Rgma_, _Snai1_, _Sdc1_, _Cpe_, _Bambi_.  The genes that are on are _Sfrp2_, _Sfrp4_ and _Bmp4_, all other other genes form a more continuous gradient of expression.  The most striking genes are _Cpe_ and _Bmp4_.  _Bmp4_ is probably expressed in more mature pre-adipocytes, which suggests this end of the pseudotime represents the more mature cells.  The cells on the left end are highest in _Myc_ and _Egr1_, so I wonder if they are actively proliferating cells.  If so, then I need to look at the impact of cell cycle as (a) a confounder, and (b) as a source of actively proliferating cells (can I also classify cells into particular parts of the cell cycle <- is that an informative analysis?).

Some of these genes look like they are quite strongly anti-correlated.  Is it possible to identify possible antagnostic gradients along this trajectory?

```{r, echo=FALSE, fig.height=9.5, fig.width=9.5}
part2.diff.cor <- cor(t(part2.diff.exprs), method='spearman')

pheatmap(part2.diff.cor,
         annotation_col=part2.genes.df,
         annotation_row=part2.genes.df,
         annotation_colors=annot.cols)
```

There are clearly some interesting behaviours amongst these genes, also indicated by the different groups of correlated genes.  Is it possible to identify genes that are associated with cell fate decisions?  For instance, _Cpe_ and _Bmp4_ have a switch-like behaviour that might suggest they are involved in making, or reinforcing a cell fate decision.  Can we also use this to determine at what point along a pseudotime that a decision is made?

Let's take a look at the wider correlation network of differentially expressed genes.  What patterns/clusters of co-expression are there?

This looks kind of cool.  It doesn't look so much like there are many discrete modules of gene expression _per se_, but gradients of gene expression changes.  This might be slightly artificial as these genes are selected for their differential expression between the middle of the pseudotime and the two arms.  Nonetheless it reinforces the idea that there is a continuum of development in these cells, and not necessarily discrete cell states.  It might be quite interesting to see how these genes compare to the ones with rapid onset/reduction in expression, and where they reach their peak/nadir along the diffusion component.  What functions do each of these clusters represent?  The blue, yellow and turquoise should be merged, but the others can stay as they are.  The background genes should be all of the DE genes, rather than all expressed genes in this case; I want to know what is specifically enriched in each cluster above and beyond what is enriched across all clusters.

## Cluster 1 up-regulated genes functional enrichment

```{r, echo=FALSE, fig.height=5.5, fig.width=12.5}
# use goseq
# use GO enrichment to test for specific pathways and function
# if there are enriched terms, this would at least imply some
# similar functional connection
# big cluster first
part2.all.genes <- unique(rownames(part2.nz))

part2.clust1.up <- unique(c(part2.de.merge.1v2$gene_id[part2.de.merge.1v2$Sig == 1 & (part2.de.merge.1v2$logFC < -1)],
                      part2.de.merge.3v1$gene_id[part2.de.merge.3v1$Sig == 1 & (part2.de.merge.3v1$logFC < -1)]))

part2.clust2.up <- unique(c(part2.de.merge.1v2$gene_id[part2.de.merge.1v2$Sig == 1 & (part2.de.merge.1v2$logFC > 1)],
                      part2.de.merge.3v2$gene_id[part2.de.merge.3v2$Sig == 1 & (part2.de.merge.3v2$logFC < -1)]))

part2.clust3.up <- unique(c(part2.de.merge.3v1$gene_id[part2.de.merge.1v2$Sig == 1 & (part2.de.merge.1v2$logFC > 1)],
                      part2.de.merge.3v2$gene_id[part2.de.merge.3v2$Sig == 1 & (part2.de.merge.3v2$logFC > 1)]))

part2.clust1.genes <- setdiff(part2.clust1.up, unique(c(part2.clust2.up, part2.clust3.up)))
part2.clust2.genes <- setdiff(part2.clust2.up, unique(c(part2.clust1.up, part2.clust3.up)))
part2.clust3.genes <- setdiff(part2.clust3.up, unique(c(part2.clust2.up, part2.clust1.up)))


de.big.vector <- as.integer(unique(part2.all.genes) %in% unique(part2.clust1.up))
names(de.big.vector) <- unique(part2.all.genes)

de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part2.gene.length))]

pwf.all <- nullp(de.big.vector, "mm10", "ensGene",
                 bias.data=part2.gene.length[intersect(names(de.big.vector), names(part2.gene.length))],
                 plot.fit=FALSE)
part2.GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
                           method="Wallenius", use_genes_without_cat=T)
# part2.GO.wall.all <-  goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
#                            method="Wallenius", use_genes_without_cat=T)

part2.GO.wall.all$padjust <- p.adjust(part2.GO.wall.all$over_represented_pvalue, method="BH")
part2.GO.wall.all$foldEnrich <- ((part2.GO.wall.all$numDEInCat/length(unique(part2.clust1.up)))/(part2.GO.wall.all$numInCat/length(part2.gene.length)))

# select on p-value and foldEnrich
part2.GO.sig.all <- part2.GO.wall.all[(part2.GO.wall.all$padjust <= 0.01), ]
part2.go_cats.all <- data.frame(part2.GO.sig.all$category, part2.GO.sig.all$foldEnrich)

part2.p.enrich <- ggplot(part2.GO.sig.all,
                   aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
                       size=foldEnrich)) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C")

part2.GO.all <- part2.GO.sig.all[order(part2.GO.sig.all$padjust, decreasing=FALSE), ]

part2.p.terms <- ggplot(na.omit(part2.GO.all[1:25, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=60)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")

plot_grid(part2.p.enrich, part2.p.terms, ncol=2, rel_widths=c(0.4, 0.6))
```

Mostly enriched for inflammatory response, proliferation, development and adhesion.


```{r}
part2.c1.pterms <- ggplot(na.omit(part2.GO.all[1:5, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1, breaks=c(min(round(-log10(part2.GO.all[1:5, ]$padjust), 1)),
                                                             max(round(-log10(part2.GO.all[1:5, ]$padjust), 1)))) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=45)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment") +
  theme(axis.text=element_text(size=18))

ggsave(part2.c1.pterms,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Cluster1_GOenrich-bar-Kmeans2.png",
       height=4.75, width=6.75, dpi=300)
```

The genes up-regulated in cluster 1 are largely enriched for cytokine and immunne response genes.

```{r, echo=FALSE, fig.height=4.5, fig.width=5.5}
psorder <- part2.dc.merge$Sample[order(part2.dc.merge$DPT1, decreasing=TRUE)]
part2.annot.df <- cbind.data.frame(part2.dc.merge[, c("ClusterLabel")])
colnames(part2.annot.df) <- c("ClusterLabel")
part2.annot.df$ClusterLabel <- as.factor(part2.annot.df$ClusterLabel)
rownames(part2.annot.df) <- part2.dc.merge$Sample

pheatmap(part2.nz[part2.clust1.up, psorder],
         show_rownames=FALSE, show_colnames=FALSE,
         cluster_cols=FALSE, annotation_col=part2.annot.df,
         annotation_colors=list(ClusterLabel=label.map))
```


#### Cluster 2 genes functional enrichment

```{r, echo=FALSE, fig.height=5.5, fig.width=12.5}
de.big.vector <- as.integer(unique(part2.all.genes) %in% unique(part2.clust2.up))
names(de.big.vector) <- unique(part2.all.genes)

de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part2.gene.length))]

pwf.all <- nullp(de.big.vector, "mm10", "ensGene", 
                 bias.data=part2.gene.length[intersect(names(de.big.vector), names(part2.gene.length))],
                 plot.fit=FALSE)
part2.GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
                           method="Wallenius", use_genes_without_cat=T)
# part2.GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
#                       method="Wallenius", use_genes_without_cat=T)

part2.GO.wall.all$padjust <- p.adjust(part2.GO.wall.all$over_represented_pvalue, method="BH")
part2.GO.wall.all$foldEnrich <- ((part2.GO.wall.all$numDEInCat/length(unique(part2.clust2.up)))/(part2.GO.wall.all$numInCat/length(part2.gene.length)))

# select on p-value and foldEnrich
part2.GO.sig.all <- part2.GO.wall.all[(part2.GO.wall.all$padjust <= 0.01), ]
part2.go_cats.all <- data.frame(part2.GO.sig.all$category, part2.GO.sig.all$foldEnrich)

part2.p.enrich <- ggplot(part2.GO.sig.all,
                   aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
                       size=foldEnrich)) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C")

part2.GO.all <- part2.GO.sig.all[order(part2.GO.sig.all$padjust, decreasing=FALSE), ]

part2.p.terms <- ggplot(na.omit(part2.GO.all[1:20, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")

plot_grid(part2.p.enrich, part2.p.terms, ncol=2)
```

Strong enrichments for cell adhesion and signalling in cluster 2.

```{r}
part2.c2.pterms <- ggplot(na.omit(part2.GO.all[1:5, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1, breaks=c(min(round(-log10(part2.GO.all[1:5, ]$padjust), 1)),
                                                             max(round(-log10(part2.GO.all[1:5, ]$padjust), 1)))) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=45)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment") +
  theme(axis.text=element_text(size=18))

ggsave(part2.c2.pterms,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Cluster2_GOenrich-bar-Kmeans2.png",
       height=4.75, width=6.75, dpi=300)
```


```{r, echo=FALSE, fig.height=4.5, fig.width=5.5}
pheatmap(part2.nz[part2.clust2.up, psorder],
         show_rownames=FALSE, show_colnames=FALSE,
         cluster_cols=FALSE, annotation_col=part2.annot.df,
         annotation_colors=list(ClusterLabel=label.map))
```

#### Cluster 3 functional enrichment

```{r, echo=FALSE, fig.height=5.5, fig.width=12.5}
de.big.vector <- as.integer(unique(part2.all.genes) %in% unique(part2.clust3.up))
names(de.big.vector) <- unique(part2.all.genes)

de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part2.gene.length))]

pwf.all <- nullp(de.big.vector, "mm10", "ensGene", 
                 bias.data=part2.gene.length[intersect(names(de.big.vector), names(part2.gene.length))],
                 plot.fit=FALSE)
part2.GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
                           method="Wallenius", use_genes_without_cat=T)
# part2.GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
#                       method="Wallenius", use_genes_without_cat=T)

part2.GO.wall.all$padjust <- p.adjust(part2.GO.wall.all$over_represented_pvalue, method="BH")
part2.GO.wall.all$foldEnrich <- ((part2.GO.wall.all$numDEInCat/length(unique(part2.clust3.up)))/(part2.GO.wall.all$numInCat/length(part2.gene.length)))

# select on p-value and foldEnrich
part2.GO.sig.all <- part2.GO.wall.all[(part2.GO.wall.all$padjust <= 0.01), ]
part2.go_cats.all <- data.frame(part2.GO.sig.all$category, part2.GO.sig.all$foldEnrich)

part2.p.enrich <- ggplot(part2.GO.sig.all,
                   aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
                       size=foldEnrich)) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C")

part2.GO.all <- part2.GO.sig.all[order(part2.GO.sig.all$padjust, decreasing=FALSE), ]

part2.p.terms <- ggplot(na.omit(part2.GO.all[1:20, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")

plot_grid(part2.p.enrich, part2.p.terms, ncol=2)
```

There is a strong enrichment for developmental processes here.

```{r}
part2.c3.pterms <- ggplot(na.omit(part2.GO.all[1:5, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1, breaks=c(min(round(-log10(part2.GO.all[1:5, ]$padjust), 1)),
                                                             max(round(-log10(part2.GO.all[1:5, ]$padjust), 1)))) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=45)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment") +
  theme(axis.text=element_text(size=18))

ggsave(part2.c3.pterms,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Cluster3_GOenrich-bar-Kmeans2.png",
       height=4.75, width=6.75, dpi=300)
```



```{r, echo=FALSE, fig.height=4.5, fig.width=5.5}
pheatmap(part2.nz[part2.clust3.up, psorder],
         show_rownames=FALSE, show_colnames=FALSE,
         cluster_cols=FALSE, annotation_col=part2.annot.df,
         annotation_colors=list(ClusterLabel=label.map))
```

These are all quite strongly expressed and broadly expressed; it is hard to see where the differences arise.

Just looking at the DE genes between clusters isn't necessarily that informative.  Perhaps finds genes that are DE along each branch separately.

#### Differential Expression Analysis: Pseudotime-dependent

```{r, echo=FALSE}
lptime <- part2.dc.merge$Sample
design.mat.lptime <- model.matrix(~ DPT1 + Plate, data=part2.dc.merge[part2.dc.merge$Sample %in% lptime, ])

# fit a linear model
fit.lptime <- lmFit(part2.nz[, colnames(part2.nz) %in% lptime], design.mat.lptime)
fit.lptime <- eBayes(fit.lptime)
sum.res.lptime <- summary(decideTests(fit.lptime))

de.res.lptime <- topTable(fit.lptime, coef=2, n=Inf, sort="p", p=1.0)
de.res.lptime$Sig <- 0
de.res.lptime$Sig[de.res.lptime$adj.P.Val <= 0.01] <- 1
de.res.lptime$Sig <- as.factor(de.res.lptime$Sig)

de.res.lptime$Diff <- 0
de.res.lptime$Diff[de.res.lptime$logFC < 0 & de.res.lptime$Sig == 1] <- -1
de.res.lptime$Diff[de.res.lptime$logFC > 0 & de.res.lptime$Sig == 1] <- 1

de.res.lptime$gene_id <- rownames(de.res.lptime)
de.merge.lptime <- merge(de.res.lptime, gene_symbol,
                  by.x='gene_id', by.y='ensembl_gene_id', all.x=TRUE)
gene.top.lptime <- de.merge.lptime$external_gene_name

# plot the top 6 and bottom 6 gene symbols
top6 <- de.merge.lptime$external_gene_name[order(de.merge.lptime$logFC, decreasing=TRUE)][1:6]
bottom6 <- de.merge.lptime$external_gene_name[order(de.merge.lptime$logFC, decreasing=FALSE)][1:6]
gene.top.lptime[!gene.top.lptime %in% c(top6, bottom6)] <- ""

# gene.top.lptime[(de.merge.lptime$logFC > -4.75) & (de.merge.lptime$logFC < 4.75)] <- ""
# gene.top.lptime[is.na(gene.top.lptime)] <- ""

# give eGFP it's proper name
de.merge.lptime$external_gene_name[de.merge.lptime$gene_id == "eGFP"] <- "eGFP"

write.table(de.merge.lptime,
            file="~/Dropbox/Adipose/Adipose-DEgenes_Leftptime-Kmeans2.tsv",
            sep="\t", quote=F)

ltime.ma <- ggplot(de.merge.lptime, aes(x=AveExpr, y=logFC, colour=Sig)) +
  geom_point(size=1, alpha=0.8) + theme_mike() +
  scale_colour_manual(values=c("#A3A3A3", "#FF0000")) +
  labs(x=expression(paste("Mean log"[2], " Expression")),
       y=expression(paste("log"[2], " Fold Change"))) +
  guides('colour'=FALSE) +
  ylim(c(-10, 10)) +
  geom_label_repel(aes(label=gene.top.lptime),
                  colour='black',
                  nudge_x=0.01,
                  nudge_y=0.01,
                  size=3)

ggsave(ltime.ma,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/MAplot-LeftPseudotime-Kmeans2.png",
       height=4.25, width=6.75, dpi=300)

ltime.ma
```


```{r}
sum.res.lptime
```

There are a fairly decent number of genes DE along the pseudotime (~linearly).

```{r, echo=FALSE, fig.height=6.5, fig.width=6.5}
# heatmap of partition 1 pseudotime-dependent DE genes
part2.diff.genes <- unique(c(de.merge.lptime$gene_id[de.merge.lptime$Sig == 1 & (abs(de.merge.lptime$logFC) > 0)]))

part2.ltime.up <- unique(c(de.merge.lptime$gene_id[de.merge.lptime$Sig == 1 & (de.merge.lptime$logFC > 0)]))
part2.ltime.down <- unique(c(de.merge.lptime$gene_id[de.merge.lptime$Sig == 1 & (de.merge.lptime$logFC < -0)]))

# only plot the genes unique to each cluster
part2.diff.exprs <- part2.nz[part2.diff.genes, part2.dc.merge$Sample[order(part2.dc.merge$DPT1, decreasing=TRUE)]]

part2.genes.df <- data.frame(cbind(part2.diff.genes))
colnames(part2.genes.df) <- c("gene_id")
rownames(part2.genes.df) <- part2.genes.df[, 1]
part2.genes.df$GeneGroup <- ""
part2.genes.df[part2.genes.df$gene_id %in% part2.ltime.down, ]$GeneGroup <- "Early"
part2.genes.df[part2.genes.df$gene_id %in% part2.ltime.up, ]$GeneGroup <- "Late"

part2.genes.df$GeneGroup <- as.factor(part2.genes.df$GeneGroup)

#genes.cols <- 
genes.cols <- c("#00F7FF", "#FF9E00")
names(genes.cols) <- c("Early", "Late")
part2.genes.df <- as.data.frame(part2.genes.df[, -1])
rownames(part2.genes.df) <- part2.diff.genes
colnames(part2.genes.df) <- "GeneGroup"

part2.annot.df <- part2.dc.merge[, c("ClusterLabel", "DPT1")]
part2.annot.df$ClusterLabel <- as.factor(part2.annot.df$ClusterLabel)
part2.annot.df$DPT1 <- as.numeric(part2.annot.df$DPT1)
rownames(part2.annot.df) <- part2.dc.merge$Sample

dc.cols <- colorRampPalette(c("#FAFF60", "#8615B0"))(dim(part2.diff.exprs)[2])
names(dc.cols) <- part2.annot.df$DPT1
  
marker.cols <- gfp.cols
names(marker.cols) <- as.factor(c("GFP+", "GFP-"))

cluster.cols <- label.map

part2.annot.cols <- list(ClusterLabel=cluster.cols,
                         GeneGroup=genes.cols,
                         DPT1=dc.cols)

gene.order <- order(part2.genes.df$GeneGroup)

pheatmap(part2.diff.exprs[gene.order, ], show_rownames=FALSE, show_colnames=FALSE,
         cluster_rows=FALSE, annotation_col=part2.annot.df,
         annotation_row=part2.genes.df,
         annotation_colors=part2.annot.cols,
         cluster_cols=FALSE,
         filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Pseudotime_DE_genes-Kmeans2-heatmap.png",
         height=6.75, width=6.75, res=300)
```

#### Left pseudotime functional enrichment

#### Up-regulated processes

```{r, echo=FALSE, fig.height=5.5, fig.width=12.5}
ltime.up <- de.merge.lptime$gene_id[de.merge.lptime$Sig == 1 & de.merge.lptime$logFC > 0]
ltime.down <- de.merge.lptime$gene_id[de.merge.lptime$Sig == 1 & de.merge.lptime$logFC < 0]

de.big.vector <- as.integer(unique(part2.all.genes) %in% unique(ltime.up))
names(de.big.vector) <- unique(part2.all.genes)

de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part2.gene.length))]

pwf.all <- nullp(de.big.vector, "mm10", "ensGene", bias.data=part2.gene.length[intersect(names(de.big.vector), names(part2.gene.length))],
                 plot.fit=FALSE)
GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
                           method="Wallenius", use_genes_without_cat=T)
# GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
#                       method="Wallenius", use_genes_without_cat=T)

GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(ltime.up)))/(GO.wall.all$numInCat/length(part2.gene.length)))

# select on p-value and foldEnrich
GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)

p.enrich <- ggplot(GO.sig.all,
                   aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
                       size=foldEnrich)) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C")

GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(GO.all[1:20, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")

plot_grid(p.enrich, p.terms, ncol=2, rel_widths = c(0.3, 0.7))
```


#### Down-regulated processes

```{r, echo=FALSE, fig.height=5.5, fig.width=12.5}
de.big.vector <- as.integer(unique(part2.all.genes) %in% unique(ltime.down))
names(de.big.vector) <- unique(part2.all.genes)

de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part2.gene.length))]

pwf.all <- nullp(de.big.vector, "mm10", "ensGene", bias.data=part2.gene.length[intersect(names(de.big.vector), names(part2.gene.length))],
                 plot.fit=FALSE)
GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
                           method="Wallenius", use_genes_without_cat=T)
# GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
#                       method="Wallenius", use_genes_without_cat=T)

GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(ltime.down)))/(GO.wall.all$numInCat/length(part2.gene.length)))

# select on p-value and foldEnrich
GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)

p.enrich <- ggplot(GO.sig.all,
                   aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
                       size=foldEnrich)) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C")

GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(GO.all[1:20, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")

plot_grid(p.enrich, p.terms, ncol=2)
```

There are quite a lot of enriched terms in each of the up and down-regulated processes along the left branch of the trajectory.  If these are co-ordinated changes then we should be able to cluster the genes together, and find modules of genes that share functional annotations.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
left.diff.exprs <- part2.nz[unique(c(ltime.down, ltime.up)),
                              !colnames(part2.nz) %in% multi.cell]
left.de.cor <- cor(t(left.diff.exprs), method='spearman')

left.de.dist <- as.dist(abs(left.de.cor - 1))
left.de.clust <- flashClust(left.de.dist, method="average")
left.de.cut <- cutreeDynamicTree(left.de.clust, deepSplit=TRUE, minModuleSize=40)
left.de.cols <- labels2colors(left.de.cut)

left.annot.df <- data.frame(cbind(left.de.cols))
rownames(left.annot.df) <- rownames(left.diff.exprs)
colnames(left.annot.df) <- "Cluster"
left.gene.clust.cols <- col2hcl(unique(left.de.cols))
names(left.gene.clust.cols) <- unique(left.de.cols)
left.annot.cols <- list("Cluster"=left.gene.clust.cols)

pheatmap(left.de.cor,
         show_rownames=FALSE, show_colnames=FALSE,
         annotation_col=left.annot.df, annotation_colors=left.annot.cols,
         clustering_distance_cols=left.de.dist,
         clustering_distance_rows=left.de.dist,
         clustering_method="average",
         filename="~/Dropbox/Adipose/Adipose-LPseudotime_DEgenes-heatmap-Kmeans2.png",
         height=5.75, width=5.75, res=300)
```

For the most part there are genes that go up and genes that go down.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
left.annot.df$gene_id <- rownames(left.annot.df)

write.table(left.annot.df,
            file="~/Dropbox/Adipose/Adipose_LeftPseudotime_DEgenes-clusters-Kmeans2.tsv",
            sep="\t", quote=FALSE, row.names=FALSE)

left.blue.genes <- rownames(left.annot.df)[left.annot.df$Cluster == 'blue']
left.yellow.genes <- rownames(left.annot.df)[left.annot.df$Cluster == 'yellow']
left.brown.genes <- rownames(left.annot.df)[left.annot.df$Cluster == 'brown']
left.turquoise.genes <- rownames(left.annot.df)[left.annot.df$Cluster == 'turquoise']
left.green.genes <- rownames(left.annot.df)[left.annot.df$Cluster == 'green']
left.red.genes <- rownames(left.annot.df)[left.annot.df$Cluster == 'red']
left.black.genes <- rownames(left.annot.df)[left.annot.df$Cluster == 'black']
left.grey.genes <- rownames(left.annot.df)[left.annot.df$Cluster == 'grey']
left.pink.genes <- rownames(left.annot.df)[left.annot.df$Cluster == 'pink']
left.magenta.genes <- rownames(left.annot.df)[left.annot.df$Cluster == 'magenta']
```


#### Functional enrichment test: blue cluster

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.5, fig.width=12.5}
de.big.vector <- as.integer(unique(part2.all.genes) %in% unique(left.blue.genes))
names(de.big.vector) <- unique(part2.all.genes)

de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part2.gene.length))]

pwf.all <- nullp(de.big.vector, "mm10", "ensGene", bias.data=part2.gene.length[intersect(names(de.big.vector), names(part2.gene.length))],
                 plot.fit=FALSE)
GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
                           method="Wallenius", use_genes_without_cat=T)
# GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
#                       method="Wallenius", use_genes_without_cat=T)

GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(left.blue.genes)))/(GO.wall.all$numInCat/length(part2.gene.length)))

# select on p-value and foldEnrich
GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)

p.enrich <- ggplot(GO.sig.all,
                   aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
                       size=foldEnrich)) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C")

GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(GO.all[1:20, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")

write.table(GO.all,
            file="~/Dropbox/Adipose/Adipose-LeftPseudotime_GOenrich-blueCluster-Kmeans2.tsv",
            sep="\t", row.names=FALSE, quote=FALSE)

plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.3, 0.7))
```

There are still `r dim(GO.all)[1]` enriched categories within this cluster of genes, many of which are quite generic and uninformative.

#### Functional enrichment test: brown cluster

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.5, fig.width=12.5}
de.big.vector <- as.integer(unique(part2.all.genes) %in% unique(left.brown.genes))
names(de.big.vector) <- unique(part2.all.genes)

de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part2.gene.length))]

pwf.all <- nullp(de.big.vector, "mm10", "ensGene", bias.data=part2.gene.length[intersect(names(de.big.vector), names(part2.gene.length))],
                 plot.fit=FALSE)
GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
                           method="Wallenius", use_genes_without_cat=T)
# GO.wal.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
#                       method="Wallenius", use_genes_without_cat=T)

GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(left.brown.genes)))/(GO.wall.all$numInCat/length(part2.gene.length)))

# select on p-value and foldEnrich
GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)

p.enrich <- ggplot(GO.sig.all,
                   aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
                       size=foldEnrich)) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C")

GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(GO.all[1:20, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")

write.table(GO.all,
            file="~/Dropbox/Adipose/Adipose-LeftPseudotime_GOenrich-brownCluster-Kmeans2.tsv",
            sep="\t", row.names=FALSE, quote=FALSE)

plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.3, 0.7))
```

Why would there be an enrichment for terms related to innate immune response when there are no macrophages in amongst these cells??  Is this an indication of a response to a pro-inflammatory signal?

#### Functional enrichment test: turquoise cluster

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.5, fig.width=12.5}
de.big.vector <- as.integer(unique(part2.all.genes) %in% unique(left.turquoise.genes))
names(de.big.vector) <- unique(part2.all.genes)

de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part2.gene.length))]

pwf.all <- nullp(de.big.vector, "mm10", "ensGene", bias.data=part2.gene.length[intersect(names(de.big.vector), names(part2.gene.length))],
                 plot.fit=FALSE)
GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
                           method="Wallenius", use_genes_without_cat=T)
# GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
#                       method="Wallenius", use_genes_without_cat=T)

GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(left.turquoise.genes)))/(GO.wall.all$numInCat/length(part2.gene.length)))

# select on p-value and foldEnrich
GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)

p.enrich <- ggplot(GO.sig.all,
                   aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
                       size=foldEnrich)) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C")

GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(GO.all[1:20, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")

write.table(GO.all,
            file="~/Dropbox/Adipose/Adipose-LeftPseudotime_GOenrich-turquoiseCluster-Kmeans2.tsv",
            sep="\t", row.names=FALSE, quote=FALSE)

plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.3, 0.7))
```

Adhesion, locomotion and differentiation imply a mesenchymal to epithelial transition.

#### Functional enrichment test: yellow cluster

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.5, fig.width=12.5}
de.big.vector <- as.integer(unique(part2.all.genes) %in% unique(left.yellow.genes))
names(de.big.vector) <- unique(part2.all.genes)

de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part2.gene.length))]

pwf.all <- nullp(de.big.vector, "mm10", "ensGene", bias.data=part2.gene.length[intersect(names(de.big.vector), names(part2.gene.length))],
                 plot.fit=FALSE)
GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
                           method="Wallenius", use_genes_without_cat=T)
# GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
#                       method="Wallenius", use_genes_without_cat=T)

GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(left.yellow.genes)))/(GO.wall.all$numInCat/length(part2.gene.length)))

# select on p-value and foldEnrich
GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)

p.enrich <- ggplot(GO.sig.all,
                   aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
                       size=foldEnrich)) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C")

GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(GO.all[1:20, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")

write.table(GO.all,
            file="~/Dropbox/Adipose/Adipose-LeftPseudotime_GOenrich-yellowCluster-Kmeans2.tsv",
            sep="\t", row.names=FALSE, quote=FALSE)

plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.3, 0.7))
```

Adhesion, locomotion and differentiation imply a mesenchymal to epithelial transition.

#### Functional enrichment test: red cluster

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.5, fig.width=12.5}
de.big.vector <- as.integer(unique(part2.all.genes) %in% unique(left.red.genes))
names(de.big.vector) <- unique(part2.all.genes)

de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part2.gene.length))]

pwf.all <- nullp(de.big.vector, "mm10", "ensGene", bias.data=part2.gene.length[intersect(names(de.big.vector), names(part2.gene.length))],
                 plot.fit=FALSE)
GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
                           method="Wallenius", use_genes_without_cat=T)
# GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
#                       method="Wallenius", use_genes_without_cat=T)

GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(left.red.genes)))/(GO.wall.all$numInCat/length(part2.gene.length)))

# select on p-value and foldEnrich
GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)

p.enrich <- ggplot(GO.sig.all,
                   aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
                       size=foldEnrich)) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C")

GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(GO.all[1:20, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")

write.table(GO.all,
            file="~/Dropbox/Adipose/Adipose-LeftPseudotime_GOenrich-redCluster-Kmeans2.tsv",
            sep="\t", row.names=FALSE, quote=FALSE)

plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.3, 0.7))
```


#### Functional enrichment test: green cluster

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.5, fig.width=12.5}
de.big.vector <- as.integer(unique(part2.all.genes) %in% unique(left.green.genes))
names(de.big.vector) <- unique(part2.all.genes)

de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part2.gene.length))]

pwf.all <- nullp(de.big.vector, "mm10", "ensGene", bias.data=part2.gene.length[intersect(names(de.big.vector), names(part2.gene.length))],
                 plot.fit=FALSE)
GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
                           method="Wallenius", use_genes_without_cat=T)
# GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
#                       method="Wallenius", use_genes_without_cat=T)

GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(left.green.genes)))/(GO.wall.all$numInCat/length(part2.gene.length)))

# select on p-value and foldEnrich
GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)

p.enrich <- ggplot(GO.sig.all,
                   aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
                       size=foldEnrich)) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C")

GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(GO.all[1:20, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")

write.table(GO.all,
            file="~/Dropbox/Adipose/Adipose-LeftPseudotime_GOenrich-greenCluster-Kmeans2.tsv",
            sep="\t", row.names=FALSE, quote=FALSE)

plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.3, 0.7))
```


Most of the differences between the left and right branches, and between the two trajectories, relate to morphological changes and response 
to some external stimulus.

#### Functional enrichment test: pink cluster

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.5, fig.width=12.5}
de.big.vector <- as.integer(unique(part2.all.genes) %in% unique(left.pink.genes))
names(de.big.vector) <- unique(part2.all.genes)

de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part2.gene.length))]

pwf.all <- nullp(de.big.vector, "mm10", "ensGene", bias.data=part2.gene.length[intersect(names(de.big.vector), names(part2.gene.length))],
                 plot.fit=FALSE)
GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
                           method="Wallenius", use_genes_without_cat=T)
# GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
#                       method="Wallenius", use_genes_without_cat=T)

GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(left.pink.genes)))/(GO.wall.all$numInCat/length(part2.gene.length)))

# select on p-value and foldEnrich
GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)

p.enrich <- ggplot(GO.sig.all,
                   aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
                       size=foldEnrich)) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C")

GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(GO.all[1:20, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")

write.table(GO.all,
            file="~/Dropbox/Adipose/Adipose-LeftPseudotime_GOenrich-pinkCluster-Kmeans2.tsv",
            sep="\t", row.names=FALSE, quote=FALSE)

plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.3, 0.7))
```

#### Functional enrichment test: black cluster

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.5, fig.width=12.5}
de.big.vector <- as.integer(unique(part2.all.genes) %in% unique(left.black.genes))
names(de.big.vector) <- unique(part2.all.genes)

de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part2.gene.length))]

pwf.all <- nullp(de.big.vector, "mm10", "ensGene", bias.data=part2.gene.length[intersect(names(de.big.vector), names(part2.gene.length))],
                 plot.fit=FALSE)
GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
                           method="Wallenius", use_genes_without_cat=T)
# GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
#                       method="Wallenius", use_genes_without_cat=T)

GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(left.black.genes)))/(GO.wall.all$numInCat/length(part2.gene.length)))

# select on p-value and foldEnrich
GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)

p.enrich <- ggplot(GO.sig.all,
                   aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
                       size=foldEnrich)) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C")

GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(GO.all[1:20, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")

write.table(GO.all,
            file="~/Dropbox/Adipose/Adipose-LeftPseudotime_GOenrich-pinkCluster-Kmeans2.tsv",
            sep="\t", row.names=FALSE, quote=FALSE)

plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.3, 0.7))
```

#### Functional enrichment test: grey cluster

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.5, fig.width=12.5}
de.big.vector <- as.integer(unique(part2.all.genes) %in% unique(left.grey.genes))
names(de.big.vector) <- unique(part2.all.genes)

de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part2.gene.length))]

pwf.all <- nullp(de.big.vector, "mm10", "ensGene", bias.data=part2.gene.length[intersect(names(de.big.vector), names(part2.gene.length))],
                 plot.fit=FALSE)
GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
                           method="Wallenius", use_genes_without_cat=T)
# GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
#                       method="Wallenius", use_genes_without_cat=T)

GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(left.grey.genes)))/(GO.wall.all$numInCat/length(part2.gene.length)))

# select on p-value and foldEnrich
GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)

p.enrich <- ggplot(GO.sig.all,
                   aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
                       size=foldEnrich)) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C")

GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(GO.all[1:20, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")

write.table(GO.all,
            file="~/Dropbox/Adipose/Adipose-LeftPseudotime_GOenrich-pinkCluster-Kmeans2.tsv",
            sep="\t", row.names=FALSE, quote=FALSE)

plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.3, 0.7))
```

#### Functional enrichment test: magenta cluster

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.5, fig.width=12.5}
de.big.vector <- as.integer(unique(part2.all.genes) %in% unique(left.magenta.genes))
names(de.big.vector) <- unique(part2.all.genes)

de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(part2.gene.length))]

pwf.all <- nullp(de.big.vector, "mm10", "ensGene", bias.data=part2.gene.length[intersect(names(de.big.vector), names(part2.gene.length))],
                 plot.fit=FALSE)
GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
                           method="Wallenius", use_genes_without_cat=T)
# GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
#                       method="Wallenius", use_genes_without_cat=T)

GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(left.magenta.genes)))/(GO.wall.all$numInCat/length(part2.gene.length)))

# select on p-value and foldEnrich
GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)

p.enrich <- ggplot(GO.sig.all,
                   aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
                       size=foldEnrich)) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C")

GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(GO.all[1:20, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")

write.table(GO.all,
            file="~/Dropbox/Adipose/Adipose-LeftPseudotime_GOenrich-pinkCluster-Kmeans2.tsv",
            sep="\t", row.names=FALSE, quote=FALSE)

plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.3, 0.7))
```

Most of the differences between the left and right branches, and between the two trajectories, relate to morphological changes and response 
to some external stimulus.


#### Differential Expression Analysis: Cluster2 Partition 1 Vs. Partition 2

There are essentially 2 groups of naive cells that are transcriptionally distinct (w.r.t. to the 2nd diffusion component).  what are the genes 
that are driving this difference?  Does this indicate cells that have previously made a fate choice, and are only now committed to their 
ultimate fate??

```{r, echo=FALSE, message=FALSE, warning=FALSE}
clust2.cells <- dc.merge$Sample[dc.merge$Cluster == 2]
clust2.exprs <- adipose.nz[, colnames(adipose.nz) %in% clust2.cells]
clust2.meta <- dc.merge[dc.merge$Sample %in% clust2.cells, ]

# force the same ordering of cells between gene expression and meta-data
rownames(clust2.meta) <- clust2.meta$Sample
clust2.meta <- clust2.meta[colnames(clust2.exprs), ]

clust2.design.mat <- model.matrix(~ 0 + Kmeans + Plate, data=clust2.meta)

# fit a linear model
clust2.fit <- lmFit(clust2.exprs, clust2.design.mat)

clust2.contrast.mat <-  makeContrasts(Kmeans2-Kmeans1,
                                     levels=clust2.design.mat)

clust2.fit <- contrasts.fit(clust2.fit, contrasts=clust2.contrast.mat)
clust2.fit <- eBayes(clust2.fit)
clust2.sum.res <- summary(decideTests(clust2.fit))

clust2.de.res <- topTable(clust2.fit, coef=1, n=Inf, sort="p", p=1.0)
clust2.de.res$Sig <- 0
clust2.de.res$Sig[clust2.de.res$adj.P.Val <= 0.01] <- 1
clust2.de.res$Sig <- as.factor(clust2.de.res$Sig)

clust2.de.res$Diff <- 0
clust2.de.res$Diff[clust2.de.res$logFC < 0 & clust2.de.res$Sig == 1] <- -1
clust2.de.res$Diff[clust2.de.res$logFC > 0 & clust2.de.res$Sig == 1] <- 1

clust2.de.res$gene_id <- rownames(clust2.de.res)
clust2.de.merge <- merge(clust2.de.res, gene_symbol,
                            by.x='gene_id', by.y='ensembl_gene_id', all.x=TRUE)

clust2.gene.top <- clust2.de.merge$external_gene_name
clust2.gene.top[clust2.de.merge$Sig == 0] <- ""

# plot the top 6 and bottom 6 gene symbols
top6 <- clust2.de.merge$external_gene_name[order(clust2.de.merge$logFC, decreasing=TRUE)][1:6]
bottom6 <- clust2.de.merge$external_gene_name[order(clust2.de.merge$logFC, decreasing=FALSE)][1:6]
clust2.gene.top[!clust2.gene.top %in% c(top6, bottom6)] <- ""

clust2.gene.top[clust2.de.merge$logFC > -2 & clust2.de.merge$logFC < 5] <- ""
clust2.gene.top[is.na(clust2.gene.top)] <- ""

write.table(clust2.de.merge,
            file="~/Dropbox/Adipose/Adipose-DEgenes-Kmeans1V2-Cluster2.tsv",
            sep="\t", quote=F)

clust2.ma <- ggplot(clust2.de.merge, aes(x=AveExpr, y=logFC, colour=Sig)) +
  geom_point(size=1, alpha=0.8) + theme_mike() +
  scale_colour_manual(values=c("#A3A3A3", "#FF0000")) +
  labs(x=expression(paste("Mean log"[2], " Expression")),
       y=expression(paste("log"[2], " Fold Change"))) +
  guides('colour'=FALSE) +
  ylim(c(-10, 10)) +
  geom_label_repel(aes(label=clust2.gene.top),
                  colour='black',
                  nudge_x=0.01,
                  nudge_y=0.01,
                  size=3)

ggsave(clust2.ma,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Cluster2-Kmeans1Vs2-MAplot.png",
       height=4.25, width=6.75, dpi=300)
clust2.ma
```

Hmm, almost all of the gene expression differences are in a single direction, i.e. they are up-regulated in the second partition 
(the smaller group with cluster 1).  These look like lots of stress, and inflammatory response genes...  
Is this a genuine feature of these cells, or some experimental artifact?  What are the functional enrichments in the up-regulated genes?


```{r, echo=FALSE, fig.height=6.75, fig.width=12.5}
clust2.up <- de.merge.lptime$gene_id[clust2.de.merge$Sig == 1 & clust2.de.merge$logFC > 0]
all.genes <- de.merge.lptime$gene_id

gene.length <- gene.tpm$Length[unique(gene.tpm$Name) %in% unique(all.genes)]
names(gene.length) <- intersect(unique(gene.tpm$Name) , unique(all.genes))

de.big.vector <- as.integer(unique(all.genes) %in% unique(clust2.up))
names(de.big.vector) <- unique(all.genes)

de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(gene.length))]

pwf.all <- nullp(de.big.vector, "mm10", "ensGene", bias.data=gene.length[intersect(names(de.big.vector), names(gene.length))],
                 plot.fit=FALSE)
GO.wall.all  <- goseq(pwf.all, genome="mm10", id="ensGene", gene2cat=hallmark.go,
                           method="Wallenius", use_genes_without_cat=T)
# GO.wall.all <- goseq(pwf.all, genome="mm10", id="ensGene", test.cats=c("GO:BP"),
#                       method="Wallenius", use_genes_without_cat=T)

GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(clust2.up)))/(GO.wall.all$numInCat/length(gene.length)))

# select on p-value and foldEnrich
GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)

p.enrich <- ggplot(GO.sig.all,
                   aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
                       size=foldEnrich)) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C")

GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(GO.all[1:20, ]),
                  aes(x=reorder(category, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  scale_x_discrete(labels=function(X) str_wrap(X, width=50)) +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")

plot_grid(p.enrich, p.terms, ncol=2)
```

It looks like these cells are responding to an inflammatory stimulus, and are consequently more transcriptionally active than the cells from 
partition 1.  This effect is not driven by a particular mouse, maybe it's something about the surgery to remove the fat depots and extract 
the SVF?

Plot marker gene expression by partition within each cluster.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
kmeans.cols <- c("#FF7800", "#002EFF")
names(kmeans.cols) <- c(1, 2)

# use the biological cluster labels
dc.merge$ClusterLabel <- ""
dc.merge$ClusterLabel[dc.merge$Cluster %in% c(2, 3)] <- "Mature"
dc.merge$ClusterLabel[dc.merge$Cluster %in% c(1)] <- "Naive"
dc.merge$ClusterLabel[dc.merge$Cluster %in% c(4)] <- "Intermediate"
dc.merge$ClusterLabel <- factor(dc.merge$ClusterLabel,
                                  levels=c("Naive", "Intermediate", "Mature"),
                                  labels=c("Naive", "Intermediate", "Mature"))

wt1.by.part <- ggplot(dc.merge, aes(x=ClusterLabel, y=Wt1, fill=Kmeans)) +
  geom_boxplot(alpha=0.6) +
  geom_jitter(position=position_jitterdodge(jitter.width = 0.5),
              aes(shape=Kmeans), size=3) +
  theme_bw() +
  scale_shape_manual(values=c(21, 24)) +
  scale_fill_manual(values=kmeans.cols) +
  labs(y=expression(bold(paste("Wt1 log"[2], " Expression"))), x="Cluster") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=12, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14), 
        panel.grid.major=element_blank(),
        panel.border=element_blank(),
        axis.line=element_line(colour="black")) +
  guides(fill=guide_legend(title="Partition"),
         shape=guide_legend(title="Partition"))

ggsave(wt1.by.part,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Wt1_partition-boxplot.png",
       height=6.5/3, width=6.5, dpi=300)

wt1.by.part
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
cd34.by.part <- ggplot(dc.merge, aes(x=ClusterLabel, y=Cd34, fill=Kmeans)) +
  geom_boxplot(alpha=0.6) +
  geom_jitter(position=position_jitterdodge(jitter.width = 0.5),
              aes(shape=Kmeans), size=3) +
  theme_bw() +
  scale_shape_manual(values=c(21, 24)) +
  scale_fill_manual(values=kmeans.cols) +
  labs(y=expression(bold(paste("Cd34 log"[2], " Expression"))), x="Cluster") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=12, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14), 
        panel.grid.major=element_blank(),
        panel.border=element_blank(),
        axis.line=element_line(colour="black")) +
  guides(fill=guide_legend(title="Partition"),
         shape=guide_legend(title="Partition"))

ggsave(cd34.by.part,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Cd34_partition-boxplot.png",
       height=6.5/3, width=6.5, dpi=300)

cd34.by.part
```

```{r, echo=FALSE, warning=FALSE, message=FALSE}
fabp4.by.part <- ggplot(dc.merge, aes(x=ClusterLabel, y=Fabp4, fill=Kmeans)) +
  geom_boxplot(alpha=0.6) +
  geom_jitter(position=position_jitterdodge(jitter.width = 0.5),
              aes(shape=Kmeans), size=3) +
  theme_bw() +
  scale_shape_manual(values=c(21, 24)) +
  scale_fill_manual(values=kmeans.cols) +
  labs(y=expression(bold(paste("Fabp4 log"[2], " Expression"))), x="Cluster") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=12, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14), 
        panel.grid.major=element_blank(),
        panel.border=element_blank(),
        axis.line=element_line(colour="black")) +
  guides(fill=guide_legend(title="Partition"),
         shape=guide_legend(title="Partition"))

ggsave(fabp4.by.part,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Fabp4_partition-boxplot.png",
       height=6.5/3, width=6.5, dpi=300)

fabp4.by.part
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
pparg.by.part <- ggplot(dc.merge, aes(x=ClusterLabel, y=Pparg, fill=Kmeans)) +
  geom_boxplot(alpha=0.6) +
  geom_jitter(position=position_jitterdodge(jitter.width = 0.5),
              aes(shape=Kmeans), size=3) +
  theme_bw() +
  scale_shape_manual(values=c(21, 24)) +
  scale_fill_manual(values=kmeans.cols) +
  labs(y=expression(bold(paste("Pparg log"[2], " Expression"))), x="Cluster") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=12, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14), 
        panel.grid.major=element_blank(),
        panel.border=element_blank(),
        axis.line=element_line(colour="black")) +
  guides(fill=guide_legend(title="Partition"),
         shape=guide_legend(title="Partition"))

ggsave(pparg.by.part,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Pparg_partition-boxplot.png",
       height=6.5/3, width=6.5, dpi=300)

pparg.by.part
```



```{r, echo=FALSE, warning=FALSE, message=FALSE}
cebpa.by.part <- ggplot(dc.merge, aes(x=ClusterLabel, y=Cebpa, fill=Kmeans)) +
  geom_boxplot(alpha=0.6) +
  geom_jitter(position=position_jitterdodge(jitter.width = 0.5),
              aes(shape=Kmeans), size=3) +
  theme_bw() +
  scale_shape_manual(values=c(21, 24)) +
  scale_fill_manual(values=kmeans.cols) +
  labs(y=expression(bold(paste("Cebpa log"[2], " Expression"))), x="Cluster") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=12, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14), 
        panel.grid.major=element_blank(),
        panel.border=element_blank(),
        axis.line=element_line(colour="black")) +
  guides(fill=guide_legend(title="Partition"),
         shape=guide_legend(title="Partition"))

ggsave(cebpa.by.part,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Cebpa_partition-boxplot.png",
       height=6.5/3, width=6.5, dpi=300)

cebpa.by.part
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
itgb1.by.part <- ggplot(dc.merge, aes(x=ClusterLabel, y=Itgb1, fill=Kmeans)) +
  geom_boxplot(alpha=0.6) +
  geom_jitter(position=position_jitterdodge(jitter.width = 0.5),
              aes(shape=Kmeans), size=3) +
  theme_bw() +
  scale_shape_manual(values=c(21, 24)) +
  scale_fill_manual(values=kmeans.cols) +
  labs(y=expression(bold(paste("Cd29 log"[2], " Expression"))), x="Cluster") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=12, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14), 
        panel.grid.major=element_blank(),
        panel.border=element_blank(),
        axis.line=element_line(colour="black")) +
  guides(fill=guide_legend(title="Partition"),
         shape=guide_legend(title="Partition"))

ggsave(itgb1.by.part,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Cd29_partition-boxplot.png",
       height=6.5/3, width=6.5, dpi=300)

itgb1.by.part
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
ly6a.by.part <- ggplot(dc.merge, aes(x=ClusterLabel, y=Ly6a, fill=Kmeans)) +
  geom_boxplot(alpha=0.6) +
  geom_jitter(position=position_jitterdodge(jitter.width = 0.5),
              aes(shape=Kmeans), size=3) +
  theme_bw() +
  scale_shape_manual(values=c(21, 24)) +
  scale_fill_manual(values=kmeans.cols) +
  labs(y=expression(bold(paste("Sca1 log"[2], " Expression"))), x="Cluster") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=12, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14), 
        panel.grid.major=element_blank(),
        panel.border=element_blank(),
        axis.line=element_line(colour="black")) +
  guides(fill=guide_legend(title="Partition"),
         shape=guide_legend(title="Partition"))

ggsave(ly6a.by.part,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Sca1_partition-boxplot.png",
       height=6.5/3, width=6.5, dpi=300)

ly6a.by.part
```

```{r, echo=FALSE, warning=FALSE, message=FALSE}
pdgfra.by.part <- ggplot(dc.merge, aes(x=ClusterLabel, y=Pdgfra, fill=Kmeans)) +
  geom_boxplot(alpha=0.6) +
  geom_jitter(position=position_jitterdodge(jitter.width = 0.5),
              aes(shape=Kmeans), size=3) +
  theme_bw() +
  scale_shape_manual(values=c(21, 24)) +
  scale_fill_manual(values=kmeans.cols) +
  labs(y=expression(bold(paste("Pdgfra log"[2], " Expression"))), x="Cluster") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=12, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14), 
        panel.grid.major=element_blank(),
        panel.border=element_blank(),
        axis.line=element_line(colour="black")) +
  guides(fill=guide_legend(title="Partition"),
         shape=guide_legend(title="Partition"))

ggsave(pdgfra.by.part,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Pdgfra_partition-boxplot.png",
       height=6.5/3, width=6.5, dpi=300)

pdgfra.by.part
```

```{r, echo=FALSE, warning=FALSE, message=FALSE}
pdgfrb.by.part <- ggplot(dc.merge, aes(x=ClusterLabel, y=Pdgfrb, fill=Kmeans)) +
  geom_boxplot(alpha=0.6) +
  geom_jitter(position=position_jitterdodge(jitter.width = 0.5),
              aes(shape=Kmeans), size=3) +
  theme_bw() +
  scale_shape_manual(values=c(21, 24)) +
  scale_fill_manual(values=kmeans.cols) +
  labs(y=expression(bold(paste("Pdgfrb log"[2], " Expression"))), x="Cluster") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=12, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14), 
        panel.grid.major=element_blank(),
        panel.border=element_blank(),
        axis.line=element_line(colour="black")) +
  guides(fill=guide_legend(title="Partition"),
         shape=guide_legend(title="Partition"))

ggsave(pdgfrb.by.part,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Pdgfrb_partition-boxplot.png",
       height=6.5/3, width=6.5, dpi=300)

pdgfrb.by.part
```


```{r}
# remerge together the two partitions meta data
part1.dc.merge$Kmeans <- "1"
part2.dc.merge$Kmeans <- "2"

all.meta <- do.call(rbind.data.frame,
                    list("Part1"=part1.dc.merge[, intersect(colnames(part1.dc.merge), colnames(part2.dc.merge))],
                         "Part2"=part2.dc.merge[, intersect(colnames(part1.dc.merge), colnames(part2.dc.merge))]))

write.table(all.meta,
            "~/Dropbox/Adipose/Adipose-Meta_partition.tsv",
            sep="\t", quote=FALSE, row.names=FALSE)
```

Plot marker genes by different GFP sorting fractions.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
wt1.by.gfp <- ggplot(dc.merge, aes(x=Marker, y=Wt1, fill=Marker)) +
  geom_boxplot(alpha=0.6) +
  geom_jitter(position=position_jitterdodge(jitter.width = 0.5),
              size=3, shape=21) +
  theme_bw() +
  scale_fill_manual(values=gfp.cols) +
  labs(y=expression(bold(paste("Wt1 log"[2], " Expression"))), x="Sorting Fraction") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=12, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14), 
        panel.grid.major=element_blank(),
        panel.border=element_blank(),
        axis.line=element_line(colour="black")) +
  guides(fill=FALSE)

ggsave(wt1.by.gfp,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Wt1_partition-boxplot-byGFP.png",
       height=(6.5/3) + 0.5, width=6.5, dpi=300)

wt1.by.gfp
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
cd34.by.gfp <- ggplot(dc.merge, aes(x=Marker, y=Cd34, fill=Marker)) +
  geom_boxplot(alpha=0.6) +
  geom_jitter(position=position_jitterdodge(jitter.width = 0.5),
              size=3, shape=21) +
  theme_bw() +
  scale_fill_manual(values=gfp.cols) +
  labs(y=expression(bold(paste("Cd34 log"[2], " Expression"))), x="Sorting Fraction") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=12, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14), 
        panel.grid.major=element_blank(),
        panel.border=element_blank(),
        axis.line=element_line(colour="black")) +
  guides(fill=FALSE)

ggsave(cd34.by.gfp,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Cd34_partition-boxplot-byGFP.png",
       height=(6.5/3) + 0.5, width=6.5, dpi=300)

cd34.by.gfp
```

```{r, echo=FALSE, warning=FALSE, message=FALSE}
fabp4.by.gfp <- ggplot(dc.merge, aes(x=Marker, y=Fabp4, fill=Marker)) +
  geom_boxplot(alpha=0.6) +
  geom_jitter(position=position_jitterdodge(jitter.width = 0.5),
              shape=21, size=3) +
  theme_bw() +
  scale_fill_manual(values=gfp.cols) +
  labs(y=expression(bold(paste("Fabp4 log"[2], " Expression"))), x="Sorting Fraction") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=12, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14), 
        panel.grid.major=element_blank(),
        panel.border=element_blank(),
        axis.line=element_line(colour="black")) +
  guides(fill=FALSE)

ggsave(fabp4.by.gfp,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Fabp4_partition-boxplot-byGFP.png",
       height=(6.5/3) + 0.5, width=6.5, dpi=300)

fabp4.by.gfp
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
pparg.by.gfp <- ggplot(dc.merge, aes(x=Marker, y=Pparg, fill=Marker)) +
  geom_boxplot(alpha=0.6) +
  geom_jitter(position=position_jitterdodge(jitter.width = 0.5),
              shape=21, size=3) +
  theme_bw() +
  scale_fill_manual(values=gfp.cols) +
  labs(y=expression(bold(paste("Pparg log"[2], " Expression"))), x="Sorting Fraction") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=12, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14), 
        panel.grid.major=element_blank(),
        panel.border=element_blank(),
        axis.line=element_line(colour="black")) +
  guides(fill=FALSE)

ggsave(pparg.by.gfp,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Pparg_partition-boxplot-byGFP.png",
       height=(6.5/3) + 0.5, width=6.5, dpi=300)

pparg.by.gfp
```

```{r, echo=FALSE, warning=FALSE, message=FALSE}
itgb1.by.gfp <- ggplot(dc.merge, aes(x=Marker, y=Itgb1, fill=Marker)) +
  geom_boxplot(alpha=0.6) +
  geom_jitter(position=position_jitterdodge(jitter.width = 0.5),
              shape=21, size=3) +
  theme_bw() +
  scale_fill_manual(values=gfp.cols) +
  labs(y=expression(bold(paste("Cd29 log"[2], " Expression"))), x="Sorting Fraction") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=12, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14), 
        panel.grid.major=element_blank(),
        panel.border=element_blank(),
        axis.line=element_line(colour="black")) +
  guides(fill=FALSE)

ggsave(itgb1.by.gfp,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Cd29_partition-boxplot-byGFP.png",
       height=(6.5/3) + 0.5, width=6.5, dpi=300)

itgb1.by.gfp
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
ly6a.by.gfp <- ggplot(dc.merge, aes(x=Marker, y=Ly6a, fill=Marker)) +
  geom_boxplot(alpha=0.6) +
  geom_jitter(position=position_jitterdodge(jitter.width = 0.5),
              shape=21, size=3) +
  theme_bw() +
  scale_fill_manual(values=gfp.cols) +
  labs(y=expression(bold(paste("Sca1 log"[2], " Expression"))), x="Sorting Fraction") +
  theme(panel.grid=element_blank(),
        axis.text=element_text(size=14, colour="black"),
        axis.title=element_text(size=12, colour="black", face="bold"),
        legend.title=element_text(size=14, colour="black", face="bold"),
        legend.text=element_text(size=14), 
        panel.grid.major=element_blank(),
        panel.border=element_blank(),
        axis.line=element_line(colour="black")) +
  guides(fill=FALSE)

ggsave(ly6a.by.gfp,
       filename="~/Dropbox/Reports_Adipose/Adipogenesis2017/Figures/Sca1_partition-boxplot-byGFP.png",
       height=(6.5/3) + 0.5, width=6.5, dpi=300)

ly6a.by.gfp
```

